{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e216d8aa-ec15-4d54-a8ff-221efdb48c7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nfs/team205/is10/miniconda/envs/gene2probe/lib/python3.11/site-packages/Bio/Application/__init__.py:40: BiopythonDeprecationWarning: The Bio.Application modules and modules relying on it have been deprecated.\n",
      "\n",
      "Due to the on going maintenance burden of keeping command line application\n",
      "wrappers up to date, we have decided to deprecate and eventually remove these\n",
      "modules.\n",
      "\n",
      "We instead now recommend building your command line and invoking it directly\n",
      "with the subprocess module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import os \n",
    "import sys\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from gene2probe import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1dffb1b9-9ecc-490d-bea4-7aefb5a26028",
   "metadata": {},
   "outputs": [],
   "source": [
    "## specify inputs\n",
    "gene_ID = 'XIST'\n",
    "feature = 'exon' ## Whether to consider only exons / introns or full gene\n",
    "gtf = '/nfs/team205/is10/resources/ucsc_genomes/hg38.ncbiRefSeq.gtf' ## Gene annotation in gtf file\n",
    "## We recommend using RefSeq as this is manually curated and more likely to contain an isoform that is present across most cell types\n",
    "## Alternatively, one can filter based on RNA-seq data for a cell type/tissue of interest\n",
    "fasta = '/nfs/team205/is10/resources/ucsc_genomes/hg38.fa' ## Genome in fasta file\n",
    "snp_db = '/nfs/team205/is10/resources/hg38_dbSNP/hg38_snp151Common_chrX.txt' ## Database of known SNPs\n",
    "masked_regions_bed = '/nfs/team205/is10/resources/hg38_repeatmasker/hg38_rmsk.bed' ## bed file with coordinates to be excluded\n",
    "blast_db = '/nfs/team205/is10/projects/thytrans/data/gene2probe/blastdb/hg38_ncbiRefSeq_exons_db' ## Database of all human transcripts to blast against\n",
    "blast_exec_path = '/nfs/team205/is10/miniconda/envs/gene2probe/bin/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cba53a9d-2f6d-4611-8b97-4a6ce8ac75e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## additional parameters\n",
    "probe_length = 50\n",
    "split_nt = 25 ## Index of nucleotide to split the probe at (start of RHS) - leave None if splitting probe is not needed\n",
    "min_GC = 0.44\n",
    "max_GC = 0.72\n",
    "required_nts = {24: 'T'} ## Dictionary of index (0-based) for required nts - by default, 25th nucleotide must be a T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dfd7548b-e88a-4e7a-b8f6-ee59759468d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create output directory\n",
    "out_dir = '/nfs/team205/is10/projects/thytrans/data/gene2probe/' + gene_ID + '/'\n",
    "os.makedirs(out_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4bb522db-fc21-421c-b37e-8154a58ae63f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read gtf file\n",
    "gene_anno = read_gtf(gtf)\n",
    "## Extract regions corresponding to gene of interest (symbol: gene_name, Ensembl ID: gene_ID)\n",
    "gene_ids = gene_anno['attribute'].apply(extract_feature_from_gtf, feature='gene_name')\n",
    "roi = gene_anno.iloc[np.where(gene_ids == gene_ID)[0],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "227330b8-43f6-4e1a-b86f-e7b924b15b56",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>325814</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>transcript</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73852753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\";  ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325815</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73827984</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325816</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73829068</td>\n",
       "      <td>73829231</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325817</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73831066</td>\n",
       "      <td>73831274</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325818</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73833238</td>\n",
       "      <td>73833374</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325819</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73837440</td>\n",
       "      <td>73837503</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325820</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73841382</td>\n",
       "      <td>73852753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname                 source     feature     start       end score  \\\n",
       "325814    chrX  ncbiRefSeq.2022-10-28  transcript  73820651  73852753     .   \n",
       "325815    chrX  ncbiRefSeq.2022-10-28        exon  73820651  73827984     .   \n",
       "325816    chrX  ncbiRefSeq.2022-10-28        exon  73829068  73829231     .   \n",
       "325817    chrX  ncbiRefSeq.2022-10-28        exon  73831066  73831274     .   \n",
       "325818    chrX  ncbiRefSeq.2022-10-28        exon  73833238  73833374     .   \n",
       "325819    chrX  ncbiRefSeq.2022-10-28        exon  73837440  73837503     .   \n",
       "325820    chrX  ncbiRefSeq.2022-10-28        exon  73841382  73852753     .   \n",
       "\n",
       "       strand frame                                          attribute  \n",
       "325814      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\";  ...  \n",
       "325815      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325816      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325817      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325818      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325819      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325820      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "95b80de7-eb8a-4cb5-8fb0-cb6bf2fe16c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Subset to feature type of interest\n",
    "roi = roi[roi['feature']==feature]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6316f862-39da-401d-9fe8-201c3d3d2cc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>325815</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73827984</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325816</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73829068</td>\n",
       "      <td>73829231</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325817</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73831066</td>\n",
       "      <td>73831274</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325818</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73833238</td>\n",
       "      <td>73833374</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325819</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73837440</td>\n",
       "      <td>73837503</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325820</th>\n",
       "      <td>chrX</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>73841382</td>\n",
       "      <td>73852753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname                 source feature     start       end score  \\\n",
       "325815    chrX  ncbiRefSeq.2022-10-28    exon  73820651  73827984     .   \n",
       "325816    chrX  ncbiRefSeq.2022-10-28    exon  73829068  73829231     .   \n",
       "325817    chrX  ncbiRefSeq.2022-10-28    exon  73831066  73831274     .   \n",
       "325818    chrX  ncbiRefSeq.2022-10-28    exon  73833238  73833374     .   \n",
       "325819    chrX  ncbiRefSeq.2022-10-28    exon  73837440  73837503     .   \n",
       "325820    chrX  ncbiRefSeq.2022-10-28    exon  73841382  73852753     .   \n",
       "\n",
       "       strand frame                                          attribute  \n",
       "325815      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325816      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325817      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325818      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325819      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  \n",
       "325820      -     .  gene_id \"XIST\"; transcript_id \"NR_001564.2\"; e...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "72223af7-a0fc-4845-b10d-622a09da87cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nfs/team205/is10/projects/thytrans/scripts/008_visiumHD/gene2probe.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_bed['start'] = df_bed['start'].astype(int) - 1\n"
     ]
    }
   ],
   "source": [
    "## Convert to bed-style df\n",
    "roi_bed = gtf_2_bed(roi, name_pref = (gene_ID + '_'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "60894d08-da20-462e-beb6-b2e672db60f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73827984</td>\n",
       "      <td>XIST_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73829067</td>\n",
       "      <td>73829231</td>\n",
       "      <td>XIST_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73831065</td>\n",
       "      <td>73831274</td>\n",
       "      <td>XIST_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73833237</td>\n",
       "      <td>73833374</td>\n",
       "      <td>XIST_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73837439</td>\n",
       "      <td>73837503</td>\n",
       "      <td>XIST_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73841381</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     start       end    name score strand\n",
       "0    chrX  73820650  73827984  XIST_0     .      -\n",
       "1    chrX  73829067  73829231  XIST_1     .      -\n",
       "2    chrX  73831065  73831274  XIST_2     .      -\n",
       "3    chrX  73833237  73833374  XIST_3     .      -\n",
       "4    chrX  73837439  73837503  XIST_4     .      -\n",
       "5    chrX  73841381  73852753  XIST_5     .      -"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "60f606b4-d196-4382-bcae-f64364dc5768",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi_kmers = generate_50mers(roi_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c981be58-decf-4f36-bce9-1efabaa02627",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18981</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18982</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18983</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852701</td>\n",
       "      <td>73852751</td>\n",
       "      <td>XIST_5_11320</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18984</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852702</td>\n",
       "      <td>73852752</td>\n",
       "      <td>XIST_5_11321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18985</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852703</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5_11322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18986 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -\n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -\n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -\n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -\n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -\n",
       "...       ...       ...       ...           ...   ...    ...\n",
       "18981    chrX  73852699  73852749  XIST_5_11318     .      -\n",
       "18982    chrX  73852700  73852750  XIST_5_11319     .      -\n",
       "18983    chrX  73852701  73852751  XIST_5_11320     .      -\n",
       "18984    chrX  73852702  73852752  XIST_5_11321     .      -\n",
       "18985    chrX  73852703  73852753  XIST_5_11322     .      -\n",
       "\n",
       "[18986 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "69c58199-798b-45c7-8799-bcd30e8dfb3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export unfiltered\n",
    "roi_kmers.to_csv((out_dir + 'kmers_all.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4aa563ab-1e10-4e63-b8a3-9fb266a70ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Convert to pybedtools\n",
    "roi_kmers_bed = pybedtools.BedTool.from_dataframe(roi_kmers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "48005706-0987-41ff-af44-77b08811b48d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove masked regions\n",
    "if masked_regions_bed is not None:\n",
    "    ## Read rmsk\n",
    "    rmsk = pybedtools.BedTool(masked_regions_bed)\n",
    "    ## Remove any kmers overlapping any masked regions\n",
    "    roi_kmers_bed = roi_kmers_bed.intersect(rmsk, v=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "179a7d62-73b5-4beb-9f14-700583fd3a9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi_kmers_filtered = roi_kmers_bed.to_dataframe(names=['seqname', 'start', 'end', 'name', 'score', 'strand'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a08f0e1a-6bfe-4d8f-9a22-ff2baaf4cbb4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16853</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16854</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16855</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852701</td>\n",
       "      <td>73852751</td>\n",
       "      <td>XIST_5_11320</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16856</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852702</td>\n",
       "      <td>73852752</td>\n",
       "      <td>XIST_5_11321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16857</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852703</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5_11322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16858 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -\n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -\n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -\n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -\n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -\n",
       "...       ...       ...       ...           ...   ...    ...\n",
       "16853    chrX  73852699  73852749  XIST_5_11318     .      -\n",
       "16854    chrX  73852700  73852750  XIST_5_11319     .      -\n",
       "16855    chrX  73852701  73852751  XIST_5_11320     .      -\n",
       "16856    chrX  73852702  73852752  XIST_5_11321     .      -\n",
       "16857    chrX  73852703  73852753  XIST_5_11322     .      -\n",
       "\n",
       "[16858 rows x 6 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_kmers_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0b801075-820c-4df3-becb-0082c24702b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Same with SNPs\n",
    "if snp_db is not None:\n",
    "    ## Read SNP file (by chromosome as this can be quite large)\n",
    "    snps = pd.read_csv(snp_db, sep='\\t')  \n",
    "    snps_bed = pybedtools.BedTool.from_dataframe(snps[['#chrom', 'chromStart', 'chromEnd']])\n",
    "    ## Remove any kmers overlapping any masked regions\n",
    "    roi_kmers_bed = roi_kmers_bed.intersect(snps_bed, v=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d732a6ee-c089-4e38-9e42-075d95b63c07",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi_kmers_filtered = roi_kmers_bed.to_dataframe(names=['seqname', 'start', 'end', 'name', 'score', 'strand'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "969e23e8-4a03-44a8-a25a-110c66ed06f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15728</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15729</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15730</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852701</td>\n",
       "      <td>73852751</td>\n",
       "      <td>XIST_5_11320</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15731</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852702</td>\n",
       "      <td>73852752</td>\n",
       "      <td>XIST_5_11321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15732</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852703</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5_11322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15733 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -\n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -\n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -\n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -\n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -\n",
       "...       ...       ...       ...           ...   ...    ...\n",
       "15728    chrX  73852699  73852749  XIST_5_11318     .      -\n",
       "15729    chrX  73852700  73852750  XIST_5_11319     .      -\n",
       "15730    chrX  73852701  73852751  XIST_5_11320     .      -\n",
       "15731    chrX  73852702  73852752  XIST_5_11321     .      -\n",
       "15732    chrX  73852703  73852753  XIST_5_11322     .      -\n",
       "\n",
       "[15733 rows x 6 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_kmers_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d0bb0340-37b0-4c07-b713-ba13e97fd25a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get DNA for transcript and probe (reverse complement)\n",
    "## Simultaneously we can monitor GC content and count the longest homopolymer stretch\n",
    "roi_kmers_seq = roi_kmers_bed.sequence(fi=fasta, s=True) \n",
    "\n",
    "kmer_coord = []\n",
    "transcript_seq = []\n",
    "probe_seq = []\n",
    "gc_content = []\n",
    "gc_content_lhs = []\n",
    "gc_content_rhs = []\n",
    "longest_homopol = []\n",
    "with open(roi_kmers_seq.seqfn) as f:\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            kmer_coord.append(line.strip().replace('>', ''))\n",
    "        else:\n",
    "            ## Get sequence\n",
    "            kmer_seq = Seq(line.strip())\n",
    "            transcript_seq.append(str(kmer_seq))\n",
    "            ## Reverse complement - this is the sequence of the probe\n",
    "            probe_seq.append(str(kmer_seq.reverse_complement()))\n",
    "            ## Estimate GC content\n",
    "            gc_content.append(gc_fraction(kmer_seq))\n",
    "            gc_content_rhs.append(gc_fraction(kmer_seq[0:split_nt])) ## RHS of the probe is LHS of the target - GC content is the same\n",
    "            gc_content_lhs.append(gc_fraction(kmer_seq[split_nt: probe_length])) ## RHS of the probe is LHS of the target - GC content is the same\n",
    "            ## Estimate longest homopolymer\n",
    "            longest_homopol.append(get_longest_homopolymer(kmer_seq))\n",
    "\n",
    "## Incorporate sequences and stats into the dataframe\n",
    "roi_kmers_filtered['kmer_coord'] = kmer_coord\n",
    "roi_kmers_filtered['transcript_seq'] = transcript_seq\n",
    "roi_kmers_filtered['probe_seq'] = probe_seq\n",
    "roi_kmers_filtered['GC_content_full'] = gc_content\n",
    "roi_kmers_filtered['GC_content_LHS'] = gc_content_lhs\n",
    "roi_kmers_filtered['GC_content_RHS'] = gc_content_rhs\n",
    "roi_kmers_filtered['longest_homopolymer'] = longest_homopol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9a98f6eb-dcdd-4406-9c7c-8e3dd37e18e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820650-73820700(-)</td>\n",
       "      <td>GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...</td>\n",
       "      <td>CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.40</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820651-73820701(-)</td>\n",
       "      <td>TGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAA...</td>\n",
       "      <td>AAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTC...</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.40</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820652-73820702(-)</td>\n",
       "      <td>CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...</td>\n",
       "      <td>AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820653-73820703(-)</td>\n",
       "      <td>ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...</td>\n",
       "      <td>AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820654-73820704(-)</td>\n",
       "      <td>AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...</td>\n",
       "      <td>GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15728</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852699-73852749(-)</td>\n",
       "      <td>CAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTAC...</td>\n",
       "      <td>GAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15729</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852700-73852750(-)</td>\n",
       "      <td>TCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTA...</td>\n",
       "      <td>AGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAA...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15730</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852701</td>\n",
       "      <td>73852751</td>\n",
       "      <td>XIST_5_11320</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852701-73852751(-)</td>\n",
       "      <td>TTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTT...</td>\n",
       "      <td>GAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAAC...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.40</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15731</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852702</td>\n",
       "      <td>73852752</td>\n",
       "      <td>XIST_5_11321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852702-73852752(-)</td>\n",
       "      <td>CTTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCT...</td>\n",
       "      <td>AGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAACT...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.40</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15732</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852703</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5_11322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852703-73852753(-)</td>\n",
       "      <td>CCTTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTC...</td>\n",
       "      <td>GTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAACTG...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15733 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand  \\\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -   \n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -   \n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -   \n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -   \n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -   \n",
       "...       ...       ...       ...           ...   ...    ...   \n",
       "15728    chrX  73852699  73852749  XIST_5_11318     .      -   \n",
       "15729    chrX  73852700  73852750  XIST_5_11319     .      -   \n",
       "15730    chrX  73852701  73852751  XIST_5_11320     .      -   \n",
       "15731    chrX  73852702  73852752  XIST_5_11321     .      -   \n",
       "15732    chrX  73852703  73852753  XIST_5_11322     .      -   \n",
       "\n",
       "                      kmer_coord  \\\n",
       "0      chrX:73820650-73820700(-)   \n",
       "1      chrX:73820651-73820701(-)   \n",
       "2      chrX:73820652-73820702(-)   \n",
       "3      chrX:73820653-73820703(-)   \n",
       "4      chrX:73820654-73820704(-)   \n",
       "...                          ...   \n",
       "15728  chrX:73852699-73852749(-)   \n",
       "15729  chrX:73852700-73852750(-)   \n",
       "15730  chrX:73852701-73852751(-)   \n",
       "15731  chrX:73852702-73852752(-)   \n",
       "15732  chrX:73852703-73852753(-)   \n",
       "\n",
       "                                          transcript_seq  \\\n",
       "0      GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...   \n",
       "1      TGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAA...   \n",
       "2      CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...   \n",
       "3      ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...   \n",
       "4      AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...   \n",
       "...                                                  ...   \n",
       "15728  CAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTAC...   \n",
       "15729  TCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTA...   \n",
       "15730  TTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTT...   \n",
       "15731  CTTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCT...   \n",
       "15732  CCTTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTC...   \n",
       "\n",
       "                                               probe_seq  GC_content_full  \\\n",
       "0      CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...             0.30   \n",
       "1      AAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTC...             0.28   \n",
       "2      AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...             0.30   \n",
       "3      AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...             0.30   \n",
       "4      GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...             0.30   \n",
       "...                                                  ...              ...   \n",
       "15728  GAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGA...             0.46   \n",
       "15729  AGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAA...             0.44   \n",
       "15730  GAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAAC...             0.44   \n",
       "15731  AGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAACT...             0.44   \n",
       "15732  GTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAACTG...             0.46   \n",
       "\n",
       "       GC_content_LHS  GC_content_RHS  longest_homopolymer  \n",
       "0                0.20            0.40                    4  \n",
       "1                0.16            0.40                    4  \n",
       "2                0.16            0.44                    4  \n",
       "3                0.16            0.44                    4  \n",
       "4                0.16            0.44                    4  \n",
       "...               ...             ...                  ...  \n",
       "15728            0.44            0.48                    3  \n",
       "15729            0.44            0.44                    3  \n",
       "15730            0.48            0.40                    3  \n",
       "15731            0.48            0.40                    3  \n",
       "15732            0.48            0.44                    3  \n",
       "\n",
       "[15733 rows x 13 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_kmers_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "84c9ff67-c6d2-4f50-9b2b-2725086f7b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Mark probes matching nucleotide requirement:\n",
    "has_required_nt = [True for i in range(roi_kmers_filtered.shape[0])]\n",
    "for nt_idx in required_nts.keys():\n",
    "    for i in range(roi_kmers_filtered.shape[0]):\n",
    "        if roi_kmers_filtered['probe_seq'][i][nt_idx] != required_nts[nt_idx]:\n",
    "            has_required_nt[i] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "442e5216-ddfe-4de3-a5b5-37b68baa1cdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi_kmers_filtered['has_required_nts'] = has_required_nt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6fac63f8-2330-4ddd-b7f6-7012a7eabe06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "has_required_nts\n",
       "False    11549\n",
       "True      4184\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_kmers_filtered['has_required_nts'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a937f445-2786-400d-ae11-38233957bbe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export kmers before filtering\n",
    "roi_kmers_filtered.to_csv((out_dir + 'kmers_candidates_unfiltered.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0fa42cd0-1b76-4948-b33f-ef5d057d9f01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820650-73820700(-)</td>\n",
       "      <td>GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...</td>\n",
       "      <td>CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.40</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820651-73820701(-)</td>\n",
       "      <td>TGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAA...</td>\n",
       "      <td>AAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTC...</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.40</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820652-73820702(-)</td>\n",
       "      <td>CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...</td>\n",
       "      <td>AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820653-73820703(-)</td>\n",
       "      <td>ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...</td>\n",
       "      <td>AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820654-73820704(-)</td>\n",
       "      <td>AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...</td>\n",
       "      <td>GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15728</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852699-73852749(-)</td>\n",
       "      <td>CAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTAC...</td>\n",
       "      <td>GAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15729</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852700-73852750(-)</td>\n",
       "      <td>TCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTA...</td>\n",
       "      <td>AGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAA...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15730</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852701</td>\n",
       "      <td>73852751</td>\n",
       "      <td>XIST_5_11320</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852701-73852751(-)</td>\n",
       "      <td>TTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTT...</td>\n",
       "      <td>GAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAAC...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.40</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15731</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852702</td>\n",
       "      <td>73852752</td>\n",
       "      <td>XIST_5_11321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852702-73852752(-)</td>\n",
       "      <td>CTTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCT...</td>\n",
       "      <td>AGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAACT...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.40</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15732</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852703</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5_11322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852703-73852753(-)</td>\n",
       "      <td>CCTTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTC...</td>\n",
       "      <td>GTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAACTG...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15733 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand  \\\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -   \n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -   \n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -   \n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -   \n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -   \n",
       "...       ...       ...       ...           ...   ...    ...   \n",
       "15728    chrX  73852699  73852749  XIST_5_11318     .      -   \n",
       "15729    chrX  73852700  73852750  XIST_5_11319     .      -   \n",
       "15730    chrX  73852701  73852751  XIST_5_11320     .      -   \n",
       "15731    chrX  73852702  73852752  XIST_5_11321     .      -   \n",
       "15732    chrX  73852703  73852753  XIST_5_11322     .      -   \n",
       "\n",
       "                      kmer_coord  \\\n",
       "0      chrX:73820650-73820700(-)   \n",
       "1      chrX:73820651-73820701(-)   \n",
       "2      chrX:73820652-73820702(-)   \n",
       "3      chrX:73820653-73820703(-)   \n",
       "4      chrX:73820654-73820704(-)   \n",
       "...                          ...   \n",
       "15728  chrX:73852699-73852749(-)   \n",
       "15729  chrX:73852700-73852750(-)   \n",
       "15730  chrX:73852701-73852751(-)   \n",
       "15731  chrX:73852702-73852752(-)   \n",
       "15732  chrX:73852703-73852753(-)   \n",
       "\n",
       "                                          transcript_seq  \\\n",
       "0      GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...   \n",
       "1      TGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAA...   \n",
       "2      CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...   \n",
       "3      ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...   \n",
       "4      AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...   \n",
       "...                                                  ...   \n",
       "15728  CAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTAC...   \n",
       "15729  TCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTA...   \n",
       "15730  TTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTT...   \n",
       "15731  CTTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCT...   \n",
       "15732  CCTTCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTC...   \n",
       "\n",
       "                                               probe_seq  GC_content_full  \\\n",
       "0      CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...             0.30   \n",
       "1      AAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTC...             0.28   \n",
       "2      AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...             0.30   \n",
       "3      AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...             0.30   \n",
       "4      GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...             0.30   \n",
       "...                                                  ...              ...   \n",
       "15728  GAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGA...             0.46   \n",
       "15729  AGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAA...             0.44   \n",
       "15730  GAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAAC...             0.44   \n",
       "15731  AGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAACT...             0.44   \n",
       "15732  GTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAACTG...             0.46   \n",
       "\n",
       "       GC_content_LHS  GC_content_RHS  longest_homopolymer  has_required_nts  \n",
       "0                0.20            0.40                    4              True  \n",
       "1                0.16            0.40                    4             False  \n",
       "2                0.16            0.44                    4              True  \n",
       "3                0.16            0.44                    4              True  \n",
       "4                0.16            0.44                    4              True  \n",
       "...               ...             ...                  ...               ...  \n",
       "15728            0.44            0.48                    3             False  \n",
       "15729            0.44            0.44                    3             False  \n",
       "15730            0.48            0.40                    3             False  \n",
       "15731            0.48            0.40                    3             False  \n",
       "15732            0.48            0.44                    3             False  \n",
       "\n",
       "[15733 rows x 14 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_kmers_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7791ef9f-63ea-45ee-b459-cd9db424d764",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter for GC content\n",
    "roi_kmers_filtered = roi_kmers_filtered[(roi_kmers_filtered['GC_content_full'] > min_GC) & \n",
    "(roi_kmers_filtered['GC_content_LHS'] > min_GC) & (roi_kmers_filtered['GC_content_RHS'] > min_GC) & \n",
    "(roi_kmers_filtered['GC_content_full'] < max_GC) & \n",
    "(roi_kmers_filtered['GC_content_LHS'] < max_GC) & (roi_kmers_filtered['GC_content_RHS'] < max_GC)].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1ed1ae93-3577-4a85-a65a-9e506d4c317a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821069</td>\n",
       "      <td>73821119</td>\n",
       "      <td>XIST_0_419</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821069-73821119(-)</td>\n",
       "      <td>CTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGATTATT...</td>\n",
       "      <td>TTCCAATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTAC...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821070</td>\n",
       "      <td>73821120</td>\n",
       "      <td>XIST_0_420</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821070-73821120(-)</td>\n",
       "      <td>ACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGATTAT...</td>\n",
       "      <td>TCCAATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACT...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821071</td>\n",
       "      <td>73821121</td>\n",
       "      <td>XIST_0_421</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821071-73821121(-)</td>\n",
       "      <td>GACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGATTA...</td>\n",
       "      <td>CCAATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTG...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821072</td>\n",
       "      <td>73821122</td>\n",
       "      <td>XIST_0_422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821072-73821122(-)</td>\n",
       "      <td>GGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGATT...</td>\n",
       "      <td>CAATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGA...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821073</td>\n",
       "      <td>73821123</td>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821073-73821123(-)</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...</td>\n",
       "      <td>AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2491</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852691</td>\n",
       "      <td>73852741</td>\n",
       "      <td>XIST_5_11310</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852691-73852741(-)</td>\n",
       "      <td>AAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGG...</td>\n",
       "      <td>AGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCG...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2492</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852692</td>\n",
       "      <td>73852742</td>\n",
       "      <td>XIST_5_11311</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852692-73852742(-)</td>\n",
       "      <td>TAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCG...</td>\n",
       "      <td>GCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGC...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2493</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2494</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852694</td>\n",
       "      <td>73852744</td>\n",
       "      <td>XIST_5_11313</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852694-73852744(-)</td>\n",
       "      <td>CTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCT...</td>\n",
       "      <td>CCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852695</td>\n",
       "      <td>73852745</td>\n",
       "      <td>XIST_5_11314</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852695-73852745(-)</td>\n",
       "      <td>TCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTC...</td>\n",
       "      <td>CCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTT...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2496 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqname     start       end          name score strand  \\\n",
       "0       chrX  73821069  73821119    XIST_0_419     .      -   \n",
       "1       chrX  73821070  73821120    XIST_0_420     .      -   \n",
       "2       chrX  73821071  73821121    XIST_0_421     .      -   \n",
       "3       chrX  73821072  73821122    XIST_0_422     .      -   \n",
       "4       chrX  73821073  73821123    XIST_0_423     .      -   \n",
       "...      ...       ...       ...           ...   ...    ...   \n",
       "2491    chrX  73852691  73852741  XIST_5_11310     .      -   \n",
       "2492    chrX  73852692  73852742  XIST_5_11311     .      -   \n",
       "2493    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "2494    chrX  73852694  73852744  XIST_5_11313     .      -   \n",
       "2495    chrX  73852695  73852745  XIST_5_11314     .      -   \n",
       "\n",
       "                     kmer_coord  \\\n",
       "0     chrX:73821069-73821119(-)   \n",
       "1     chrX:73821070-73821120(-)   \n",
       "2     chrX:73821071-73821121(-)   \n",
       "3     chrX:73821072-73821122(-)   \n",
       "4     chrX:73821073-73821123(-)   \n",
       "...                         ...   \n",
       "2491  chrX:73852691-73852741(-)   \n",
       "2492  chrX:73852692-73852742(-)   \n",
       "2493  chrX:73852693-73852743(-)   \n",
       "2494  chrX:73852694-73852744(-)   \n",
       "2495  chrX:73852695-73852745(-)   \n",
       "\n",
       "                                         transcript_seq  \\\n",
       "0     CTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGATTATT...   \n",
       "1     ACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGATTAT...   \n",
       "2     GACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGATTA...   \n",
       "3     GGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGATT...   \n",
       "4     TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...   \n",
       "...                                                 ...   \n",
       "2491  AAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGG...   \n",
       "2492  TAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCG...   \n",
       "2493  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "2494  CTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCT...   \n",
       "2495  TCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTC...   \n",
       "\n",
       "                                              probe_seq  GC_content_full  \\\n",
       "0     TTCCAATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTAC...             0.48   \n",
       "1     TCCAATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACT...             0.48   \n",
       "2     CCAATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTG...             0.50   \n",
       "3     CAATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGA...             0.50   \n",
       "4     AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...             0.48   \n",
       "...                                                 ...              ...   \n",
       "2491  AGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCG...             0.52   \n",
       "2492  GCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGC...             0.52   \n",
       "2493  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "2494  CCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTT...             0.50   \n",
       "2495  CCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTT...             0.48   \n",
       "\n",
       "      GC_content_LHS  GC_content_RHS  longest_homopolymer  has_required_nts  \n",
       "0               0.48            0.48                    4             False  \n",
       "1               0.48            0.48                    4             False  \n",
       "2               0.52            0.48                    4             False  \n",
       "3               0.52            0.48                    4             False  \n",
       "4               0.48            0.48                    4              True  \n",
       "...              ...             ...                  ...               ...  \n",
       "2491            0.48            0.56                    4             False  \n",
       "2492            0.52            0.52                    4             False  \n",
       "2493            0.48            0.52                    4              True  \n",
       "2494            0.48            0.52                    3             False  \n",
       "2495            0.48            0.48                    3             False  \n",
       "\n",
       "[2496 rows x 14 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_kmers_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3a8d3d13-5233-4a90-827e-020d53b4dfdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter for required nucleotides\n",
    "roi_kmers_filtered = roi_kmers_filtered[roi_kmers_filtered['has_required_nts']==True].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1a7eb377-dce8-4e92-aa44-bbdec9172d27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821073</td>\n",
       "      <td>73821123</td>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821073-73821123(-)</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...</td>\n",
       "      <td>AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821085</td>\n",
       "      <td>73821135</td>\n",
       "      <td>XIST_0_435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821085-73821135(-)</td>\n",
       "      <td>TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...</td>\n",
       "      <td>CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821087</td>\n",
       "      <td>73821137</td>\n",
       "      <td>XIST_0_437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821087-73821137(-)</td>\n",
       "      <td>TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...</td>\n",
       "      <td>CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821088</td>\n",
       "      <td>73821138</td>\n",
       "      <td>XIST_0_438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821088-73821138(-)</td>\n",
       "      <td>ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...</td>\n",
       "      <td>ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821091</td>\n",
       "      <td>73821141</td>\n",
       "      <td>XIST_0_441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821091-73821141(-)</td>\n",
       "      <td>TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...</td>\n",
       "      <td>CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852654</td>\n",
       "      <td>73852704</td>\n",
       "      <td>XIST_5_11273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852654-73852704(-)</td>\n",
       "      <td>CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...</td>\n",
       "      <td>CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.64</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852663</td>\n",
       "      <td>73852713</td>\n",
       "      <td>XIST_5_11282</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852663-73852713(-)</td>\n",
       "      <td>TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...</td>\n",
       "      <td>CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>526 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73821073  73821123    XIST_0_423     .      -   \n",
       "1      chrX  73821085  73821135    XIST_0_435     .      -   \n",
       "2      chrX  73821087  73821137    XIST_0_437     .      -   \n",
       "3      chrX  73821088  73821138    XIST_0_438     .      -   \n",
       "4      chrX  73821091  73821141    XIST_0_441     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "521    chrX  73852654  73852704  XIST_5_11273     .      -   \n",
       "522    chrX  73852663  73852713  XIST_5_11282     .      -   \n",
       "523    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "524    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "525    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73821073-73821123(-)   \n",
       "1    chrX:73821085-73821135(-)   \n",
       "2    chrX:73821087-73821137(-)   \n",
       "3    chrX:73821088-73821138(-)   \n",
       "4    chrX:73821091-73821141(-)   \n",
       "..                         ...   \n",
       "521  chrX:73852654-73852704(-)   \n",
       "522  chrX:73852663-73852713(-)   \n",
       "523  chrX:73852687-73852737(-)   \n",
       "524  chrX:73852689-73852739(-)   \n",
       "525  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...   \n",
       "1    TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...   \n",
       "2    TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...   \n",
       "3    ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...   \n",
       "4    TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...   \n",
       "..                                                 ...   \n",
       "521  CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...   \n",
       "522  TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...   \n",
       "523  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "524  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "525  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...             0.48   \n",
       "1    CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...             0.54   \n",
       "2    CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...             0.52   \n",
       "3    ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...             0.50   \n",
       "4    CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...             0.50   \n",
       "..                                                 ...              ...   \n",
       "521  CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...             0.58   \n",
       "522  CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...             0.48   \n",
       "523  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "524  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "525  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "     GC_content_LHS  GC_content_RHS  longest_homopolymer  has_required_nts  \n",
       "0              0.48            0.48                    4              True  \n",
       "1              0.60            0.48                    4              True  \n",
       "2              0.56            0.48                    4              True  \n",
       "3              0.52            0.48                    4              True  \n",
       "4              0.52            0.48                    4              True  \n",
       "..              ...             ...                  ...               ...  \n",
       "521            0.52            0.64                    4              True  \n",
       "522            0.48            0.48                    4              True  \n",
       "523            0.48            0.60                    4              True  \n",
       "524            0.48            0.64                    4              True  \n",
       "525            0.48            0.52                    4              True  \n",
       "\n",
       "[526 rows x 14 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Candidate kmers\n",
    "roi_kmers_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e482ac72-5ed2-46dd-a8bb-1950d1a8aa14",
   "metadata": {},
   "outputs": [],
   "source": [
    "roi_kmers_filtered.to_csv((out_dir + 'kmers_candidates_filtered.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3ace2aec-f433-49c6-90c6-8a971232be41",
   "metadata": {},
   "outputs": [],
   "source": [
    "## We are exporting the transcript sequence as that's the one that has to be blasted against the human transcriptome\n",
    "write_fasta(roi_kmers_filtered['name'], roi_kmers_filtered['transcript_seq'], (out_dir + 'kmers_candidates_filtered_transcript_seqs.fa'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0f5255b6-3eb9-4912-a75a-c2dcf6cad93a",
   "metadata": {},
   "outputs": [],
   "source": [
    "command = [\n",
    "    (blast_exec_path + 'blastn'),\n",
    "    '-query', (out_dir + 'kmers_candidates_filtered_transcript_seqs.fa'), ## Sequences of targeted regions\n",
    "    '-db', blast_db,                     # BLAST database\n",
    "    '-out', (out_dir + 'kmers_candidates_filtered_blast_output.txt'),  # Output file\n",
    "    '-outfmt', '6',\n",
    "    '-strand', 'plus',\n",
    "    '-evalue', '1e-3',  # More lenient e-value to detect potential off-targets\n",
    "    '-dust', 'no'       # Turn off low-complexity filter\n",
    "]\n",
    "\n",
    "# Run the command\n",
    "result = subprocess.run(command, capture_output=True, text=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0cb79219-58bb-4344-b02d-4eef45ff18e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read in the blast output\n",
    "blast_columns = [\n",
    "    \"name\",   # Query Seq-id\n",
    "    \"sseqid\",   # Subject Seq-id\n",
    "    \"pident\",   # Percentage of identical matches\n",
    "    \"length\",   # Alignment length\n",
    "    \"mismatch\", # Number of mismatches\n",
    "    \"gapopen\",  # Number of gap openings\n",
    "    \"qstart\",   # Start of alignment in query\n",
    "    \"qend\",     # End of alignment in query\n",
    "    \"sstart\",   # Start of alignment in subject\n",
    "    \"send\",     # End of alignment in subject\n",
    "    \"evalue\",   # Expect value\n",
    "    \"bitscore\"  # Bit score\n",
    "]\n",
    "\n",
    "blast_res = pd.read_csv(\n",
    "    (out_dir + 'kmers_candidates_filtered_blast_output.txt'),\n",
    "    sep='\\t', \n",
    "    header=None, \n",
    "    names=blast_columns\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6aae54ac-97a5-426f-a88f-8352885a6493",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Extract the gene ID\n",
    "blast_res['sgeneid'] = blast_res['sseqid'].str.split('::').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "66bd2f88-1ce9-4953-b238-4fbcb1349a2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sgeneid\n",
       "XIST    611\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_res['sgeneid'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "b219ece2-679e-4132-9c2a-50a5f21f5974",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    6.110000e+02\n",
       "mean     1.116134e-12\n",
       "std      1.632751e-11\n",
       "min      1.300000e-18\n",
       "25%      1.300000e-18\n",
       "50%      1.300000e-18\n",
       "75%      1.300000e-18\n",
       "max      2.840000e-10\n",
       "Name: evalue, dtype: float64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_res['evalue'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f2d394e1-35a7-461a-b94e-fa9db71984b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "offtargets = blast_res['name'][blast_res['sgeneid']!=gene_ID].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "73ace4d0-5850-47e7-9770-3358b78237c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "offtargets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b25ebea4-fa7c-44f9-b7f0-e6fb32149341",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove off targets\n",
    "roi_kmers_filtered = roi_kmers_filtered[roi_kmers_filtered['name'].isin(offtargets)==False].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1a1d18ca-0824-4094-94f5-d4a5a0645366",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821073</td>\n",
       "      <td>73821123</td>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821073-73821123(-)</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...</td>\n",
       "      <td>AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821085</td>\n",
       "      <td>73821135</td>\n",
       "      <td>XIST_0_435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821085-73821135(-)</td>\n",
       "      <td>TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...</td>\n",
       "      <td>CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821087</td>\n",
       "      <td>73821137</td>\n",
       "      <td>XIST_0_437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821087-73821137(-)</td>\n",
       "      <td>TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...</td>\n",
       "      <td>CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821088</td>\n",
       "      <td>73821138</td>\n",
       "      <td>XIST_0_438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821088-73821138(-)</td>\n",
       "      <td>ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...</td>\n",
       "      <td>ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821091</td>\n",
       "      <td>73821141</td>\n",
       "      <td>XIST_0_441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821091-73821141(-)</td>\n",
       "      <td>TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...</td>\n",
       "      <td>CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852654</td>\n",
       "      <td>73852704</td>\n",
       "      <td>XIST_5_11273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852654-73852704(-)</td>\n",
       "      <td>CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...</td>\n",
       "      <td>CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.64</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852663</td>\n",
       "      <td>73852713</td>\n",
       "      <td>XIST_5_11282</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852663-73852713(-)</td>\n",
       "      <td>TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...</td>\n",
       "      <td>CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>526 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73821073  73821123    XIST_0_423     .      -   \n",
       "1      chrX  73821085  73821135    XIST_0_435     .      -   \n",
       "2      chrX  73821087  73821137    XIST_0_437     .      -   \n",
       "3      chrX  73821088  73821138    XIST_0_438     .      -   \n",
       "4      chrX  73821091  73821141    XIST_0_441     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "521    chrX  73852654  73852704  XIST_5_11273     .      -   \n",
       "522    chrX  73852663  73852713  XIST_5_11282     .      -   \n",
       "523    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "524    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "525    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73821073-73821123(-)   \n",
       "1    chrX:73821085-73821135(-)   \n",
       "2    chrX:73821087-73821137(-)   \n",
       "3    chrX:73821088-73821138(-)   \n",
       "4    chrX:73821091-73821141(-)   \n",
       "..                         ...   \n",
       "521  chrX:73852654-73852704(-)   \n",
       "522  chrX:73852663-73852713(-)   \n",
       "523  chrX:73852687-73852737(-)   \n",
       "524  chrX:73852689-73852739(-)   \n",
       "525  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...   \n",
       "1    TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...   \n",
       "2    TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...   \n",
       "3    ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...   \n",
       "4    TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...   \n",
       "..                                                 ...   \n",
       "521  CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...   \n",
       "522  TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...   \n",
       "523  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "524  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "525  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...             0.48   \n",
       "1    CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...             0.54   \n",
       "2    CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...             0.52   \n",
       "3    ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...             0.50   \n",
       "4    CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...             0.50   \n",
       "..                                                 ...              ...   \n",
       "521  CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...             0.58   \n",
       "522  CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...             0.48   \n",
       "523  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "524  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "525  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "     GC_content_LHS  GC_content_RHS  longest_homopolymer  has_required_nts  \n",
       "0              0.48            0.48                    4              True  \n",
       "1              0.60            0.48                    4              True  \n",
       "2              0.56            0.48                    4              True  \n",
       "3              0.52            0.48                    4              True  \n",
       "4              0.52            0.48                    4              True  \n",
       "..              ...             ...                  ...               ...  \n",
       "521            0.52            0.64                    4              True  \n",
       "522            0.48            0.48                    4              True  \n",
       "523            0.48            0.60                    4              True  \n",
       "524            0.48            0.64                    4              True  \n",
       "525            0.48            0.52                    4              True  \n",
       "\n",
       "[526 rows x 14 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_kmers_filtered "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c0d8c8f9-0c4c-46a4-bb8f-16782e47d379",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sort in increasing homopolymer length\n",
    "roi_kmers_filtered = roi_kmers_filtered.sort_values('longest_homopolymer', ascending=True).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d951231a-8c94-4473-978d-b202c2ded194",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73827199</td>\n",
       "      <td>73827249</td>\n",
       "      <td>XIST_0_6549</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73827199-73827249(-)</td>\n",
       "      <td>GTGCAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTT...</td>\n",
       "      <td>GCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGTTCT...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.56</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73851704</td>\n",
       "      <td>73851754</td>\n",
       "      <td>XIST_5_10323</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73851704-73851754(-)</td>\n",
       "      <td>AGATGGATGATAGCAGGTCAGGCAGAGGAAGTCATGTGCATTGCAT...</td>\n",
       "      <td>GCTCATGCAATGCACATGACTTCCTCTGCCTGACCTGCTATCATCC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850595</td>\n",
       "      <td>73850645</td>\n",
       "      <td>XIST_5_9214</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850595-73850645(-)</td>\n",
       "      <td>TGCTCCAGGCCTGCTTGGTGTGGACATGGTGGTGAGCCGTGGCAAG...</td>\n",
       "      <td>GGTCCTTGCCACGGCTCACCACCATGTCCACACCAAGCAGGCCTGG...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850597</td>\n",
       "      <td>73850647</td>\n",
       "      <td>XIST_5_9216</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850597-73850647(-)</td>\n",
       "      <td>AGTGCTCCAGGCCTGCTTGGTGTGGACATGGTGGTGAGCCGTGGCA...</td>\n",
       "      <td>TCCTTGCCACGGCTCACCACCATGTCCACACCAAGCAGGCCTGGAG...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.64</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73826123</td>\n",
       "      <td>73826173</td>\n",
       "      <td>XIST_0_5473</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73826123-73826173(-)</td>\n",
       "      <td>CAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCTTCCT...</td>\n",
       "      <td>AAGAAGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852326</td>\n",
       "      <td>73852376</td>\n",
       "      <td>XIST_5_10945</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852326-73852376(-)</td>\n",
       "      <td>GGGGCTGCGGATACCTGGTTTTATTATTTTTTCTTTGCCCAACGGG...</td>\n",
       "      <td>CGGCCCCGTTGGGCAAAGAAAAAATAATAAAACCAGGTATCCGCAG...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850763</td>\n",
       "      <td>73850813</td>\n",
       "      <td>XIST_5_9382</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850763-73850813(-)</td>\n",
       "      <td>TTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCcccc...</td>\n",
       "      <td>gtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGA...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850758</td>\n",
       "      <td>73850808</td>\n",
       "      <td>XIST_5_9377</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850758-73850808(-)</td>\n",
       "      <td>CTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccacc...</td>\n",
       "      <td>ggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAAT...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.52</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>524</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850757</td>\n",
       "      <td>73850807</td>\n",
       "      <td>XIST_5_9376</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850757-73850807(-)</td>\n",
       "      <td>TGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccaccc...</td>\n",
       "      <td>tggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAA...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.48</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>525</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850764</td>\n",
       "      <td>73850814</td>\n",
       "      <td>XIST_5_9383</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850764-73850814(-)</td>\n",
       "      <td>ATTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccc...</td>\n",
       "      <td>tggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGAG...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>526 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73827199  73827249   XIST_0_6549     .      -   \n",
       "1      chrX  73851704  73851754  XIST_5_10323     .      -   \n",
       "2      chrX  73850595  73850645   XIST_5_9214     .      -   \n",
       "3      chrX  73850597  73850647   XIST_5_9216     .      -   \n",
       "4      chrX  73826123  73826173   XIST_0_5473     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "521    chrX  73852326  73852376  XIST_5_10945     .      -   \n",
       "522    chrX  73850763  73850813   XIST_5_9382     .      -   \n",
       "523    chrX  73850758  73850808   XIST_5_9377     .      -   \n",
       "524    chrX  73850757  73850807   XIST_5_9376     .      -   \n",
       "525    chrX  73850764  73850814   XIST_5_9383     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73827199-73827249(-)   \n",
       "1    chrX:73851704-73851754(-)   \n",
       "2    chrX:73850595-73850645(-)   \n",
       "3    chrX:73850597-73850647(-)   \n",
       "4    chrX:73826123-73826173(-)   \n",
       "..                         ...   \n",
       "521  chrX:73852326-73852376(-)   \n",
       "522  chrX:73850763-73850813(-)   \n",
       "523  chrX:73850758-73850808(-)   \n",
       "524  chrX:73850757-73850807(-)   \n",
       "525  chrX:73850764-73850814(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    GTGCAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTT...   \n",
       "1    AGATGGATGATAGCAGGTCAGGCAGAGGAAGTCATGTGCATTGCAT...   \n",
       "2    TGCTCCAGGCCTGCTTGGTGTGGACATGGTGGTGAGCCGTGGCAAG...   \n",
       "3    AGTGCTCCAGGCCTGCTTGGTGTGGACATGGTGGTGAGCCGTGGCA...   \n",
       "4    CAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCTTCCT...   \n",
       "..                                                 ...   \n",
       "521  GGGGCTGCGGATACCTGGTTTTATTATTTTTTCTTTGCCCAACGGG...   \n",
       "522  TTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCcccc...   \n",
       "523  CTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccacc...   \n",
       "524  TGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccaccc...   \n",
       "525  ATTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccc...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    GCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGTTCT...             0.54   \n",
       "1    GCTCATGCAATGCACATGACTTCCTCTGCCTGACCTGCTATCATCC...             0.50   \n",
       "2    GGTCCTTGCCACGGCTCACCACCATGTCCACACCAAGCAGGCCTGG...             0.64   \n",
       "3    TCCTTGCCACGGCTCACCACCATGTCCACACCAAGCAGGCCTGGAG...             0.62   \n",
       "4    AAGAAGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACA...             0.52   \n",
       "..                                                 ...              ...   \n",
       "521  CGGCCCCGTTGGGCAAAGAAAAAATAATAAAACCAGGTATCCGCAG...             0.52   \n",
       "522  gtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGA...             0.54   \n",
       "523  ggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAAT...             0.60   \n",
       "524  tggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAA...             0.58   \n",
       "525  tggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGAG...             0.52   \n",
       "\n",
       "     GC_content_LHS  GC_content_RHS  longest_homopolymer  has_required_nts  \n",
       "0              0.52            0.56                    2              True  \n",
       "1              0.48            0.52                    2              True  \n",
       "2              0.64            0.64                    2              True  \n",
       "3              0.60            0.64                    2              True  \n",
       "4              0.52            0.52                    2              True  \n",
       "..              ...             ...                  ...               ...  \n",
       "521            0.52            0.52                    6              True  \n",
       "522            0.60            0.48                    6              True  \n",
       "523            0.68            0.52                    6              True  \n",
       "524            0.68            0.48                    6              True  \n",
       "525            0.56            0.48                    6              True  \n",
       "\n",
       "[526 rows x 14 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_kmers_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3504dcc6-d55f-4491-bcdd-5049eab3b40e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nfs/team205/is10/projects/thytrans/scripts/008_visiumHD/gene2probe.py:117: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  start = int(probe_df['start'].iloc[i])\n",
      "/nfs/team205/is10/projects/thytrans/scripts/008_visiumHD/gene2probe.py:118: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  end = int(probe_df['end'].iloc[i])\n"
     ]
    }
   ],
   "source": [
    "selected_probes_list = []\n",
    "df = roi_kmers_filtered.copy()\n",
    "i = 0\n",
    "while i < 3:\n",
    "    selected_probes_list.append(df.iloc[0:1,:])\n",
    "    ## Going for a high offset as we have a large span:\n",
    "    df = remove_overlapping_probes(df,\n",
    "                                   df['name'][0], \n",
    "                                   offset=1000).reset_index(drop=True).copy()\n",
    "    i +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "069f0bab-21de-4f48-a902-767f77ea4260",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_probes_df = pd.concat(selected_probes_list, axis=0).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6ed7dbf9-aef5-4723-ad14-f6892251e7e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73827199</td>\n",
       "      <td>73827249</td>\n",
       "      <td>XIST_0_6549</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73827199-73827249(-)</td>\n",
       "      <td>GTGCAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTT...</td>\n",
       "      <td>GCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGTTCT...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.56</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73851704</td>\n",
       "      <td>73851754</td>\n",
       "      <td>XIST_5_10323</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73851704-73851754(-)</td>\n",
       "      <td>AGATGGATGATAGCAGGTCAGGCAGAGGAAGTCATGTGCATTGCAT...</td>\n",
       "      <td>GCTCATGCAATGCACATGACTTCCTCTGCCTGACCTGCTATCATCC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850595</td>\n",
       "      <td>73850645</td>\n",
       "      <td>XIST_5_9214</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850595-73850645(-)</td>\n",
       "      <td>TGCTCCAGGCCTGCTTGGTGTGGACATGGTGGTGAGCCGTGGCAAG...</td>\n",
       "      <td>GGTCCTTGCCACGGCTCACCACCATGTCCACACCAAGCAGGCCTGG...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     start       end          name score strand  \\\n",
       "0    chrX  73827199  73827249   XIST_0_6549     .      -   \n",
       "1    chrX  73851704  73851754  XIST_5_10323     .      -   \n",
       "2    chrX  73850595  73850645   XIST_5_9214     .      -   \n",
       "\n",
       "                  kmer_coord  \\\n",
       "0  chrX:73827199-73827249(-)   \n",
       "1  chrX:73851704-73851754(-)   \n",
       "2  chrX:73850595-73850645(-)   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  GTGCAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTT...   \n",
       "1  AGATGGATGATAGCAGGTCAGGCAGAGGAAGTCATGTGCATTGCAT...   \n",
       "2  TGCTCCAGGCCTGCTTGGTGTGGACATGGTGGTGAGCCGTGGCAAG...   \n",
       "\n",
       "                                           probe_seq  GC_content_full  \\\n",
       "0  GCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGTTCT...             0.54   \n",
       "1  GCTCATGCAATGCACATGACTTCCTCTGCCTGACCTGCTATCATCC...             0.50   \n",
       "2  GGTCCTTGCCACGGCTCACCACCATGTCCACACCAAGCAGGCCTGG...             0.64   \n",
       "\n",
       "   GC_content_LHS  GC_content_RHS  longest_homopolymer  has_required_nts  \n",
       "0            0.52            0.56                    2              True  \n",
       "1            0.48            0.52                    2              True  \n",
       "2            0.64            0.64                    2              True  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_probes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9ab3f7e6-a39f-417c-8b40-92eb9e3e2908",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTGCAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATGC\n",
      "AGATGGATGATAGCAGGTCAGGCAGAGGAAGTCATGTGCATTGCATGAGC\n",
      "TGCTCCAGGCCTGCTTGGTGTGGACATGGTGGTGAGCCGTGGCAAGGACC\n"
     ]
    }
   ],
   "source": [
    "for seq in selected_probes_df['transcript_seq']:\n",
    "    print(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3ec41fa9-1992-49cc-8f8f-88da8fdd07e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Make split probes\n",
    "selected_probes_df['probe_seq_LHS'] = [seq[0:split_nt] for seq in selected_probes_df['probe_seq']]\n",
    "selected_probes_df['probe_seq_RHS'] = [seq[split_nt: probe_length] for seq in selected_probes_df['probe_seq']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "76b0b300-b3ca-49db-a600-c82961d30ce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Also add gene_ID for completeness\n",
    "selected_probes_df['gene_ID'] = gene_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "1ce62661-8a1b-4ef6-83e1-2bd603806975",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export selected probes as dataframe:\n",
    "selected_probes_df.to_csv((out_dir + 'kmers_selected_probes.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cec17e82-1d55-4479-b506-84b8796e4355",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gene2probe",
   "language": "python",
   "name": "gene2probe"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
