{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e216d8aa-ec15-4d54-a8ff-221efdb48c7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import sys\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "from gene2probe import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f708aff-11ce-4ac7-8e30-952a5902419e",
   "metadata": {},
   "source": [
    "This tutorial guides you through the design of custom probes against a gene of interest.\n",
    "\n",
    "First, we need to specify the gene symbol, and which feature we are interested in designing probes agains (in this case we are using exons).\n",
    "We also need to specify an output directory for the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "26aa4295-dc32-413a-92c9-71c01dad41e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Specify gene of interest and feature of interest\n",
    "gene_ID = 'XIST'\n",
    "mode = 'exon' ## Whether to consider only exons / introns or full gene\n",
    "\n",
    "## specify output directory\n",
    "out_dir = '../sample_run/probeDesign_' + gene_ID + '_' + mode + '/'\n",
    "## Create output directory\n",
    "os.makedirs(out_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17f89fba-8fd7-4a21-9f27-67060bd21a44",
   "metadata": {},
   "source": [
    "Additionally, we need to provide the path to several resource files. Many of these files can be obtained from [UCSC table browser](https://genome.ucsc.edu/cgi-bin/hgTables).\n",
    "\n",
    "We also need a blast database, such as the one we generated in the [previous tutorial](https://github.com/Teichlab/gene2probe/blob/main/notebooks/001_make_blast_database.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1dffb1b9-9ecc-490d-bea4-7aefb5a26028",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Required resources (most can be downloaded from \n",
    "gtf = '../hg38_resources/hg38.ncbiRefSeq.gtf' ## Gene annotation in gtf file\n",
    "## We recommend using RefSeq as this is manually curated and more likely to contain an isoform that is present across most cell types\n",
    "## Alternatively, one can filter based on RNA-seq data for a cell type/tissue of interest\n",
    "fasta = '../hg38_resources/hg38.fa' ## Genome in fasta file\n",
    "snp_db = '../hg38_resources/hg38_snp151Common.txt' ## Database of known SNPs and small indels\n",
    "repeats = '../hg38_resources/hg38_rmsk.bed' ## bed file with repeats/low complexity regions to be excluded\n",
    "gaps = '../hg38_resources/hg38_rmsk.bed' ## bed file with gaps in the genome assembly to be excluded\n",
    "blast_db = '../hg38_resources/blastdb/hg38_ncbiRefSeq_exons_db' ## Database of all human transcripts to blast against"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a65bb4d6-13fd-4d49-9a1b-496c7d0e6a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to blast binaries.\n",
    "## Replace with your conda environment\n",
    "## This can also be omitted if you started the jupyter session from within the gene2probe conda environment\n",
    "blast_exec_path = '/nfs/team205/is10/miniconda/envs/gene2probe_env/bin/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa752587-9f96-4ca4-8d44-fae0021f0cf2",
   "metadata": {},
   "source": [
    "Finally, we need to provide a set of parameters related to our probe's length, at which nucleotide it's split (if at all), the acceptable range for GC content and any specific requirements for individual nucleotides.\n",
    "\n",
    "Here we are following the [recommendations of 10x Genomics for custom probes for VisiumHD/VisiumFFPE/Flex](https://cdn.10xgenomics.com/image/upload/v1697739385/support-documents/CG000621_CustomProbeDesign_TechNote_RevC.pdf)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cba53a9d-2f6d-4611-8b97-4a6ce8ac75e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Additional parameters regarding how the probe should look like\n",
    "probe_length = 50 ## Length of probe in nucleotides\n",
    "split_nt = 25 ## Index of nucleotide to split the probe at (start of RHS) - set to None if splitting probe is not needed\n",
    "min_GC = 0.44 ## Minimum GC content for probe (if split probe, applied to both LHS and RHS)\n",
    "max_GC = 0.72 ## Maximum GC content for probe (if split probe, applied to both LHS and RHS)\n",
    "required_nts = {24: 'T'} ##Â Dictionary of index (0-based) for required nts - by default, 25th nucleotide must be a T - set to None if no requirements\n",
    "probe_offset = 100 ## Minimum distance between probes - 10 bp is the recommended minimum by 10x, this can also be adjusted depending on how many probes pass other cutoffs\n",
    "n_desired_probes =3 ## Number of probes to be designed."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0392b3e9-f101-4386-a981-1338c431ec02",
   "metadata": {},
   "source": [
    "Now we can start by reading the gene annotation and filtering for our gene of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4bb522db-fc21-421c-b37e-8154a58ae63f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read gtf file\n",
    "gene_anno = read_gtf(gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "049129b1-a807-491e-a725-0a8d9f7c1113",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"TRNP\"; transcript_id \"rna-TRNP\";  gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"TRNP\"; transcript_id \"rna-TRNP\"; exon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"TRNT\"; transcript_id \"rna-TRNT\";  gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"TRNT\"; transcript_id \"rna-TRNT\"; exon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"CYTB\"; transcript_id \"rna-CYTB\";  gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886697</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>29321</td>\n",
       "      <td>29370</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"WASH7P\"; transcript_id \"NR_024540.1\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886698</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886699</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886700</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886701</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4886702 rows Ã 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname                 source     feature  start    end score strand  \\\n",
       "0          chrM  ncbiRefSeq.2022-10-28  transcript  15956  16023     .      -   \n",
       "1          chrM  ncbiRefSeq.2022-10-28        exon  15956  16023     .      -   \n",
       "2          chrM  ncbiRefSeq.2022-10-28  transcript  15888  15953     .      +   \n",
       "3          chrM  ncbiRefSeq.2022-10-28        exon  15888  15953     .      +   \n",
       "4          chrM  ncbiRefSeq.2022-10-28  transcript  14747  15887     .      +   \n",
       "...         ...                    ...         ...    ...    ...   ...    ...   \n",
       "4886697    chr1  ncbiRefSeq.2022-10-28        exon  29321  29370     .      -   \n",
       "4886698    chr1  ncbiRefSeq.2022-10-28  transcript  11874  14409     .      +   \n",
       "4886699    chr1  ncbiRefSeq.2022-10-28        exon  11874  12227     .      +   \n",
       "4886700    chr1  ncbiRefSeq.2022-10-28        exon  12613  12721     .      +   \n",
       "4886701    chr1  ncbiRefSeq.2022-10-28        exon  13221  14409     .      +   \n",
       "\n",
       "        frame                                          attribute  \n",
       "0           .  gene_id \"TRNP\"; transcript_id \"rna-TRNP\";  gen...  \n",
       "1           .  gene_id \"TRNP\"; transcript_id \"rna-TRNP\"; exon...  \n",
       "2           .  gene_id \"TRNT\"; transcript_id \"rna-TRNT\";  gen...  \n",
       "3           .  gene_id \"TRNT\"; transcript_id \"rna-TRNT\"; exon...  \n",
       "4           .  gene_id \"CYTB\"; transcript_id \"rna-CYTB\";  gen...  \n",
       "...       ...                                                ...  \n",
       "4886697     .  gene_id \"WASH7P\"; transcript_id \"NR_024540.1\";...  \n",
       "4886698     .  gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...  \n",
       "4886699     .  gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...  \n",
       "4886700     .  gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...  \n",
       "4886701     .  gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...  \n",
       "\n",
       "[4886702 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b510a00e-bf9f-4476-85f8-8d643859dc0d",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gene_ID' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m## Extract regions corresponding to gene of interest (symbol: gene_name, Ensembl ID: gene_ID), subset to feature of interest and convert to bed style dataframe:\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m roi_bed \u001b[38;5;241m=\u001b[39m \u001b[43mget_region_of_interest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgene_anno\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene_ID\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgene_id_type\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgene_name\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfeature\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/nfs/team205/is10/miniconda/envs/gene2probe_env/lib/python3.11/site-packages/gene2probe/gene2probe.py:62\u001b[0m, in \u001b[0;36mget_region_of_interest\u001b[0;34m(gtf, gene_id, gene_id_type, feature, distance_from_exons)\u001b[0m\n\u001b[1;32m     59\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFeature has to be one of [\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mexon\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCDS\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtranscript\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mintron\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m, \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msplice_junction\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     61\u001b[0m \u001b[38;5;66;03m## Filter for gene_id\u001b[39;00m\n\u001b[0;32m---> 62\u001b[0m roi \u001b[38;5;241m=\u001b[39m gtf\u001b[38;5;241m.\u001b[39miloc[np\u001b[38;5;241m.\u001b[39mwhere(gene_ids \u001b[38;5;241m==\u001b[39m \u001b[43mgene_ID\u001b[49m)[\u001b[38;5;241m0\u001b[39m],:]\n\u001b[1;32m     64\u001b[0m \u001b[38;5;66;03m## If feature is exon, CDS or transcript, our job is easy:\u001b[39;00m\n\u001b[1;32m     65\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m feature \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mexon\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCDS\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtranscript\u001b[39m\u001b[38;5;124m'\u001b[39m]:\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gene_ID' is not defined"
     ]
    }
   ],
   "source": [
    "## Extract regions corresponding to gene of interest (symbol: gene_name, Ensembl ID: gene_ID), subset to feature of interest and convert to bed style dataframe:\n",
    "roi_bed = get_region_of_interest(gene_anno, gene_ID, gene_id_type = 'gene_name', feature=mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "60894d08-da20-462e-beb6-b2e672db60f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73827984</td>\n",
       "      <td>XIST_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73829067</td>\n",
       "      <td>73829231</td>\n",
       "      <td>XIST_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73831065</td>\n",
       "      <td>73831274</td>\n",
       "      <td>XIST_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73833237</td>\n",
       "      <td>73833374</td>\n",
       "      <td>XIST_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73837439</td>\n",
       "      <td>73837503</td>\n",
       "      <td>XIST_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73841381</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     start       end    name score strand\n",
       "0    chrX  73820650  73827984  XIST_0     .      -\n",
       "1    chrX  73829067  73829231  XIST_1     .      -\n",
       "2    chrX  73831065  73831274  XIST_2     .      -\n",
       "3    chrX  73833237  73833374  XIST_3     .      -\n",
       "4    chrX  73837439  73837503  XIST_4     .      -\n",
       "5    chrX  73841381  73852753  XIST_5     .      -"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "019cdc9b-d3fa-4d46-91f7-5308681f6e6c",
   "metadata": {},
   "source": [
    "After extracting the coordinates of interest and converting to a bed-like format, we can generate all possible kmers that fall within these regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "60f606b4-d196-4382-bcae-f64364dc5768",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers = generate_kmers(roi_bed, k=probe_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c981be58-decf-4f36-bce9-1efabaa02627",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18981</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18982</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18983</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852701</td>\n",
       "      <td>73852751</td>\n",
       "      <td>XIST_5_11320</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18984</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852702</td>\n",
       "      <td>73852752</td>\n",
       "      <td>XIST_5_11321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18985</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852703</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5_11322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18986 rows Ã 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -\n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -\n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -\n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -\n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -\n",
       "...       ...       ...       ...           ...   ...    ...\n",
       "18981    chrX  73852699  73852749  XIST_5_11318     .      -\n",
       "18982    chrX  73852700  73852750  XIST_5_11319     .      -\n",
       "18983    chrX  73852701  73852751  XIST_5_11320     .      -\n",
       "18984    chrX  73852702  73852752  XIST_5_11321     .      -\n",
       "18985    chrX  73852703  73852753  XIST_5_11322     .      -\n",
       "\n",
       "[18986 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "69c58199-798b-45c7-8799-bcd30e8dfb3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export unfiltered\n",
    "kmers.to_csv((out_dir + 'kmers_all.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4aa563ab-1e10-4e63-b8a3-9fb266a70ddf",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Convert to pybedtools\n",
    "kmers_bed = pybedtools.BedTool.from_dataframe(kmers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8715663-c967-47e5-be22-1094b2721689",
   "metadata": {},
   "source": [
    "We can next exclude kmers overlapping undesired regions (repeats, low complexity regions, common polymorphism, gaps in the assembly) from further consideration.\n",
    "\n",
    "Ideally, we will exclude everything that overlaps a repeat or polymorphism, but if we only have too few kmers available, we might need to relax these requirements (e.g., to only exclude kmers overlapping SNPs around the ligation junction, if the probes are split)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "48005706-0987-41ff-af44-77b08811b48d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove repeat/low complexity regions\n",
    "if repeats is not None:\n",
    "    ## Read rmsk\n",
    "    repeats_bed = pybedtools.BedTool(repeats)\n",
    "    ## Remove any kmers overlapping any masked regions\n",
    "    kmers_bed = kmers_bed.intersect(repeats_bed, v=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "26aa01c2-8e9c-480b-8a4b-b94260d69da8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Same for gaps\n",
    "if gaps is not None:\n",
    "    ## Read rmsk\n",
    "    gaps_bed = pybedtools.BedTool(gaps)\n",
    "    ## Remove any kmers overlapping any masked regions\n",
    "    kmers_bed = kmers_bed.intersect(gaps_bed, v=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0b801075-820c-4df3-becb-0082c24702b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Same with SNPs\n",
    "if snp_db is not None:\n",
    "    snps_bed = pybedtools.BedTool(snp_db)\n",
    "    ## Remove any kmers overlapping any masked regions\n",
    "    kmers_bed = kmers_bed.intersect(snps_bed, v=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d732a6ee-c089-4e38-9e42-075d95b63c07",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers = kmers_bed.to_dataframe(names=['seqname', 'start', 'end', 'name', 'score', 'strand'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "969e23e8-4a03-44a8-a25a-110c66ed06f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15711</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852696</td>\n",
       "      <td>73852746</td>\n",
       "      <td>XIST_5_11315</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15712</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852697</td>\n",
       "      <td>73852747</td>\n",
       "      <td>XIST_5_11316</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15713</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852698</td>\n",
       "      <td>73852748</td>\n",
       "      <td>XIST_5_11317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15714</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15715</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15716 rows Ã 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -\n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -\n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -\n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -\n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -\n",
       "...       ...       ...       ...           ...   ...    ...\n",
       "15711    chrX  73852696  73852746  XIST_5_11315     .      -\n",
       "15712    chrX  73852697  73852747  XIST_5_11316     .      -\n",
       "15713    chrX  73852698  73852748  XIST_5_11317     .      -\n",
       "15714    chrX  73852699  73852749  XIST_5_11318     .      -\n",
       "15715    chrX  73852700  73852750  XIST_5_11319     .      -\n",
       "\n",
       "[15716 rows x 6 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "090e41ad-d72b-4b91-bbb8-f6c259530c96",
   "metadata": {},
   "source": [
    "Having excluded undesirable kmers based on intersection with genomic annotations, the next step is to consider their sequence features.\n",
    "\n",
    "For this, we first extract the sequences of each k-mer, and then estimate features such as GC content and the presence of desired nucleotides in specific positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6b91e694-c124-483a-941e-5b134aa3e69e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get DNA for the transcript\n",
    "kmers_seq = kmers_bed.sequence(fi=fasta, s=True) \n",
    "\n",
    "## We can read in the sequences and simultaneously monitor GC content and count the longest homopolymer stretch\n",
    "kmers_seq_stats = get_sequence_stats(kmers_seq.seqfn, probe_length, split_nt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "6db0119a-9c26-47f8-bb25-363d738fede9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Combining with our dataframe\n",
    "kmers = pd.merge(kmers, kmers_seq_stats, left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9a98f6eb-dcdd-4406-9c7c-8e3dd37e18e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820650-73820700(-)</td>\n",
       "      <td>GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...</td>\n",
       "      <td>CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820651-73820701(-)</td>\n",
       "      <td>TGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAA...</td>\n",
       "      <td>AAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTC...</td>\n",
       "      <td>0.28</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820652-73820702(-)</td>\n",
       "      <td>CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...</td>\n",
       "      <td>AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820653-73820703(-)</td>\n",
       "      <td>ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...</td>\n",
       "      <td>AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820654-73820704(-)</td>\n",
       "      <td>AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...</td>\n",
       "      <td>GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15711</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852696</td>\n",
       "      <td>73852746</td>\n",
       "      <td>XIST_5_11315</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852696-73852746(-)</td>\n",
       "      <td>TTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCT...</td>\n",
       "      <td>CGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15712</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852697</td>\n",
       "      <td>73852747</td>\n",
       "      <td>XIST_5_11316</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852697-73852747(-)</td>\n",
       "      <td>GTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTC...</td>\n",
       "      <td>GAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15713</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852698</td>\n",
       "      <td>73852748</td>\n",
       "      <td>XIST_5_11317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852698-73852748(-)</td>\n",
       "      <td>AGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACT...</td>\n",
       "      <td>AGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAG...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15714</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852699-73852749(-)</td>\n",
       "      <td>CAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTAC...</td>\n",
       "      <td>GAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15715</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852700-73852750(-)</td>\n",
       "      <td>TCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTA...</td>\n",
       "      <td>AGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAA...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15716 rows Ã 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand  \\\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -   \n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -   \n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -   \n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -   \n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -   \n",
       "...       ...       ...       ...           ...   ...    ...   \n",
       "15711    chrX  73852696  73852746  XIST_5_11315     .      -   \n",
       "15712    chrX  73852697  73852747  XIST_5_11316     .      -   \n",
       "15713    chrX  73852698  73852748  XIST_5_11317     .      -   \n",
       "15714    chrX  73852699  73852749  XIST_5_11318     .      -   \n",
       "15715    chrX  73852700  73852750  XIST_5_11319     .      -   \n",
       "\n",
       "                      kmer_coord  \\\n",
       "0      chrX:73820650-73820700(-)   \n",
       "1      chrX:73820651-73820701(-)   \n",
       "2      chrX:73820652-73820702(-)   \n",
       "3      chrX:73820653-73820703(-)   \n",
       "4      chrX:73820654-73820704(-)   \n",
       "...                          ...   \n",
       "15711  chrX:73852696-73852746(-)   \n",
       "15712  chrX:73852697-73852747(-)   \n",
       "15713  chrX:73852698-73852748(-)   \n",
       "15714  chrX:73852699-73852749(-)   \n",
       "15715  chrX:73852700-73852750(-)   \n",
       "\n",
       "                                          transcript_seq  \\\n",
       "0      GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...   \n",
       "1      TGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAA...   \n",
       "2      CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...   \n",
       "3      ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...   \n",
       "4      AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...   \n",
       "...                                                  ...   \n",
       "15711  TTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCT...   \n",
       "15712  GTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTC...   \n",
       "15713  AGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACT...   \n",
       "15714  CAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTAC...   \n",
       "15715  TCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTA...   \n",
       "\n",
       "                                               probe_seq  GC_content_full  \\\n",
       "0      CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...             0.30   \n",
       "1      AAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTC...             0.28   \n",
       "2      AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...             0.30   \n",
       "3      AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...             0.30   \n",
       "4      GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...             0.30   \n",
       "...                                                  ...              ...   \n",
       "15711  CGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTA...             0.46   \n",
       "15712  GAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAA...             0.46   \n",
       "15713  AGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAG...             0.44   \n",
       "15714  GAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGA...             0.46   \n",
       "15715  AGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAA...             0.44   \n",
       "\n",
       "       longest_homopolymer  GC_content_LHS  GC_content_RHS  \n",
       "0                        4            0.20            0.40  \n",
       "1                        4            0.16            0.40  \n",
       "2                        4            0.16            0.44  \n",
       "3                        4            0.16            0.44  \n",
       "4                        4            0.16            0.44  \n",
       "...                    ...             ...             ...  \n",
       "15711                    3            0.44            0.48  \n",
       "15712                    3            0.44            0.48  \n",
       "15713                    3            0.44            0.44  \n",
       "15714                    3            0.44            0.48  \n",
       "15715                    3            0.44            0.44  \n",
       "\n",
       "[15716 rows x 13 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "44dfc696-c7b4-4b79-a382-8d3d9511db2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "has_required_nts\n",
      "False    11525\n",
      "True      4191\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## Check for required nucleotides in specific positions:\n",
    "if required_nts is not None:\n",
    "    kmers['has_required_nts'] = check_for_required_nts(kmers, required_nts)\n",
    "    print(kmers['has_required_nts'].value_counts())\n",
    "    ## Filter for required nucleotides\n",
    "    kmers = kmers[kmers['has_required_nts']==True].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a937f445-2786-400d-ae11-38233957bbe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export kmers before filtering\n",
    "kmers.to_csv((out_dir + 'kmers_candidates_unfiltered.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0fa42cd0-1b76-4948-b33f-ef5d057d9f01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820650-73820700(-)</td>\n",
       "      <td>GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...</td>\n",
       "      <td>CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.40</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820652-73820702(-)</td>\n",
       "      <td>CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...</td>\n",
       "      <td>AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820653-73820703(-)</td>\n",
       "      <td>ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...</td>\n",
       "      <td>AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820654-73820704(-)</td>\n",
       "      <td>AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...</td>\n",
       "      <td>GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820655</td>\n",
       "      <td>73820705</td>\n",
       "      <td>XIST_0_5</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820655-73820705(-)</td>\n",
       "      <td>AAACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTG...</td>\n",
       "      <td>TTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCCA...</td>\n",
       "      <td>0.28</td>\n",
       "      <td>4</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4186</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852664</td>\n",
       "      <td>73852714</td>\n",
       "      <td>XIST_5_11283</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852664-73852714(-)</td>\n",
       "      <td>ATATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCT...</td>\n",
       "      <td>AGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGAA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>4</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4187</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852680</td>\n",
       "      <td>73852730</td>\n",
       "      <td>XIST_5_11299</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852680-73852730(-)</td>\n",
       "      <td>ATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCTGGAAGCT...</td>\n",
       "      <td>AGGAAGCTTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGC...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4188</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4189</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4191 rows Ã 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqname     start       end          name score strand  \\\n",
       "0       chrX  73820650  73820700      XIST_0_0     .      -   \n",
       "1       chrX  73820652  73820702      XIST_0_2     .      -   \n",
       "2       chrX  73820653  73820703      XIST_0_3     .      -   \n",
       "3       chrX  73820654  73820704      XIST_0_4     .      -   \n",
       "4       chrX  73820655  73820705      XIST_0_5     .      -   \n",
       "...      ...       ...       ...           ...   ...    ...   \n",
       "4186    chrX  73852664  73852714  XIST_5_11283     .      -   \n",
       "4187    chrX  73852680  73852730  XIST_5_11299     .      -   \n",
       "4188    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "4189    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "4190    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                     kmer_coord  \\\n",
       "0     chrX:73820650-73820700(-)   \n",
       "1     chrX:73820652-73820702(-)   \n",
       "2     chrX:73820653-73820703(-)   \n",
       "3     chrX:73820654-73820704(-)   \n",
       "4     chrX:73820655-73820705(-)   \n",
       "...                         ...   \n",
       "4186  chrX:73852664-73852714(-)   \n",
       "4187  chrX:73852680-73852730(-)   \n",
       "4188  chrX:73852687-73852737(-)   \n",
       "4189  chrX:73852689-73852739(-)   \n",
       "4190  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                         transcript_seq  \\\n",
       "0     GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...   \n",
       "1     CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...   \n",
       "2     ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...   \n",
       "3     AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...   \n",
       "4     AAACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTG...   \n",
       "...                                                 ...   \n",
       "4186  ATATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCT...   \n",
       "4187  ATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCTGGAAGCT...   \n",
       "4188  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "4189  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "4190  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                              probe_seq  GC_content_full  \\\n",
       "0     CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...             0.30   \n",
       "1     AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...             0.30   \n",
       "2     AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...             0.30   \n",
       "3     GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...             0.30   \n",
       "4     TTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCCA...             0.28   \n",
       "...                                                 ...              ...   \n",
       "4186  AGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGAA...             0.46   \n",
       "4187  AGGAAGCTTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGC...             0.52   \n",
       "4188  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "4189  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "4190  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "      longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                       4            0.20            0.40              True  \n",
       "1                       4            0.16            0.44              True  \n",
       "2                       4            0.16            0.44              True  \n",
       "3                       4            0.16            0.44              True  \n",
       "4                       4            0.12            0.44              True  \n",
       "...                   ...             ...             ...               ...  \n",
       "4186                    4            0.44            0.48              True  \n",
       "4187                    4            0.60            0.44              True  \n",
       "4188                    4            0.48            0.60              True  \n",
       "4189                    4            0.48            0.64              True  \n",
       "4190                    4            0.48            0.52              True  \n",
       "\n",
       "[4191 rows x 14 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6b903375-7a4b-4878-9479-574bb359c563",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter for GC content\n",
    "kmers = filter_by_GC_content(kmers, min_GC, max_GC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1ed1ae93-3577-4a85-a65a-9e506d4c317a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821073</td>\n",
       "      <td>73821123</td>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821073-73821123(-)</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...</td>\n",
       "      <td>AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821085</td>\n",
       "      <td>73821135</td>\n",
       "      <td>XIST_0_435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821085-73821135(-)</td>\n",
       "      <td>TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...</td>\n",
       "      <td>CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821087</td>\n",
       "      <td>73821137</td>\n",
       "      <td>XIST_0_437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821087-73821137(-)</td>\n",
       "      <td>TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...</td>\n",
       "      <td>CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821088</td>\n",
       "      <td>73821138</td>\n",
       "      <td>XIST_0_438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821088-73821138(-)</td>\n",
       "      <td>ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...</td>\n",
       "      <td>ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821091</td>\n",
       "      <td>73821141</td>\n",
       "      <td>XIST_0_441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821091-73821141(-)</td>\n",
       "      <td>TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...</td>\n",
       "      <td>CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852654</td>\n",
       "      <td>73852704</td>\n",
       "      <td>XIST_5_11273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852654-73852704(-)</td>\n",
       "      <td>CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...</td>\n",
       "      <td>CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852663</td>\n",
       "      <td>73852713</td>\n",
       "      <td>XIST_5_11282</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852663-73852713(-)</td>\n",
       "      <td>TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...</td>\n",
       "      <td>CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>534 rows Ã 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73821073  73821123    XIST_0_423     .      -   \n",
       "1      chrX  73821085  73821135    XIST_0_435     .      -   \n",
       "2      chrX  73821087  73821137    XIST_0_437     .      -   \n",
       "3      chrX  73821088  73821138    XIST_0_438     .      -   \n",
       "4      chrX  73821091  73821141    XIST_0_441     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "529    chrX  73852654  73852704  XIST_5_11273     .      -   \n",
       "530    chrX  73852663  73852713  XIST_5_11282     .      -   \n",
       "531    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "532    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "533    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73821073-73821123(-)   \n",
       "1    chrX:73821085-73821135(-)   \n",
       "2    chrX:73821087-73821137(-)   \n",
       "3    chrX:73821088-73821138(-)   \n",
       "4    chrX:73821091-73821141(-)   \n",
       "..                         ...   \n",
       "529  chrX:73852654-73852704(-)   \n",
       "530  chrX:73852663-73852713(-)   \n",
       "531  chrX:73852687-73852737(-)   \n",
       "532  chrX:73852689-73852739(-)   \n",
       "533  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...   \n",
       "1    TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...   \n",
       "2    TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...   \n",
       "3    ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...   \n",
       "4    TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...   \n",
       "..                                                 ...   \n",
       "529  CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...   \n",
       "530  TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...   \n",
       "531  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "532  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "533  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...             0.48   \n",
       "1    CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...             0.54   \n",
       "2    CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...             0.52   \n",
       "3    ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...             0.50   \n",
       "4    CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...             0.50   \n",
       "..                                                 ...              ...   \n",
       "529  CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...             0.58   \n",
       "530  CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...             0.48   \n",
       "531  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "532  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "533  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                      4            0.48            0.48              True  \n",
       "1                      4            0.60            0.48              True  \n",
       "2                      4            0.56            0.48              True  \n",
       "3                      4            0.52            0.48              True  \n",
       "4                      4            0.52            0.48              True  \n",
       "..                   ...             ...             ...               ...  \n",
       "529                    4            0.52            0.64              True  \n",
       "530                    4            0.48            0.48              True  \n",
       "531                    4            0.48            0.60              True  \n",
       "532                    4            0.48            0.64              True  \n",
       "533                    4            0.48            0.52              True  \n",
       "\n",
       "[534 rows x 14 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1a7eb377-dce8-4e92-aa44-bbdec9172d27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821073</td>\n",
       "      <td>73821123</td>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821073-73821123(-)</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...</td>\n",
       "      <td>AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821085</td>\n",
       "      <td>73821135</td>\n",
       "      <td>XIST_0_435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821085-73821135(-)</td>\n",
       "      <td>TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...</td>\n",
       "      <td>CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821087</td>\n",
       "      <td>73821137</td>\n",
       "      <td>XIST_0_437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821087-73821137(-)</td>\n",
       "      <td>TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...</td>\n",
       "      <td>CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821088</td>\n",
       "      <td>73821138</td>\n",
       "      <td>XIST_0_438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821088-73821138(-)</td>\n",
       "      <td>ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...</td>\n",
       "      <td>ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821091</td>\n",
       "      <td>73821141</td>\n",
       "      <td>XIST_0_441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821091-73821141(-)</td>\n",
       "      <td>TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...</td>\n",
       "      <td>CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852654</td>\n",
       "      <td>73852704</td>\n",
       "      <td>XIST_5_11273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852654-73852704(-)</td>\n",
       "      <td>CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...</td>\n",
       "      <td>CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852663</td>\n",
       "      <td>73852713</td>\n",
       "      <td>XIST_5_11282</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852663-73852713(-)</td>\n",
       "      <td>TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...</td>\n",
       "      <td>CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>534 rows Ã 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73821073  73821123    XIST_0_423     .      -   \n",
       "1      chrX  73821085  73821135    XIST_0_435     .      -   \n",
       "2      chrX  73821087  73821137    XIST_0_437     .      -   \n",
       "3      chrX  73821088  73821138    XIST_0_438     .      -   \n",
       "4      chrX  73821091  73821141    XIST_0_441     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "529    chrX  73852654  73852704  XIST_5_11273     .      -   \n",
       "530    chrX  73852663  73852713  XIST_5_11282     .      -   \n",
       "531    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "532    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "533    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73821073-73821123(-)   \n",
       "1    chrX:73821085-73821135(-)   \n",
       "2    chrX:73821087-73821137(-)   \n",
       "3    chrX:73821088-73821138(-)   \n",
       "4    chrX:73821091-73821141(-)   \n",
       "..                         ...   \n",
       "529  chrX:73852654-73852704(-)   \n",
       "530  chrX:73852663-73852713(-)   \n",
       "531  chrX:73852687-73852737(-)   \n",
       "532  chrX:73852689-73852739(-)   \n",
       "533  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...   \n",
       "1    TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...   \n",
       "2    TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...   \n",
       "3    ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...   \n",
       "4    TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...   \n",
       "..                                                 ...   \n",
       "529  CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...   \n",
       "530  TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...   \n",
       "531  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "532  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "533  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...             0.48   \n",
       "1    CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...             0.54   \n",
       "2    CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...             0.52   \n",
       "3    ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...             0.50   \n",
       "4    CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...             0.50   \n",
       "..                                                 ...              ...   \n",
       "529  CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...             0.58   \n",
       "530  CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...             0.48   \n",
       "531  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "532  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "533  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                      4            0.48            0.48              True  \n",
       "1                      4            0.60            0.48              True  \n",
       "2                      4            0.56            0.48              True  \n",
       "3                      4            0.52            0.48              True  \n",
       "4                      4            0.52            0.48              True  \n",
       "..                   ...             ...             ...               ...  \n",
       "529                    4            0.52            0.64              True  \n",
       "530                    4            0.48            0.48              True  \n",
       "531                    4            0.48            0.60              True  \n",
       "532                    4            0.48            0.64              True  \n",
       "533                    4            0.48            0.52              True  \n",
       "\n",
       "[534 rows x 14 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Candidate kmers\n",
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e482ac72-5ed2-46dd-a8bb-1950d1a8aa14",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers.to_csv((out_dir + 'kmers_candidates_filtered.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bede588-f85c-4321-a15d-d48498d802d2",
   "metadata": {},
   "source": [
    "After filtering for desired sequence features, the next step is to make sure that our k-mers uniquely target our gene of interest (as opposed to other genes).\n",
    "\n",
    "For this, we rely on using BLAST against our custom database which contains all exons in the human transcriptome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3ace2aec-f433-49c6-90c6-8a971232be41",
   "metadata": {},
   "outputs": [],
   "source": [
    "## We are exporting the transcript sequence as that's the one that has to be blasted against the human transcriptome\n",
    "write_fasta(kmers['name'], kmers['transcript_seq'], (out_dir + 'kmers_candidates_filtered_transcript_seqs.fa'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0f5255b6-3eb9-4912-a75a-c2dcf6cad93a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run BLAST - you can also adjust the parameters here, though the outfmt needs to be consistent for downstream functions to work\n",
    "command = [\n",
    "    (blast_exec_path + 'blastn'),\n",
    "    '-query', (out_dir + 'kmers_candidates_filtered_transcript_seqs.fa'), ## Sequences of targeted regions\n",
    "    '-db', blast_db,                     # BLAST database\n",
    "    '-out', (out_dir + 'kmers_candidates_filtered_blast_output.txt'),  # Output file\n",
    "    '-outfmt', '6',\n",
    "    '-strand', 'plus',\n",
    "    '-evalue', '1e-3',  # More lenient e-value to detect potential off-targets\n",
    "    '-dust', 'no'       # Turn off low-complexity filter\n",
    "]\n",
    "\n",
    "# Run the command\n",
    "result = subprocess.run(command, capture_output=True, text=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0cb79219-58bb-4344-b02d-4eef45ff18e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Â Read in the blast output\n",
    "blast_columns = [\n",
    "    \"name\",   # Query Seq-id\n",
    "    \"sseqid\",   # Subject Seq-id\n",
    "    \"pident\",   # Percentage of identical matches\n",
    "    \"length\",   # Alignment length\n",
    "    \"mismatch\", # Number of mismatches\n",
    "    \"gapopen\",  # Number of gap openings\n",
    "    \"qstart\",   # Start of alignment in query\n",
    "    \"qend\",     # End of alignment in query\n",
    "    \"sstart\",   # Start of alignment in subject\n",
    "    \"send\",     # End of alignment in subject\n",
    "    \"evalue\",   # Expect value\n",
    "    \"bitscore\"  # Bit score\n",
    "]\n",
    "\n",
    "blast_res = pd.read_csv(\n",
    "    (out_dir + 'kmers_candidates_filtered_blast_output.txt'),\n",
    "    sep='\\t', \n",
    "    header=None, \n",
    "    names=blast_columns\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6aae54ac-97a5-426f-a88f-8352885a6493",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Â Extract the gene ID\n",
    "blast_res['sgeneid'] = blast_res['sseqid'].str.split('::').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "66bd2f88-1ce9-4953-b238-4fbcb1349a2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sgeneid\n",
       "XIST    641\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_res['sgeneid'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "b219ece2-679e-4132-9c2a-50a5f21f5974",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    6.410000e+02\n",
       "mean     1.067253e-12\n",
       "std      1.594191e-11\n",
       "min      1.300000e-18\n",
       "25%      1.300000e-18\n",
       "50%      1.300000e-18\n",
       "75%      1.300000e-18\n",
       "max      2.840000e-10\n",
       "Name: evalue, dtype: float64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_res['evalue'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0d97396-3b89-4a62-9581-0027bff69048",
   "metadata": {},
   "source": [
    "Not all BLAST hits will be off-targets. Hopefully, our gene of interest is included in the BLAST output. We therefore need to filter for hits with different gene IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f2d394e1-35a7-461a-b94e-fa9db71984b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "offtargets = blast_res['name'][blast_res['sgeneid']!=gene_ID].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "73ace4d0-5850-47e7-9770-3358b78237c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([], dtype=object)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "offtargets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b25ebea4-fa7c-44f9-b7f0-e6fb32149341",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove off targets\n",
    "kmers = kmers[kmers['name'].isin(offtargets)==False].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "1a1d18ca-0824-4094-94f5-d4a5a0645366",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821073</td>\n",
       "      <td>73821123</td>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821073-73821123(-)</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...</td>\n",
       "      <td>AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821085</td>\n",
       "      <td>73821135</td>\n",
       "      <td>XIST_0_435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821085-73821135(-)</td>\n",
       "      <td>TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...</td>\n",
       "      <td>CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821087</td>\n",
       "      <td>73821137</td>\n",
       "      <td>XIST_0_437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821087-73821137(-)</td>\n",
       "      <td>TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...</td>\n",
       "      <td>CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821088</td>\n",
       "      <td>73821138</td>\n",
       "      <td>XIST_0_438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821088-73821138(-)</td>\n",
       "      <td>ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...</td>\n",
       "      <td>ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821091</td>\n",
       "      <td>73821141</td>\n",
       "      <td>XIST_0_441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821091-73821141(-)</td>\n",
       "      <td>TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...</td>\n",
       "      <td>CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852654</td>\n",
       "      <td>73852704</td>\n",
       "      <td>XIST_5_11273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852654-73852704(-)</td>\n",
       "      <td>CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...</td>\n",
       "      <td>CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852663</td>\n",
       "      <td>73852713</td>\n",
       "      <td>XIST_5_11282</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852663-73852713(-)</td>\n",
       "      <td>TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...</td>\n",
       "      <td>CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>534 rows Ã 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73821073  73821123    XIST_0_423     .      -   \n",
       "1      chrX  73821085  73821135    XIST_0_435     .      -   \n",
       "2      chrX  73821087  73821137    XIST_0_437     .      -   \n",
       "3      chrX  73821088  73821138    XIST_0_438     .      -   \n",
       "4      chrX  73821091  73821141    XIST_0_441     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "529    chrX  73852654  73852704  XIST_5_11273     .      -   \n",
       "530    chrX  73852663  73852713  XIST_5_11282     .      -   \n",
       "531    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "532    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "533    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73821073-73821123(-)   \n",
       "1    chrX:73821085-73821135(-)   \n",
       "2    chrX:73821087-73821137(-)   \n",
       "3    chrX:73821088-73821138(-)   \n",
       "4    chrX:73821091-73821141(-)   \n",
       "..                         ...   \n",
       "529  chrX:73852654-73852704(-)   \n",
       "530  chrX:73852663-73852713(-)   \n",
       "531  chrX:73852687-73852737(-)   \n",
       "532  chrX:73852689-73852739(-)   \n",
       "533  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...   \n",
       "1    TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...   \n",
       "2    TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...   \n",
       "3    ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...   \n",
       "4    TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...   \n",
       "..                                                 ...   \n",
       "529  CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...   \n",
       "530  TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...   \n",
       "531  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "532  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "533  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...             0.48   \n",
       "1    CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...             0.54   \n",
       "2    CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...             0.52   \n",
       "3    ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...             0.50   \n",
       "4    CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...             0.50   \n",
       "..                                                 ...              ...   \n",
       "529  CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...             0.58   \n",
       "530  CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...             0.48   \n",
       "531  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "532  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "533  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                      4            0.48            0.48              True  \n",
       "1                      4            0.60            0.48              True  \n",
       "2                      4            0.56            0.48              True  \n",
       "3                      4            0.52            0.48              True  \n",
       "4                      4            0.52            0.48              True  \n",
       "..                   ...             ...             ...               ...  \n",
       "529                    4            0.52            0.64              True  \n",
       "530                    4            0.48            0.48              True  \n",
       "531                    4            0.48            0.60              True  \n",
       "532                    4            0.48            0.64              True  \n",
       "533                    4            0.48            0.52              True  \n",
       "\n",
       "[534 rows x 14 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56719d4a-ad9f-4ff0-9e89-0c2fcbc55041",
   "metadata": {},
   "source": [
    "At this point, we have effectively acquired a set of usable probes. They don't overlap undesirable regions (repeats/polymorphism), have desirable sequence features (GC content, specific nucleotides) and are specific to our gene of interest.\n",
    "\n",
    "In this particular case, we still have a lot of possible k-mers (much more than the number of probes we intend to design). We can therefore choose to prioritise k-mers with shorter homopolymer stretches, as these are also discouraged by the [10x recommendations](https://cdn.10xgenomics.com/image/upload/v1697739385/support-documents/CG000621_CustomProbeDesign_TechNote_RevC.pdf).\n",
    "\n",
    "However, at this stage you might want to consider ranking probes in a diffferent way, depending on your application.\n",
    "\n",
    "After having ranked our k-mers in whatever way we think is reasonable at this stage, we can proceed with selecting the top probe, then removing all overlapping/adjacent probes (within a window determined by `probe_offset`).\n",
    "\n",
    "Since we have so many available probes, we will be increasing `probe_offset` from `100 (default)` to `1000`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "c0d8c8f9-0c4c-46a4-bb8f-16782e47d379",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sort in increasing homopolymer length\n",
    "kmers = kmers.sort_values('longest_homopolymer', ascending=True).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d951231a-8c94-4473-978d-b202c2ded194",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850556</td>\n",
       "      <td>73850606</td>\n",
       "      <td>XIST_5_9175</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850556-73850606(-)</td>\n",
       "      <td>TGGCAAGGACCAGAATGGATCACAGATGATCGTTGGCCAACAGGTG...</td>\n",
       "      <td>CTGCCACCTGTTGGCCAACGATCATCTGTGATCCATTCTGGTCCTT...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850818</td>\n",
       "      <td>73850868</td>\n",
       "      <td>XIST_5_9437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850818-73850868(-)</td>\n",
       "      <td>AGTGCTGCCTCTTGCAGTGCTGGATATCTGGCTGTGTGGTCTGAAC...</td>\n",
       "      <td>GGAGGTTCAGACCACACAGCCAGATATCCAGCACTGCAAGAGGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>2</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73826123</td>\n",
       "      <td>73826173</td>\n",
       "      <td>XIST_0_5473</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73826123-73826173(-)</td>\n",
       "      <td>CAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCTTCCT...</td>\n",
       "      <td>AAGAAGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73826126</td>\n",
       "      <td>73826176</td>\n",
       "      <td>XIST_0_5476</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73826126-73826176(-)</td>\n",
       "      <td>ACACAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCTT...</td>\n",
       "      <td>AAGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACATCT...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73826127</td>\n",
       "      <td>73826177</td>\n",
       "      <td>XIST_0_5477</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73826127-73826177(-)</td>\n",
       "      <td>TACACAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCT...</td>\n",
       "      <td>AGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACATCTG...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850757</td>\n",
       "      <td>73850807</td>\n",
       "      <td>XIST_5_9376</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850757-73850807(-)</td>\n",
       "      <td>TGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccaccc...</td>\n",
       "      <td>tggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAA...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>6</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850758</td>\n",
       "      <td>73850808</td>\n",
       "      <td>XIST_5_9377</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850758-73850808(-)</td>\n",
       "      <td>CTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccacc...</td>\n",
       "      <td>ggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAAT...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>6</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850763</td>\n",
       "      <td>73850813</td>\n",
       "      <td>XIST_5_9382</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850763-73850813(-)</td>\n",
       "      <td>TTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCcccc...</td>\n",
       "      <td>gtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGA...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>6</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850764</td>\n",
       "      <td>73850814</td>\n",
       "      <td>XIST_5_9383</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850764-73850814(-)</td>\n",
       "      <td>ATTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccc...</td>\n",
       "      <td>tggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGAG...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>6</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852326</td>\n",
       "      <td>73852376</td>\n",
       "      <td>XIST_5_10945</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852326-73852376(-)</td>\n",
       "      <td>GGGGCTGCGGATACCTGGTTTTATTATTTTTTCTTTGCCCAACGGG...</td>\n",
       "      <td>CGGCCCCGTTGGGCAAAGAAAAAATAATAAAACCAGGTATCCGCAG...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>6</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>534 rows Ã 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73850556  73850606   XIST_5_9175     .      -   \n",
       "1      chrX  73850818  73850868   XIST_5_9437     .      -   \n",
       "2      chrX  73826123  73826173   XIST_0_5473     .      -   \n",
       "3      chrX  73826126  73826176   XIST_0_5476     .      -   \n",
       "4      chrX  73826127  73826177   XIST_0_5477     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "529    chrX  73850757  73850807   XIST_5_9376     .      -   \n",
       "530    chrX  73850758  73850808   XIST_5_9377     .      -   \n",
       "531    chrX  73850763  73850813   XIST_5_9382     .      -   \n",
       "532    chrX  73850764  73850814   XIST_5_9383     .      -   \n",
       "533    chrX  73852326  73852376  XIST_5_10945     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73850556-73850606(-)   \n",
       "1    chrX:73850818-73850868(-)   \n",
       "2    chrX:73826123-73826173(-)   \n",
       "3    chrX:73826126-73826176(-)   \n",
       "4    chrX:73826127-73826177(-)   \n",
       "..                         ...   \n",
       "529  chrX:73850757-73850807(-)   \n",
       "530  chrX:73850758-73850808(-)   \n",
       "531  chrX:73850763-73850813(-)   \n",
       "532  chrX:73850764-73850814(-)   \n",
       "533  chrX:73852326-73852376(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    TGGCAAGGACCAGAATGGATCACAGATGATCGTTGGCCAACAGGTG...   \n",
       "1    AGTGCTGCCTCTTGCAGTGCTGGATATCTGGCTGTGTGGTCTGAAC...   \n",
       "2    CAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCTTCCT...   \n",
       "3    ACACAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCTT...   \n",
       "4    TACACAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCT...   \n",
       "..                                                 ...   \n",
       "529  TGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccaccc...   \n",
       "530  CTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccacc...   \n",
       "531  TTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCcccc...   \n",
       "532  ATTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccc...   \n",
       "533  GGGGCTGCGGATACCTGGTTTTATTATTTTTTCTTTGCCCAACGGG...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    CTGCCACCTGTTGGCCAACGATCATCTGTGATCCATTCTGGTCCTT...             0.54   \n",
       "1    GGAGGTTCAGACCACACAGCCAGATATCCAGCACTGCAAGAGGCAG...             0.56   \n",
       "2    AAGAAGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACA...             0.52   \n",
       "3    AAGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACATCT...             0.52   \n",
       "4    AGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACATCTG...             0.52   \n",
       "..                                                 ...              ...   \n",
       "529  tggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAA...             0.58   \n",
       "530  ggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAAT...             0.60   \n",
       "531  gtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGA...             0.54   \n",
       "532  tggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGAG...             0.52   \n",
       "533  CGGCCCCGTTGGGCAAAGAAAAAATAATAAAACCAGGTATCCGCAG...             0.52   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                      2            0.56            0.52              True  \n",
       "1                      2            0.56            0.56              True  \n",
       "2                      2            0.52            0.52              True  \n",
       "3                      2            0.56            0.48              True  \n",
       "4                      2            0.56            0.48              True  \n",
       "..                   ...             ...             ...               ...  \n",
       "529                    6            0.68            0.48              True  \n",
       "530                    6            0.68            0.52              True  \n",
       "531                    6            0.60            0.48              True  \n",
       "532                    6            0.56            0.48              True  \n",
       "533                    6            0.52            0.52              True  \n",
       "\n",
       "[534 rows x 14 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "9d476c7a-f47e-4389-abf1-8b2a567ce668",
   "metadata": {},
   "outputs": [],
   "source": [
    "## We have a lot of probes here - increasing the offset to 1000 bp to space them out\n",
    "probe_offset = 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "3504dcc6-d55f-4491-bcdd-5049eab3b40e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nfs/team205/is10/miniconda/envs/gene2probe_env/lib/python3.11/site-packages/gene2probe/gene2probe.py:203: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  start = int(probe_df['start'].iloc[i])\n",
      "/nfs/team205/is10/miniconda/envs/gene2probe_env/lib/python3.11/site-packages/gene2probe/gene2probe.py:204: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  end = int(probe_df['end'].iloc[i])\n"
     ]
    }
   ],
   "source": [
    "selected_probes_list = []\n",
    "df = kmers.copy()\n",
    "i = 0\n",
    "while i < n_desired_probes:\n",
    "    selected_probes_list.append(df.iloc[0:1,:])\n",
    "    ## Going for a high offset as we have a large span:\n",
    "    df = remove_overlapping_probes(df,\n",
    "                                   df['name'][0], \n",
    "                                   offset=probe_offset).reset_index(drop=True).copy()\n",
    "    i +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "069f0bab-21de-4f48-a902-767f77ea4260",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_probes_df = pd.concat(selected_probes_list, axis=0).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6ed7dbf9-aef5-4723-ad14-f6892251e7e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850556</td>\n",
       "      <td>73850606</td>\n",
       "      <td>XIST_5_9175</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850556-73850606(-)</td>\n",
       "      <td>TGGCAAGGACCAGAATGGATCACAGATGATCGTTGGCCAACAGGTG...</td>\n",
       "      <td>CTGCCACCTGTTGGCCAACGATCATCTGTGATCCATTCTGGTCCTT...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73826123</td>\n",
       "      <td>73826173</td>\n",
       "      <td>XIST_0_5473</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73826123-73826173(-)</td>\n",
       "      <td>CAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCTTCCT...</td>\n",
       "      <td>AAGAAGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73851704</td>\n",
       "      <td>73851754</td>\n",
       "      <td>XIST_5_10323</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73851704-73851754(-)</td>\n",
       "      <td>AGATGGATGATAGCAGGTCAGGCAGAGGAAGTCATGTGCATTGCAT...</td>\n",
       "      <td>GCTCATGCAATGCACATGACTTCCTCTGCCTGACCTGCTATCATCC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     start       end          name score strand  \\\n",
       "0    chrX  73850556  73850606   XIST_5_9175     .      -   \n",
       "1    chrX  73826123  73826173   XIST_0_5473     .      -   \n",
       "2    chrX  73851704  73851754  XIST_5_10323     .      -   \n",
       "\n",
       "                  kmer_coord  \\\n",
       "0  chrX:73850556-73850606(-)   \n",
       "1  chrX:73826123-73826173(-)   \n",
       "2  chrX:73851704-73851754(-)   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  TGGCAAGGACCAGAATGGATCACAGATGATCGTTGGCCAACAGGTG...   \n",
       "1  CAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCTTCCT...   \n",
       "2  AGATGGATGATAGCAGGTCAGGCAGAGGAAGTCATGTGCATTGCAT...   \n",
       "\n",
       "                                           probe_seq  GC_content_full  \\\n",
       "0  CTGCCACCTGTTGGCCAACGATCATCTGTGATCCATTCTGGTCCTT...             0.54   \n",
       "1  AAGAAGGAAGCTGTTGCTGGCAGGTGCTTCTCAGAAGTCTGGCACA...             0.52   \n",
       "2  GCTCATGCAATGCACATGACTTCCTCTGCCTGACCTGCTATCATCC...             0.50   \n",
       "\n",
       "   longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                    2            0.56            0.52              True  \n",
       "1                    2            0.52            0.52              True  \n",
       "2                    2            0.48            0.52              True  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_probes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "9ab3f7e6-a39f-417c-8b40-92eb9e3e2908",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGGCAAGGACCAGAATGGATCACAGATGATCGTTGGCCAACAGGTGGCAG\n",
      "CAGATGTGCCAGACTTCTGAGAAGCACCTGCCAGCAACAGCTTCCTTCTT\n",
      "AGATGGATGATAGCAGGTCAGGCAGAGGAAGTCATGTGCATTGCATGAGC\n"
     ]
    }
   ],
   "source": [
    "for seq in selected_probes_df['transcript_seq']:\n",
    "    print(seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52a8e31d-f567-410c-bd05-54b1e2731f6f",
   "metadata": {},
   "source": [
    "And we are done! We have now selected three potential probes for our gene.\n",
    "\n",
    "If our probes are meant to be split, we can additionally generate these columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "3ec41fa9-1992-49cc-8f8f-88da8fdd07e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "if split_nt is not None: \n",
    "    ## Make split probes\n",
    "    selected_probes_df['probe_seq_LHS'] = [seq[0:split_nt] for seq in selected_probes_df['probe_seq']]\n",
    "    selected_probes_df['probe_seq_RHS'] = [seq[split_nt: probe_length] for seq in selected_probes_df['probe_seq']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "76b0b300-b3ca-49db-a600-c82961d30ce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Also add gene_ID for completeness\n",
    "selected_probes_df['gene_ID'] = gene_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "1ce62661-8a1b-4ef6-83e1-2bd603806975",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export selected probes as dataframe:\n",
    "selected_probes_df.to_csv((out_dir + 'kmers_selected_probes.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77b84ece-a367-4a6b-87c1-1c2a3858ed8a",
   "metadata": {},
   "source": [
    "We always recommend additionally performing a manual BLAST of these probe sequences to make sure that there are no off-target effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e5eb6cd-cc4e-4354-bc83-647c2003a9fc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gene2probe_env",
   "language": "python",
   "name": "gene2probe_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
