{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "rising-carnival",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import sys\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "from gene2probe import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "organic-parameter",
   "metadata": {},
   "source": [
    "In the previous tutorials, we have seen how to generate probes starting from a [gene](https://github.com/Teichlab/gene2probe/blob/main/notebooks/002_make_exon_probes_XIST.ipynb) or a [specific region](https://github.com/Teichlab/gene2probe/blob/main/notebooks/003_make_probes_for_custom_region_CD45RA.ipynb).\n",
    "\n",
    "But what if we are interested in designing probes against a sequence that is not included in the human genome (e.g., GFP)?\n",
    "We can achieve that by directly providing the sequence as input and following all subsequent steps.\n",
    "\n",
    "We note that we don't recommend this approach for the detection of bacterial or viral reads for metagenomics analyses, as there are many additional considerations that have to be taken into account to ensure that a probe is specifically detecting a certain species."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "great-corruption",
   "metadata": {},
   "source": [
    "### 1. Specify parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "wanted-merchant",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Specify gene of interest and feature of interest\n",
    "gene_ID = 'EGFP'\n",
    "sequence = 'GTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACAAGTAA'\n",
    "\n",
    "## specify output directory\n",
    "out_dir = '../sample_run/probeDesign_' + gene_ID +'/'\n",
    "## Create output directory\n",
    "os.makedirs(out_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "interested-trail",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACAAGTAA'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "internal-rapid",
   "metadata": {},
   "source": [
    "Since we already provide the sequence, we just need a blast database to check for offtargets. \n",
    "\n",
    "We can use the one we generated in the [previous tutorial](https://github.com/Teichlab/gene2probe/blob/main/notebooks/001_make_blast_database.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "traditional-bronze",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_db = '../hg38_resources/blastdb/hg38_ncbiRefSeq_transcripts_db' ## Database of all human transcripts to blast against"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "professional-stable",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to blast binaries.\n",
    "## Replace with your conda environment\n",
    "## This can also be omitted if you started the jupyter session from within the gene2probe conda environment\n",
    "blast_exec_path = '/nfs/team205/is10/miniconda/envs/gene2probe_env/bin/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "outside-acrobat",
   "metadata": {},
   "source": [
    "Finally, we need to provide a set of parameters related to our probe's length, at which nucleotide it's split (if at all), the acceptable range for GC content and any specific requirements for individual nucleotides.\n",
    "\n",
    "Here we are following the [recommendations of 10x Genomics for custom probes for VisiumHD/VisiumFFPE/Flex](https://cdn.10xgenomics.com/image/upload/v1697739385/support-documents/CG000621_CustomProbeDesign_TechNote_RevC.pdf)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "exterior-saturday",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Additional parameters regarding how the probe should look like\n",
    "probe_length = 50 ## Length of probe in nucleotides\n",
    "split_nt = 25 ## Index of nucleotide to split the probe at (start of RHS) - set to None if splitting probe is not needed\n",
    "min_GC = 0.44 ## Minimum GC content for probe (if split probe, applied to both LHS and RHS)\n",
    "max_GC = 0.72 ## Maximum GC content for probe (if split probe, applied to both LHS and RHS)\n",
    "required_nts = {24: 'T'} ## Dictionary of index (0-based) for required nts - by default, 25th nucleotide must be a T - set to None if no requirements\n",
    "probe_offset = 100 ## Minimum distance between probes - 10 bp is the recommended minimum by 10x, this can also be adjusted depending on how many probes pass other cutoffs\n",
    "n_desired_probes =3 ## Number of probes to be designed.\n",
    "min_mismatches = 5 ## Minimum number of mismatches (in at least LHS or RHS) - here we require in both to be more conservative"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "saved-behavior",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Optionally, we can also specify adapters that have to be added to the probes.\n",
    "## For example, for visiumHD:\n",
    "LHS_pref = 'CCTTGGCACCCGAGAATTCCA' ## Will be added to the 5' of the LHS probe\n",
    "LHS_suff = '' ## Will be added to the 3' of the LHS probe\n",
    "RHS_pref = '/5Phos/' ## Will be added to the 5' of the RHS probe\n",
    "RHS_suff = 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA' ## Will be added to the 3' of the RHS probe\n",
    "\n",
    "## Leave as empty strings if you don't want to use them"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hazardous-bridal",
   "metadata": {},
   "source": [
    "### 2. Generate k-mers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cardiovascular-title",
   "metadata": {},
   "source": [
    "Since we started with a sequence, we can directly make kmers from it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "literary-excuse",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_kmers_from_seq(seq, k, outfasta, name_pref):\n",
    "    \"\"\"\n",
    "    Given a sequence and an integer k, split the sequence into all possible kmers.\n",
    "    Kmers are named based on their index, preceded by an optional prefix.\n",
    "    The function exports all kmers into a fasta file (path specified as outfasta)\n",
    "    \"\"\"\n",
    "    l = len(seq)\n",
    "    if l < k:\n",
    "        raise ValueError('k should be smaller than the length of the sequence')\n",
    "    i = 0\n",
    "    kmers_names = []\n",
    "    kmers_seqs = []\n",
    "    kmers_start = []\n",
    "    kmers_end = []\n",
    "    while i < (l-k):\n",
    "        kmers_names.append((name_pref + str(i)))\n",
    "        kmers_seqs.append(seq[i: (i+k)])\n",
    "        kmers_start.append(i)\n",
    "        kmers_end.append((i + k))\n",
    "        i+=1\n",
    "    ## Export fasta\n",
    "    write_fasta(kmers_names, kmers_seqs, outfasta)\n",
    "\n",
    "    ## Create dataframe\n",
    "    kmers_df = pd.DataFrame({\n",
    "        'seqname': 'custom_seq',\n",
    "        'start': kmers_start,\n",
    "        'end': kmers_end,\n",
    "        'name': kmers_names,\n",
    "        'transcript_seq': kmers_seqs\n",
    "    })\n",
    "    return kmers_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3f33cf5c-da37-4481-86e3-01dce61f240e",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers = generate_kmers_from_seq(sequence, k = probe_length, outfasta=(out_dir + 'all_kmers_unfiltered.fa'), name_pref=(gene_ID + '_'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "controlling-stake",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>transcript_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>EGFP_0</td>\n",
       "      <td>GTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "      <td>EGFP_1</td>\n",
       "      <td>TGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>2</td>\n",
       "      <td>52</td>\n",
       "      <td>EGFP_2</td>\n",
       "      <td>GCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>3</td>\n",
       "      <td>53</td>\n",
       "      <td>EGFP_3</td>\n",
       "      <td>CCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>4</td>\n",
       "      <td>54</td>\n",
       "      <td>EGFP_4</td>\n",
       "      <td>CCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>629</td>\n",
       "      <td>679</td>\n",
       "      <td>EGFP_629</td>\n",
       "      <td>GGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>630</td>\n",
       "      <td>680</td>\n",
       "      <td>EGFP_630</td>\n",
       "      <td>GAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>631</td>\n",
       "      <td>681</td>\n",
       "      <td>EGFP_631</td>\n",
       "      <td>AGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>632</td>\n",
       "      <td>682</td>\n",
       "      <td>EGFP_632</td>\n",
       "      <td>GTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>633</td>\n",
       "      <td>683</td>\n",
       "      <td>EGFP_633</td>\n",
       "      <td>TTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>634 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  start  end      name  \\\n",
       "0    custom_seq      0   50    EGFP_0   \n",
       "1    custom_seq      1   51    EGFP_1   \n",
       "2    custom_seq      2   52    EGFP_2   \n",
       "3    custom_seq      3   53    EGFP_3   \n",
       "4    custom_seq      4   54    EGFP_4   \n",
       "..          ...    ...  ...       ...   \n",
       "629  custom_seq    629  679  EGFP_629   \n",
       "630  custom_seq    630  680  EGFP_630   \n",
       "631  custom_seq    631  681  EGFP_631   \n",
       "632  custom_seq    632  682  EGFP_632   \n",
       "633  custom_seq    633  683  EGFP_633   \n",
       "\n",
       "                                        transcript_seq  \n",
       "0    GTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGT...  \n",
       "1    TGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTT...  \n",
       "2    GCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTC...  \n",
       "3    CCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCA...  \n",
       "4    CCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAG...  \n",
       "..                                                 ...  \n",
       "629  GGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTG...  \n",
       "630  GAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGT...  \n",
       "631  AGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTA...  \n",
       "632  GTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTAC...  \n",
       "633  TTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACA...  \n",
       "\n",
       "[634 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "abstract-container",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export unfiltered\n",
    "kmers.to_csv((out_dir + 'kmers_all.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "driving-announcement",
   "metadata": {},
   "source": [
    "### 3. Exclude annotated repeats/polymorphism"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "piano-staff",
   "metadata": {},
   "source": [
    "Since we started from a custom sequence, we skip this part.\n",
    "(alternatively, if a SNP is known, we can provide it as a relative coordinate to the custom sequence)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "handed-signature",
   "metadata": {},
   "source": [
    "### 4. Filter for desirable sequence features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "protective-greeting",
   "metadata": {},
   "source": [
    "The next step is to consider their sequence features.\n",
    "\n",
    "For this, we first extract the sequences of each k-mer, and then estimate features such as GC content and the presence of desired nucleotides in specific positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "toxic-image",
   "metadata": {},
   "outputs": [],
   "source": [
    "## We can read in the sequences and simultaneously monitor GC content and count the longest homopolymer stretch\n",
    "kmers_seq_stats = get_sequence_stats((out_dir + 'all_kmers_unfiltered.fa'), probe_length, split_nt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "weighted-period",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Combining with our dataframe\n",
    "kmers = pd.merge(kmers, kmers_seq_stats, on = 'transcript_seq', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "julian-electronics",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>0</td>\n",
       "      <td>50</td>\n",
       "      <td>EGFP_0</td>\n",
       "      <td>GTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGT...</td>\n",
       "      <td>EGFP_0</td>\n",
       "      <td>CTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGATGG...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>1</td>\n",
       "      <td>51</td>\n",
       "      <td>EGFP_1</td>\n",
       "      <td>TGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTT...</td>\n",
       "      <td>EGFP_1</td>\n",
       "      <td>GCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGATG...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>2</td>\n",
       "      <td>52</td>\n",
       "      <td>EGFP_2</td>\n",
       "      <td>GCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTC...</td>\n",
       "      <td>EGFP_2</td>\n",
       "      <td>CGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGAT...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>3</td>\n",
       "      <td>53</td>\n",
       "      <td>EGFP_3</td>\n",
       "      <td>CCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCA...</td>\n",
       "      <td>EGFP_3</td>\n",
       "      <td>ACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGA...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>4</td>\n",
       "      <td>54</td>\n",
       "      <td>EGFP_4</td>\n",
       "      <td>CCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAG...</td>\n",
       "      <td>EGFP_4</td>\n",
       "      <td>CACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGG...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>629</td>\n",
       "      <td>679</td>\n",
       "      <td>EGFP_629</td>\n",
       "      <td>GGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTG...</td>\n",
       "      <td>EGFP_629</td>\n",
       "      <td>TGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACGAA...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>630</td>\n",
       "      <td>680</td>\n",
       "      <td>EGFP_630</td>\n",
       "      <td>GAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGT...</td>\n",
       "      <td>EGFP_630</td>\n",
       "      <td>TTGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACGA...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>631</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>631</td>\n",
       "      <td>681</td>\n",
       "      <td>EGFP_631</td>\n",
       "      <td>AGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTA...</td>\n",
       "      <td>EGFP_631</td>\n",
       "      <td>CTTGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACG...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>632</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>632</td>\n",
       "      <td>682</td>\n",
       "      <td>EGFP_632</td>\n",
       "      <td>GTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTAC...</td>\n",
       "      <td>EGFP_632</td>\n",
       "      <td>ACTTGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCAC...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>633</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>633</td>\n",
       "      <td>683</td>\n",
       "      <td>EGFP_633</td>\n",
       "      <td>TTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACA...</td>\n",
       "      <td>EGFP_633</td>\n",
       "      <td>TACTTGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCA...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>3</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>634 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  start  end      name  \\\n",
       "0    custom_seq      0   50    EGFP_0   \n",
       "1    custom_seq      1   51    EGFP_1   \n",
       "2    custom_seq      2   52    EGFP_2   \n",
       "3    custom_seq      3   53    EGFP_3   \n",
       "4    custom_seq      4   54    EGFP_4   \n",
       "..          ...    ...  ...       ...   \n",
       "629  custom_seq    629  679  EGFP_629   \n",
       "630  custom_seq    630  680  EGFP_630   \n",
       "631  custom_seq    631  681  EGFP_631   \n",
       "632  custom_seq    632  682  EGFP_632   \n",
       "633  custom_seq    633  683  EGFP_633   \n",
       "\n",
       "                                        transcript_seq kmer_coord  \\\n",
       "0    GTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGT...     EGFP_0   \n",
       "1    TGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTT...     EGFP_1   \n",
       "2    GCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTC...     EGFP_2   \n",
       "3    CCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCA...     EGFP_3   \n",
       "4    CCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAG...     EGFP_4   \n",
       "..                                                 ...        ...   \n",
       "629  GGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTG...   EGFP_629   \n",
       "630  GAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGT...   EGFP_630   \n",
       "631  AGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTA...   EGFP_631   \n",
       "632  GTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTAC...   EGFP_632   \n",
       "633  TTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACA...   EGFP_633   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    CTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGATGG...             0.62   \n",
       "1    GCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGATG...             0.62   \n",
       "2    CGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGAT...             0.64   \n",
       "3    ACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGA...             0.62   \n",
       "4    CACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGG...             0.62   \n",
       "..                                                 ...              ...   \n",
       "629  TGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACGAA...             0.64   \n",
       "630  TTGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACGA...             0.62   \n",
       "631  CTTGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACG...             0.62   \n",
       "632  ACTTGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCAC...             0.62   \n",
       "633  TACTTGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCA...             0.60   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  \n",
       "0                      3            0.56            0.68  \n",
       "1                      3            0.56            0.68  \n",
       "2                      3            0.56            0.72  \n",
       "3                      3            0.52            0.72  \n",
       "4                      3            0.56            0.68  \n",
       "..                   ...             ...             ...  \n",
       "629                    3            0.56            0.72  \n",
       "630                    3            0.56            0.68  \n",
       "631                    3            0.56            0.68  \n",
       "632                    3            0.56            0.68  \n",
       "633                    3            0.52            0.68  \n",
       "\n",
       "[634 rows x 11 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ideal-greek",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "has_required_nts\n",
      "False    478\n",
      "True     156\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## Check for required nucleotides in specific positions:\n",
    "if required_nts is not None:\n",
    "    kmers['has_required_nts'] = check_for_required_nts(kmers, required_nts)\n",
    "    print(kmers['has_required_nts'].value_counts())\n",
    "    ## Filter for required nucleotides\n",
    "    kmers = kmers[kmers['has_required_nts']==True].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "mysterious-episode",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export kmers before filtering\n",
    "kmers.to_csv((out_dir + 'kmers_candidates_unfiltered.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "royal-perception",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>3</td>\n",
       "      <td>53</td>\n",
       "      <td>EGFP_3</td>\n",
       "      <td>CCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCA...</td>\n",
       "      <td>EGFP_3</td>\n",
       "      <td>ACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGA...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.72</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>EGFP_7</td>\n",
       "      <td>TCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGT...</td>\n",
       "      <td>EGFP_7</td>\n",
       "      <td>GGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACC...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>58</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>9</td>\n",
       "      <td>59</td>\n",
       "      <td>EGFP_9</td>\n",
       "      <td>CTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGT...</td>\n",
       "      <td>EGFP_9</td>\n",
       "      <td>CCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGA...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>15</td>\n",
       "      <td>65</td>\n",
       "      <td>EGFP_15</td>\n",
       "      <td>GAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCG...</td>\n",
       "      <td>EGFP_15</td>\n",
       "      <td>CCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCA...</td>\n",
       "      <td>0.66</td>\n",
       "      <td>3</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>593</td>\n",
       "      <td>643</td>\n",
       "      <td>EGFP_593</td>\n",
       "      <td>AGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTG...</td>\n",
       "      <td>EGFP_593</td>\n",
       "      <td>CGGTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGG...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>606</td>\n",
       "      <td>656</td>\n",
       "      <td>EGFP_606</td>\n",
       "      <td>AAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGA...</td>\n",
       "      <td>EGFP_606</td>\n",
       "      <td>GTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCATGTGATCGC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>614</td>\n",
       "      <td>664</td>\n",
       "      <td>EGFP_614</td>\n",
       "      <td>TCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTC...</td>\n",
       "      <td>EGFP_614</td>\n",
       "      <td>TGCCGAGAGTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCAT...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>626</td>\n",
       "      <td>676</td>\n",
       "      <td>EGFP_626</td>\n",
       "      <td>GCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAG...</td>\n",
       "      <td>EGFP_626</td>\n",
       "      <td>ACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACGAACTC...</td>\n",
       "      <td>0.66</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.76</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>629</td>\n",
       "      <td>679</td>\n",
       "      <td>EGFP_629</td>\n",
       "      <td>GGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTG...</td>\n",
       "      <td>EGFP_629</td>\n",
       "      <td>TGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACGAA...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.72</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>156 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  start  end      name  \\\n",
       "0    custom_seq      3   53    EGFP_3   \n",
       "1    custom_seq      7   57    EGFP_7   \n",
       "2    custom_seq      8   58    EGFP_8   \n",
       "3    custom_seq      9   59    EGFP_9   \n",
       "4    custom_seq     15   65   EGFP_15   \n",
       "..          ...    ...  ...       ...   \n",
       "151  custom_seq    593  643  EGFP_593   \n",
       "152  custom_seq    606  656  EGFP_606   \n",
       "153  custom_seq    614  664  EGFP_614   \n",
       "154  custom_seq    626  676  EGFP_626   \n",
       "155  custom_seq    629  679  EGFP_629   \n",
       "\n",
       "                                        transcript_seq kmer_coord  \\\n",
       "0    CCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCA...     EGFP_3   \n",
       "1    TCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGT...     EGFP_7   \n",
       "2    CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...     EGFP_8   \n",
       "3    CTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGT...     EGFP_9   \n",
       "4    GAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCG...    EGFP_15   \n",
       "..                                                 ...        ...   \n",
       "151  AGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTG...   EGFP_593   \n",
       "152  AAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGA...   EGFP_606   \n",
       "153  TCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTC...   EGFP_614   \n",
       "154  GCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAG...   EGFP_626   \n",
       "155  GGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTG...   EGFP_629   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    ACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACCAGGA...             0.62   \n",
       "1    GGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACC...             0.62   \n",
       "2    CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...             0.64   \n",
       "3    CCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGA...             0.64   \n",
       "4    CCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCA...             0.66   \n",
       "..                                                 ...              ...   \n",
       "151  CGGTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGG...             0.60   \n",
       "152  GTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCATGTGATCGC...             0.64   \n",
       "153  TGCCGAGAGTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCAT...             0.64   \n",
       "154  ACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACGAACTC...             0.66   \n",
       "155  TGTACAGCTCGTCCATGCCGAGAGTGATCCCGGCGGCGGTCACGAA...             0.64   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                      3            0.52            0.72              True  \n",
       "1                      3            0.56            0.68              True  \n",
       "2                      3            0.60            0.68              True  \n",
       "3                      3            0.64            0.64              True  \n",
       "4                      3            0.64            0.68              True  \n",
       "..                   ...             ...             ...               ...  \n",
       "151                    4            0.60            0.60              True  \n",
       "152                    3            0.68            0.60              True  \n",
       "153                    3            0.72            0.56              True  \n",
       "154                    3            0.56            0.76              True  \n",
       "155                    3            0.56            0.72              True  \n",
       "\n",
       "[156 rows x 12 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "massive-philippines",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter for GC content\n",
    "kmers = filter_by_GC_content(kmers, min_GC, max_GC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "characteristic-leather",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>EGFP_7</td>\n",
       "      <td>TCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGT...</td>\n",
       "      <td>EGFP_7</td>\n",
       "      <td>GGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACC...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>58</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>9</td>\n",
       "      <td>59</td>\n",
       "      <td>EGFP_9</td>\n",
       "      <td>CTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGT...</td>\n",
       "      <td>EGFP_9</td>\n",
       "      <td>CCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGA...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>15</td>\n",
       "      <td>65</td>\n",
       "      <td>EGFP_15</td>\n",
       "      <td>GAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCG...</td>\n",
       "      <td>EGFP_15</td>\n",
       "      <td>CCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCA...</td>\n",
       "      <td>0.66</td>\n",
       "      <td>3</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>17</td>\n",
       "      <td>67</td>\n",
       "      <td>EGFP_17</td>\n",
       "      <td>GCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAG...</td>\n",
       "      <td>EGFP_17</td>\n",
       "      <td>CGCCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTC...</td>\n",
       "      <td>0.68</td>\n",
       "      <td>3</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>582</td>\n",
       "      <td>632</td>\n",
       "      <td>EGFP_582</td>\n",
       "      <td>GCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGC...</td>\n",
       "      <td>EGFP_582</td>\n",
       "      <td>TCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTTTGCTCA...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>588</td>\n",
       "      <td>638</td>\n",
       "      <td>EGFP_588</td>\n",
       "      <td>AGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGT...</td>\n",
       "      <td>EGFP_588</td>\n",
       "      <td>ACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTT...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>591</td>\n",
       "      <td>641</td>\n",
       "      <td>EGFP_591</td>\n",
       "      <td>AAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCG...</td>\n",
       "      <td>EGFP_591</td>\n",
       "      <td>GTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGT...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>593</td>\n",
       "      <td>643</td>\n",
       "      <td>EGFP_593</td>\n",
       "      <td>AGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTG...</td>\n",
       "      <td>EGFP_593</td>\n",
       "      <td>CGGTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGG...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>606</td>\n",
       "      <td>656</td>\n",
       "      <td>EGFP_606</td>\n",
       "      <td>AAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGA...</td>\n",
       "      <td>EGFP_606</td>\n",
       "      <td>GTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCATGTGATCGC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname  start  end      name  \\\n",
       "0   custom_seq      7   57    EGFP_7   \n",
       "1   custom_seq      8   58    EGFP_8   \n",
       "2   custom_seq      9   59    EGFP_9   \n",
       "3   custom_seq     15   65   EGFP_15   \n",
       "4   custom_seq     17   67   EGFP_17   \n",
       "..         ...    ...  ...       ...   \n",
       "94  custom_seq    582  632  EGFP_582   \n",
       "95  custom_seq    588  638  EGFP_588   \n",
       "96  custom_seq    591  641  EGFP_591   \n",
       "97  custom_seq    593  643  EGFP_593   \n",
       "98  custom_seq    606  656  EGFP_606   \n",
       "\n",
       "                                       transcript_seq kmer_coord  \\\n",
       "0   TCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGT...     EGFP_7   \n",
       "1   CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...     EGFP_8   \n",
       "2   CTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGT...     EGFP_9   \n",
       "3   GAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCG...    EGFP_15   \n",
       "4   GCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAG...    EGFP_17   \n",
       "..                                                ...        ...   \n",
       "94  GCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGC...   EGFP_582   \n",
       "95  AGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGT...   EGFP_588   \n",
       "96  AAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCG...   EGFP_591   \n",
       "97  AGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTG...   EGFP_593   \n",
       "98  AAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGA...   EGFP_606   \n",
       "\n",
       "                                            probe_seq  GC_content_full  \\\n",
       "0   GGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACC...             0.62   \n",
       "1   CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...             0.64   \n",
       "2   CCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGA...             0.64   \n",
       "3   CCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCA...             0.66   \n",
       "4   CGCCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTC...             0.68   \n",
       "..                                                ...              ...   \n",
       "94  TCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTTTGCTCA...             0.60   \n",
       "95  ACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTT...             0.56   \n",
       "96  GTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGT...             0.56   \n",
       "97  CGGTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGG...             0.60   \n",
       "98  GTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCATGTGATCGC...             0.64   \n",
       "\n",
       "    longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                     3            0.56            0.68              True  \n",
       "1                     3            0.60            0.68              True  \n",
       "2                     3            0.64            0.64              True  \n",
       "3                     3            0.64            0.68              True  \n",
       "4                     3            0.68            0.68              True  \n",
       "..                  ...             ...             ...               ...  \n",
       "94                    4            0.60            0.60              True  \n",
       "95                    4            0.52            0.60              True  \n",
       "96                    4            0.56            0.56              True  \n",
       "97                    4            0.60            0.60              True  \n",
       "98                    3            0.68            0.60              True  \n",
       "\n",
       "[99 rows x 12 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "saving-exception",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>EGFP_7</td>\n",
       "      <td>TCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGT...</td>\n",
       "      <td>EGFP_7</td>\n",
       "      <td>GGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACC...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>58</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>9</td>\n",
       "      <td>59</td>\n",
       "      <td>EGFP_9</td>\n",
       "      <td>CTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGT...</td>\n",
       "      <td>EGFP_9</td>\n",
       "      <td>CCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGA...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>15</td>\n",
       "      <td>65</td>\n",
       "      <td>EGFP_15</td>\n",
       "      <td>GAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCG...</td>\n",
       "      <td>EGFP_15</td>\n",
       "      <td>CCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCA...</td>\n",
       "      <td>0.66</td>\n",
       "      <td>3</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>17</td>\n",
       "      <td>67</td>\n",
       "      <td>EGFP_17</td>\n",
       "      <td>GCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAG...</td>\n",
       "      <td>EGFP_17</td>\n",
       "      <td>CGCCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTC...</td>\n",
       "      <td>0.68</td>\n",
       "      <td>3</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>582</td>\n",
       "      <td>632</td>\n",
       "      <td>EGFP_582</td>\n",
       "      <td>GCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGC...</td>\n",
       "      <td>EGFP_582</td>\n",
       "      <td>TCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTTTGCTCA...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>588</td>\n",
       "      <td>638</td>\n",
       "      <td>EGFP_588</td>\n",
       "      <td>AGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGT...</td>\n",
       "      <td>EGFP_588</td>\n",
       "      <td>ACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTT...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>591</td>\n",
       "      <td>641</td>\n",
       "      <td>EGFP_591</td>\n",
       "      <td>AAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCG...</td>\n",
       "      <td>EGFP_591</td>\n",
       "      <td>GTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGT...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>593</td>\n",
       "      <td>643</td>\n",
       "      <td>EGFP_593</td>\n",
       "      <td>AGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTG...</td>\n",
       "      <td>EGFP_593</td>\n",
       "      <td>CGGTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGG...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>606</td>\n",
       "      <td>656</td>\n",
       "      <td>EGFP_606</td>\n",
       "      <td>AAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGA...</td>\n",
       "      <td>EGFP_606</td>\n",
       "      <td>GTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCATGTGATCGC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname  start  end      name  \\\n",
       "0   custom_seq      7   57    EGFP_7   \n",
       "1   custom_seq      8   58    EGFP_8   \n",
       "2   custom_seq      9   59    EGFP_9   \n",
       "3   custom_seq     15   65   EGFP_15   \n",
       "4   custom_seq     17   67   EGFP_17   \n",
       "..         ...    ...  ...       ...   \n",
       "94  custom_seq    582  632  EGFP_582   \n",
       "95  custom_seq    588  638  EGFP_588   \n",
       "96  custom_seq    591  641  EGFP_591   \n",
       "97  custom_seq    593  643  EGFP_593   \n",
       "98  custom_seq    606  656  EGFP_606   \n",
       "\n",
       "                                       transcript_seq kmer_coord  \\\n",
       "0   TCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGT...     EGFP_7   \n",
       "1   CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...     EGFP_8   \n",
       "2   CTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGT...     EGFP_9   \n",
       "3   GAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCG...    EGFP_15   \n",
       "4   GCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAG...    EGFP_17   \n",
       "..                                                ...        ...   \n",
       "94  GCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGC...   EGFP_582   \n",
       "95  AGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGT...   EGFP_588   \n",
       "96  AAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCG...   EGFP_591   \n",
       "97  AGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTG...   EGFP_593   \n",
       "98  AAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGA...   EGFP_606   \n",
       "\n",
       "                                            probe_seq  GC_content_full  \\\n",
       "0   GGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACC...             0.62   \n",
       "1   CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...             0.64   \n",
       "2   CCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGA...             0.64   \n",
       "3   CCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCA...             0.66   \n",
       "4   CGCCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTC...             0.68   \n",
       "..                                                ...              ...   \n",
       "94  TCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTTTGCTCA...             0.60   \n",
       "95  ACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTT...             0.56   \n",
       "96  GTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGT...             0.56   \n",
       "97  CGGTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGG...             0.60   \n",
       "98  GTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCATGTGATCGC...             0.64   \n",
       "\n",
       "    longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                     3            0.56            0.68              True  \n",
       "1                     3            0.60            0.68              True  \n",
       "2                     3            0.64            0.64              True  \n",
       "3                     3            0.64            0.68              True  \n",
       "4                     3            0.68            0.68              True  \n",
       "..                  ...             ...             ...               ...  \n",
       "94                    4            0.60            0.60              True  \n",
       "95                    4            0.52            0.60              True  \n",
       "96                    4            0.56            0.56              True  \n",
       "97                    4            0.60            0.60              True  \n",
       "98                    3            0.68            0.60              True  \n",
       "\n",
       "[99 rows x 12 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Candidate kmers\n",
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "patient-period",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers.to_csv((out_dir + 'kmers_candidates_filtered.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "buried-courage",
   "metadata": {},
   "source": [
    "### 5. Remove probes with potential off-targets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "specific-sweet",
   "metadata": {},
   "source": [
    "Having identified a set of kmers that fulfill our sequence requirements, we can next proceed with testing whether they are specific to our transcript/exon of interest.\n",
    "\n",
    "For this we rely on using BLAST. \n",
    "\n",
    "We recommend blasting against transcripts (i.e., exons and introns combined) to be as conservative as possible in terms of off-targets. \n",
    "However, in cases where it is not possible to obtain enough suitable kmers (e.g., for  short transcripts), it is reasonable to relax this requirement by BLASTing against exons only (a much smaller search space).\n",
    "\n",
    "At this step, we also want to consider whether our probes are split (as in the current specifications for VisiumHD) or a single oligo. If probes are split, it's best to BLAST each side separately, to make sure that both sides are specific."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "developing-galaxy",
   "metadata": {},
   "outputs": [],
   "source": [
    "## The first thing to do is to export our sequences in fasta format, so that we can use them for BLAST\n",
    "write_fasta(kmers['name'], kmers['transcript_seq'], (out_dir + 'kmers_candidates_filtered_transcript_seqs.fa'))\n",
    "## If our probes are meant to be split, we should additionally blast them separately \n",
    "## Note that the LHS/RHS in the transcript are reversed compared to the probe (i.e., the LHS of the transcript is complementary to the RHS of the probe)\n",
    "if split_nt is not None: \n",
    "    ## Make split probes\n",
    "    kmers['transcript_seq_LHS'] = [seq[0:split_nt] for seq in kmers['transcript_seq']]\n",
    "    kmers['transcript_seq_RHS'] = [seq[split_nt: probe_length] for seq in kmers['transcript_seq']]\n",
    "\n",
    "    ## We are exporting the transcript sequence as that's the one that has to be blasted against the human transcriptome\n",
    "    write_fasta(kmers['name'], kmers['transcript_seq_LHS'], (out_dir + 'kmers_candidates_filtered_transcript_seqs_LHS.fa'))\n",
    "    write_fasta(kmers['name'], kmers['transcript_seq_RHS'], (out_dir + 'kmers_candidates_filtered_transcript_seqs_RHS.fa'))    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "toxic-cylinder",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_res = {}\n",
    "## First, blast the full probe\n",
    "blast_res['full'] = run_blast(fasta=(out_dir + 'kmers_candidates_filtered_transcript_seqs.fa'),\n",
    "                              blastdb = blast_db,\n",
    "                              path2blastn=(blast_exec_path + 'blastn'),\n",
    "                              outfile = (out_dir + 'kmers_candidates_filtered_blast_output.txt'))\n",
    "\n",
    "## Additionally, if probe is split, blast each side separately\n",
    "if split_nt is not None: \n",
    "    blast_res['LHS'] = run_blast(fasta=(out_dir + 'kmers_candidates_filtered_transcript_seqs_LHS.fa'),\n",
    "                                     blastdb = blast_db,\n",
    "                                     path2blastn=(blast_exec_path + 'blastn'),\n",
    "                                     outfile = (out_dir + 'kmers_candidates_filtered_blast_output_LHS.txt'))\n",
    "    blast_res['RHS'] = run_blast(fasta=(out_dir + 'kmers_candidates_filtered_transcript_seqs_RHS.fa'),\n",
    "                                     blastdb = blast_db,\n",
    "                                     path2blastn=(blast_exec_path + 'blastn'),\n",
    "                                     outfile = (out_dir + 'kmers_candidates_filtered_blast_output_RHS.txt'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "agreed-jonathan",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following genes were detected in mode: full\n",
      "sgeneid\n",
      "CCDC85C      4\n",
      "LINC00630    3\n",
      "Name: count, dtype: int64\n",
      "The following genes were detected in mode: LHS\n",
      "sgeneid\n",
      "LMTK2           2\n",
      "LHFPL6          2\n",
      "LOC105374785    1\n",
      "Name: count, dtype: int64\n",
      "The following genes were detected in mode: RHS\n",
      "sgeneid\n",
      "LOC105374785    1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for k in blast_res.keys():\n",
    "    print((\"The following genes were detected in mode: \" +  k))\n",
    "    print(blast_res[k]['sgeneid'].value_counts().head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "timely-fellowship",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>sgeneid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EGFP_42</td>\n",
       "      <td>CCDC85C::chr14:99500189-99604207(-)</td>\n",
       "      <td>100.000</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>36</td>\n",
       "      <td>211</td>\n",
       "      <td>231</td>\n",
       "      <td>0.47</td>\n",
       "      <td>39.2</td>\n",
       "      <td>CCDC85C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EGFP_42</td>\n",
       "      <td>CCDC85C::chr14:99524540-99604207(-)</td>\n",
       "      <td>100.000</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>36</td>\n",
       "      <td>211</td>\n",
       "      <td>231</td>\n",
       "      <td>0.47</td>\n",
       "      <td>39.2</td>\n",
       "      <td>CCDC85C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EGFP_42</td>\n",
       "      <td>CCDC85C::chr14:99524540-99604207(-)</td>\n",
       "      <td>100.000</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>36</td>\n",
       "      <td>211</td>\n",
       "      <td>231</td>\n",
       "      <td>0.47</td>\n",
       "      <td>39.2</td>\n",
       "      <td>CCDC85C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EGFP_42</td>\n",
       "      <td>CCDC85C::chr14:99526286-99604207(-)</td>\n",
       "      <td>100.000</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>36</td>\n",
       "      <td>211</td>\n",
       "      <td>231</td>\n",
       "      <td>0.47</td>\n",
       "      <td>39.2</td>\n",
       "      <td>CCDC85C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EGFP_68</td>\n",
       "      <td>LINC00630::chrX:102769152-102964523(+)</td>\n",
       "      <td>82.051</td>\n",
       "      <td>39</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>47</td>\n",
       "      <td>192909</td>\n",
       "      <td>192947</td>\n",
       "      <td>0.47</td>\n",
       "      <td>39.2</td>\n",
       "      <td>LINC00630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EGFP_74</td>\n",
       "      <td>LINC00630::chrX:102769152-102964523(+)</td>\n",
       "      <td>82.051</td>\n",
       "      <td>39</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>41</td>\n",
       "      <td>192909</td>\n",
       "      <td>192947</td>\n",
       "      <td>0.47</td>\n",
       "      <td>39.2</td>\n",
       "      <td>LINC00630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EGFP_75</td>\n",
       "      <td>LINC00630::chrX:102769152-102964523(+)</td>\n",
       "      <td>82.051</td>\n",
       "      <td>39</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>40</td>\n",
       "      <td>192909</td>\n",
       "      <td>192947</td>\n",
       "      <td>0.47</td>\n",
       "      <td>39.2</td>\n",
       "      <td>LINC00630</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      name                                  sseqid   pident  length  mismatch  \\\n",
       "0  EGFP_42     CCDC85C::chr14:99500189-99604207(-)  100.000      21         0   \n",
       "1  EGFP_42     CCDC85C::chr14:99524540-99604207(-)  100.000      21         0   \n",
       "2  EGFP_42     CCDC85C::chr14:99524540-99604207(-)  100.000      21         0   \n",
       "3  EGFP_42     CCDC85C::chr14:99526286-99604207(-)  100.000      21         0   \n",
       "4  EGFP_68  LINC00630::chrX:102769152-102964523(+)   82.051      39         7   \n",
       "5  EGFP_74  LINC00630::chrX:102769152-102964523(+)   82.051      39         7   \n",
       "6  EGFP_75  LINC00630::chrX:102769152-102964523(+)   82.051      39         7   \n",
       "\n",
       "   gapopen  qstart  qend  sstart    send  evalue  bitscore    sgeneid  \n",
       "0        0      16    36     211     231    0.47      39.2    CCDC85C  \n",
       "1        0      16    36     211     231    0.47      39.2    CCDC85C  \n",
       "2        0      16    36     211     231    0.47      39.2    CCDC85C  \n",
       "3        0      16    36     211     231    0.47      39.2    CCDC85C  \n",
       "4        0       9    47  192909  192947    0.47      39.2  LINC00630  \n",
       "5        0       3    41  192909  192947    0.47      39.2  LINC00630  \n",
       "6        0       2    40  192909  192947    0.47      39.2  LINC00630  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_res['full']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "offshore-great",
   "metadata": {},
   "source": [
    "Not all BLAST hits will be off-targets. Hopefully, our gene of interest is included in the BLAST output. We therefore need to filter for hits with different gene IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "aggressive-terrorist",
   "metadata": {},
   "outputs": [],
   "source": [
    "offtargets = []\n",
    "for k in blast_res.keys():\n",
    "    offtargets += (detect_offtargets(blast_res[k], gene_ID, min_mismatches=min_mismatches))\n",
    "## Remove redundancies\n",
    "offtargets = list(set(offtargets))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "continent-encyclopedia",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(offtargets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "novel-dispatch",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove off-targets\n",
    "kmers = kmers[kmers['name'].isin(offtargets)==False].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "alternative-wildlife",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "      <th>transcript_seq_LHS</th>\n",
       "      <th>transcript_seq_RHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>7</td>\n",
       "      <td>57</td>\n",
       "      <td>EGFP_7</td>\n",
       "      <td>TCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGT...</td>\n",
       "      <td>EGFP_7</td>\n",
       "      <td>GGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACC...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>3</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>TCCTGGTCGAGCTGGACGGCGACGT</td>\n",
       "      <td>AAACGGCCACAAGTTCAGCGTGTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>58</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>CCTGGTCGAGCTGGACGGCGACGTA</td>\n",
       "      <td>AACGGCCACAAGTTCAGCGTGTCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>9</td>\n",
       "      <td>59</td>\n",
       "      <td>EGFP_9</td>\n",
       "      <td>CTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGT...</td>\n",
       "      <td>EGFP_9</td>\n",
       "      <td>CCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGA...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "      <td>CTGGTCGAGCTGGACGGCGACGTAA</td>\n",
       "      <td>ACGGCCACAAGTTCAGCGTGTCCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>15</td>\n",
       "      <td>65</td>\n",
       "      <td>EGFP_15</td>\n",
       "      <td>GAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCG...</td>\n",
       "      <td>EGFP_15</td>\n",
       "      <td>CCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCA...</td>\n",
       "      <td>0.66</td>\n",
       "      <td>3</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>GAGCTGGACGGCGACGTAAACGGCC</td>\n",
       "      <td>ACAAGTTCAGCGTGTCCGGCGAGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>17</td>\n",
       "      <td>67</td>\n",
       "      <td>EGFP_17</td>\n",
       "      <td>GCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAG...</td>\n",
       "      <td>EGFP_17</td>\n",
       "      <td>CGCCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTC...</td>\n",
       "      <td>0.68</td>\n",
       "      <td>3</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>GCTGGACGGCGACGTAAACGGCCAC</td>\n",
       "      <td>AAGTTCAGCGTGTCCGGCGAGGGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>582</td>\n",
       "      <td>632</td>\n",
       "      <td>EGFP_582</td>\n",
       "      <td>GCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGC...</td>\n",
       "      <td>EGFP_582</td>\n",
       "      <td>TCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTTTGCTCA...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>GCCCTGAGCAAAGACCCCAACGAGA</td>\n",
       "      <td>AGCGCGATCACATGGTCCTGCTGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>588</td>\n",
       "      <td>638</td>\n",
       "      <td>EGFP_588</td>\n",
       "      <td>AGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGT...</td>\n",
       "      <td>EGFP_588</td>\n",
       "      <td>ACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTT...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>AGCAAAGACCCCAACGAGAAGCGCG</td>\n",
       "      <td>ATCACATGGTCCTGCTGGAGTTCGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>591</td>\n",
       "      <td>641</td>\n",
       "      <td>EGFP_591</td>\n",
       "      <td>AAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCG...</td>\n",
       "      <td>EGFP_591</td>\n",
       "      <td>GTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGT...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "      <td>AAAGACCCCAACGAGAAGCGCGATC</td>\n",
       "      <td>ACATGGTCCTGCTGGAGTTCGTGAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>593</td>\n",
       "      <td>643</td>\n",
       "      <td>EGFP_593</td>\n",
       "      <td>AGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTG...</td>\n",
       "      <td>EGFP_593</td>\n",
       "      <td>CGGTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGG...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>AGACCCCAACGAGAAGCGCGATCAC</td>\n",
       "      <td>ATGGTCCTGCTGGAGTTCGTGACCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>606</td>\n",
       "      <td>656</td>\n",
       "      <td>EGFP_606</td>\n",
       "      <td>AAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGA...</td>\n",
       "      <td>EGFP_606</td>\n",
       "      <td>GTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCATGTGATCGC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>AAGCGCGATCACATGGTCCTGCTGG</td>\n",
       "      <td>AGTTCGTGACCGCCGCCGGGATCAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname  start  end      name  \\\n",
       "0   custom_seq      7   57    EGFP_7   \n",
       "1   custom_seq      8   58    EGFP_8   \n",
       "2   custom_seq      9   59    EGFP_9   \n",
       "3   custom_seq     15   65   EGFP_15   \n",
       "4   custom_seq     17   67   EGFP_17   \n",
       "..         ...    ...  ...       ...   \n",
       "89  custom_seq    582  632  EGFP_582   \n",
       "90  custom_seq    588  638  EGFP_588   \n",
       "91  custom_seq    591  641  EGFP_591   \n",
       "92  custom_seq    593  643  EGFP_593   \n",
       "93  custom_seq    606  656  EGFP_606   \n",
       "\n",
       "                                       transcript_seq kmer_coord  \\\n",
       "0   TCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGT...     EGFP_7   \n",
       "1   CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...     EGFP_8   \n",
       "2   CTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGT...     EGFP_9   \n",
       "3   GAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCG...    EGFP_15   \n",
       "4   GCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAG...    EGFP_17   \n",
       "..                                                ...        ...   \n",
       "89  GCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGC...   EGFP_582   \n",
       "90  AGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGT...   EGFP_588   \n",
       "91  AAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCG...   EGFP_591   \n",
       "92  AGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTG...   EGFP_593   \n",
       "93  AAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGA...   EGFP_606   \n",
       "\n",
       "                                            probe_seq  GC_content_full  \\\n",
       "0   GGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGACC...             0.62   \n",
       "1   CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...             0.64   \n",
       "2   CCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGA...             0.64   \n",
       "3   CCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCA...             0.66   \n",
       "4   CGCCCTCGCCGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTC...             0.68   \n",
       "..                                                ...              ...   \n",
       "89  TCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTTTGCTCA...             0.60   \n",
       "90  ACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGTCTT...             0.56   \n",
       "91  GTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGGGT...             0.56   \n",
       "92  CGGTCACGAACTCCAGCAGGACCATGTGATCGCGCTTCTCGTTGGG...             0.60   \n",
       "93  GTGATCCCGGCGGCGGTCACGAACTCCAGCAGGACCATGTGATCGC...             0.64   \n",
       "\n",
       "    longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \\\n",
       "0                     3            0.56            0.68              True   \n",
       "1                     3            0.60            0.68              True   \n",
       "2                     3            0.64            0.64              True   \n",
       "3                     3            0.64            0.68              True   \n",
       "4                     3            0.68            0.68              True   \n",
       "..                  ...             ...             ...               ...   \n",
       "89                    4            0.60            0.60              True   \n",
       "90                    4            0.52            0.60              True   \n",
       "91                    4            0.56            0.56              True   \n",
       "92                    4            0.60            0.60              True   \n",
       "93                    3            0.68            0.60              True   \n",
       "\n",
       "           transcript_seq_LHS         transcript_seq_RHS  \n",
       "0   TCCTGGTCGAGCTGGACGGCGACGT  AAACGGCCACAAGTTCAGCGTGTCC  \n",
       "1   CCTGGTCGAGCTGGACGGCGACGTA  AACGGCCACAAGTTCAGCGTGTCCG  \n",
       "2   CTGGTCGAGCTGGACGGCGACGTAA  ACGGCCACAAGTTCAGCGTGTCCGG  \n",
       "3   GAGCTGGACGGCGACGTAAACGGCC  ACAAGTTCAGCGTGTCCGGCGAGGG  \n",
       "4   GCTGGACGGCGACGTAAACGGCCAC  AAGTTCAGCGTGTCCGGCGAGGGCG  \n",
       "..                        ...                        ...  \n",
       "89  GCCCTGAGCAAAGACCCCAACGAGA  AGCGCGATCACATGGTCCTGCTGGA  \n",
       "90  AGCAAAGACCCCAACGAGAAGCGCG  ATCACATGGTCCTGCTGGAGTTCGT  \n",
       "91  AAAGACCCCAACGAGAAGCGCGATC  ACATGGTCCTGCTGGAGTTCGTGAC  \n",
       "92  AGACCCCAACGAGAAGCGCGATCAC  ATGGTCCTGCTGGAGTTCGTGACCG  \n",
       "93  AAGCGCGATCACATGGTCCTGCTGG  AGTTCGTGACCGCCGCCGGGATCAC  \n",
       "\n",
       "[94 rows x 14 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dried-model",
   "metadata": {},
   "source": [
    "### 6. Select non-overlapping probes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accredited-holiday",
   "metadata": {},
   "source": [
    "At this point, we have effectively acquired a set of usable probes. They don't overlap undesirable regions (repeats/polymorphism), have desirable sequence features (GC content, specific nucleotides) and are specific to our gene of interest.\n",
    "\n",
    "In this particular case, we still have a lot of possible k-mers (much more than the number of probes we intend to design). We can therefore choose to prioritise k-mers with shorter homopolymer stretches, as these are also discouraged by the [10x recommendations](https://cdn.10xgenomics.com/image/upload/v1697739385/support-documents/CG000621_CustomProbeDesign_TechNote_RevC.pdf).\n",
    "\n",
    "However, at this stage you might want to consider ranking probes in a diffferent way, depending on your application.\n",
    "\n",
    "After having ranked our k-mers in whatever way we think is reasonable at this stage, we can proceed with selecting the top probe, then removing all overlapping/adjacent probes (within a window determined by `probe_offset`).\n",
    "\n",
    "Since we have so many available probes, we will be increasing `probe_offset` from `100 (default)` to `1000`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "virgin-apparatus",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sort in increasing homopolymer length\n",
    "kmers = kmers.sort_values('longest_homopolymer', ascending=True).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "laden-penguin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "      <th>transcript_seq_LHS</th>\n",
       "      <th>transcript_seq_RHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>450</td>\n",
       "      <td>500</td>\n",
       "      <td>EGFP_450</td>\n",
       "      <td>AAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGC...</td>\n",
       "      <td>EGFP_450</td>\n",
       "      <td>AGCTGCACGCTGCCGTCCTCGATGTTGTGGCGGATCTTGAAGTTCA...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>2</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>AAGGTGAACTTCAAGATCCGCCACA</td>\n",
       "      <td>ACATCGAGGACGGCAGCGTGCAGCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>452</td>\n",
       "      <td>502</td>\n",
       "      <td>EGFP_452</td>\n",
       "      <td>GGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAG...</td>\n",
       "      <td>EGFP_452</td>\n",
       "      <td>CGAGCTGCACGCTGCCGTCCTCGATGTTGTGGCGGATCTTGAAGTT...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>2</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "      <td>GGTGAACTTCAAGATCCGCCACAAC</td>\n",
       "      <td>ATCGAGGACGGCAGCGTGCAGCTCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>464</td>\n",
       "      <td>514</td>\n",
       "      <td>EGFP_464</td>\n",
       "      <td>GATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCAC...</td>\n",
       "      <td>EGFP_464</td>\n",
       "      <td>GGTAGTGGTCGGCGAGCTGCACGCTGCCGTCCTCGATGTTGTGGCG...</td>\n",
       "      <td>0.66</td>\n",
       "      <td>2</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "      <td>GATCCGCCACAACATCGAGGACGGC</td>\n",
       "      <td>AGCGTGCAGCTCGCCGACCACTACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>471</td>\n",
       "      <td>521</td>\n",
       "      <td>EGFP_471</td>\n",
       "      <td>CACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGC...</td>\n",
       "      <td>EGFP_471</td>\n",
       "      <td>TTCTGCTGGTAGTGGTCGGCGAGCTGCACGCTGCCGTCCTCGATGT...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>2</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "      <td>CACAACATCGAGGACGGCAGCGTGC</td>\n",
       "      <td>AGCTCGCCGACCACTACCAGCAGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>422</td>\n",
       "      <td>472</td>\n",
       "      <td>EGFP_422</td>\n",
       "      <td>CATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATC...</td>\n",
       "      <td>EGFP_422</td>\n",
       "      <td>GGCGGATCTTGAAGTTCACCTTGATGCCGTTCTTCTGCTTGTCGGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>CATGGCCGACAAGCAGAAGAACGGC</td>\n",
       "      <td>ATCAAGGTGAACTTCAAGATCCGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>186</td>\n",
       "      <td>236</td>\n",
       "      <td>EGFP_186</td>\n",
       "      <td>TACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGC...</td>\n",
       "      <td>EGFP_186</td>\n",
       "      <td>TCGGGCATGGCGGACTTGAAGAAGTCGTGCTGCTTCATGTGGTCGG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>TACCCCGACCACATGAAGCAGCACG</td>\n",
       "      <td>ACTTCTTCAAGTCCGCCATGCCCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>168</td>\n",
       "      <td>218</td>\n",
       "      <td>EGFP_168</td>\n",
       "      <td>GTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACT...</td>\n",
       "      <td>EGFP_168</td>\n",
       "      <td>AAGAAGTCGTGCTGCTTCATGTGGTCGGGGTAGCGGCTGAAGCACT...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>GTGCAGTGCTTCAGCCGCTACCCCG</td>\n",
       "      <td>ACCACATGAAGCAGCACGACTTCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>162</td>\n",
       "      <td>212</td>\n",
       "      <td>EGFP_162</td>\n",
       "      <td>TACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGC...</td>\n",
       "      <td>EGFP_162</td>\n",
       "      <td>TCGTGCTGCTTCATGTGGTCGGGGTAGCGGCTGAAGCACTGCACGC...</td>\n",
       "      <td>0.62</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "      <td>TACGGCGTGCAGTGCTTCAGCCGCT</td>\n",
       "      <td>ACCCCGACCACATGAAGCAGCACGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>155</td>\n",
       "      <td>205</td>\n",
       "      <td>EGFP_155</td>\n",
       "      <td>CCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATG...</td>\n",
       "      <td>EGFP_155</td>\n",
       "      <td>GCTTCATGTGGTCGGGGTAGCGGCTGAAGCACTGCACGCCGTAGGT...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>4</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "      <td>CCTGACCTACGGCGTGCAGTGCTTC</td>\n",
       "      <td>AGCCGCTACCCCGACCACATGAAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>489</td>\n",
       "      <td>539</td>\n",
       "      <td>EGFP_489</td>\n",
       "      <td>AGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCG...</td>\n",
       "      <td>EGFP_489</td>\n",
       "      <td>CCGTCGCCGATGGGGGTGTTCTGCTGGTAGTGGTCGGCGAGCTGCA...</td>\n",
       "      <td>0.68</td>\n",
       "      <td>5</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>AGCGTGCAGCTCGCCGACCACTACC</td>\n",
       "      <td>AGCAGAACACCCCCATCGGCGACGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname  start  end      name  \\\n",
       "0   custom_seq    450  500  EGFP_450   \n",
       "1   custom_seq    452  502  EGFP_452   \n",
       "2   custom_seq    464  514  EGFP_464   \n",
       "3   custom_seq    471  521  EGFP_471   \n",
       "4   custom_seq    422  472  EGFP_422   \n",
       "..         ...    ...  ...       ...   \n",
       "89  custom_seq    186  236  EGFP_186   \n",
       "90  custom_seq    168  218  EGFP_168   \n",
       "91  custom_seq    162  212  EGFP_162   \n",
       "92  custom_seq    155  205  EGFP_155   \n",
       "93  custom_seq    489  539  EGFP_489   \n",
       "\n",
       "                                       transcript_seq kmer_coord  \\\n",
       "0   AAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGC...   EGFP_450   \n",
       "1   GGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAG...   EGFP_452   \n",
       "2   GATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCAC...   EGFP_464   \n",
       "3   CACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGC...   EGFP_471   \n",
       "4   CATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATC...   EGFP_422   \n",
       "..                                                ...        ...   \n",
       "89  TACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGC...   EGFP_186   \n",
       "90  GTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACT...   EGFP_168   \n",
       "91  TACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGC...   EGFP_162   \n",
       "92  CCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATG...   EGFP_155   \n",
       "93  AGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCG...   EGFP_489   \n",
       "\n",
       "                                            probe_seq  GC_content_full  \\\n",
       "0   AGCTGCACGCTGCCGTCCTCGATGTTGTGGCGGATCTTGAAGTTCA...             0.56   \n",
       "1   CGAGCTGCACGCTGCCGTCCTCGATGTTGTGGCGGATCTTGAAGTT...             0.60   \n",
       "2   GGTAGTGGTCGGCGAGCTGCACGCTGCCGTCCTCGATGTTGTGGCG...             0.66   \n",
       "3   TTCTGCTGGTAGTGGTCGGCGAGCTGCACGCTGCCGTCCTCGATGT...             0.62   \n",
       "4   GGCGGATCTTGAAGTTCACCTTGATGCCGTTCTTCTGCTTGTCGGC...             0.54   \n",
       "..                                                ...              ...   \n",
       "89  TCGGGCATGGCGGACTTGAAGAAGTCGTGCTGCTTCATGTGGTCGG...             0.58   \n",
       "90  AAGAAGTCGTGCTGCTTCATGTGGTCGGGGTAGCGGCTGAAGCACT...             0.58   \n",
       "91  TCGTGCTGCTTCATGTGGTCGGGGTAGCGGCTGAAGCACTGCACGC...             0.62   \n",
       "92  GCTTCATGTGGTCGGGGTAGCGGCTGAAGCACTGCACGCCGTAGGT...             0.64   \n",
       "93  CCGTCGCCGATGGGGGTGTTCTGCTGGTAGTGGTCGGCGAGCTGCA...             0.68   \n",
       "\n",
       "    longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \\\n",
       "0                     2            0.64            0.48              True   \n",
       "1                     2            0.68            0.52              True   \n",
       "2                     2            0.68            0.64              True   \n",
       "3                     2            0.60            0.64              True   \n",
       "4                     2            0.48            0.60              True   \n",
       "..                  ...             ...             ...               ...   \n",
       "89                    4            0.56            0.60              True   \n",
       "90                    4            0.48            0.68              True   \n",
       "91                    4            0.60            0.64              True   \n",
       "92                    4            0.64            0.64              True   \n",
       "93                    5            0.68            0.68              True   \n",
       "\n",
       "           transcript_seq_LHS         transcript_seq_RHS  \n",
       "0   AAGGTGAACTTCAAGATCCGCCACA  ACATCGAGGACGGCAGCGTGCAGCT  \n",
       "1   GGTGAACTTCAAGATCCGCCACAAC  ATCGAGGACGGCAGCGTGCAGCTCG  \n",
       "2   GATCCGCCACAACATCGAGGACGGC  AGCGTGCAGCTCGCCGACCACTACC  \n",
       "3   CACAACATCGAGGACGGCAGCGTGC  AGCTCGCCGACCACTACCAGCAGAA  \n",
       "4   CATGGCCGACAAGCAGAAGAACGGC  ATCAAGGTGAACTTCAAGATCCGCC  \n",
       "..                        ...                        ...  \n",
       "89  TACCCCGACCACATGAAGCAGCACG  ACTTCTTCAAGTCCGCCATGCCCGA  \n",
       "90  GTGCAGTGCTTCAGCCGCTACCCCG  ACCACATGAAGCAGCACGACTTCTT  \n",
       "91  TACGGCGTGCAGTGCTTCAGCCGCT  ACCCCGACCACATGAAGCAGCACGA  \n",
       "92  CCTGACCTACGGCGTGCAGTGCTTC  AGCCGCTACCCCGACCACATGAAGC  \n",
       "93  AGCGTGCAGCTCGCCGACCACTACC  AGCAGAACACCCCCATCGGCGACGG  \n",
       "\n",
       "[94 rows x 14 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "italian-packing",
   "metadata": {},
   "outputs": [],
   "source": [
    "## We have a lot of probes here - increasing the offset to 1000 bp to space them out\n",
    "probe_offset = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "closing-control",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nfs/team205/is10/miniconda/envs/gene2probe_env/lib/python3.11/site-packages/gene2probe/gene2probe.py:436: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  start = int(probe_df['start'].iloc[i])\n",
      "/nfs/team205/is10/miniconda/envs/gene2probe_env/lib/python3.11/site-packages/gene2probe/gene2probe.py:437: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  end = int(probe_df['end'].iloc[i])\n"
     ]
    }
   ],
   "source": [
    "selected_probes_list = []\n",
    "df = kmers.copy()\n",
    "i = 0\n",
    "while i < n_desired_probes:\n",
    "    selected_probes_list.append(df.iloc[0:1,:])\n",
    "    ## Going for a high offset as we have a large span:\n",
    "    df = remove_overlapping_probes(df,\n",
    "                                   df['name'][0], \n",
    "                                   offset=probe_offset).reset_index(drop=True).copy()\n",
    "    i +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "julian-sunday",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_probes_df = pd.concat(selected_probes_list, axis=0).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "sound-modem",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "      <th>transcript_seq_LHS</th>\n",
       "      <th>transcript_seq_RHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>450</td>\n",
       "      <td>500</td>\n",
       "      <td>EGFP_450</td>\n",
       "      <td>AAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGC...</td>\n",
       "      <td>EGFP_450</td>\n",
       "      <td>AGCTGCACGCTGCCGTCCTCGATGTTGTGGCGGATCTTGAAGTTCA...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>2</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>AAGGTGAACTTCAAGATCCGCCACA</td>\n",
       "      <td>ACATCGAGGACGGCAGCGTGCAGCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>233</td>\n",
       "      <td>283</td>\n",
       "      <td>EGFP_233</td>\n",
       "      <td>CGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGC...</td>\n",
       "      <td>EGFP_233</td>\n",
       "      <td>AGTTGCCGTCGTCCTTGAAGAAGATGGTGCGCTCCTGGACGTAGCC...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>CGAAGGCTACGTCCAGGAGCGCACC</td>\n",
       "      <td>ATCTTCTTCAAGGACGACGGCAACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>58</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>CCTGGTCGAGCTGGACGGCGACGTA</td>\n",
       "      <td>AACGGCCACAAGTTCAGCGTGTCCG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname  start  end      name  \\\n",
       "0  custom_seq    450  500  EGFP_450   \n",
       "1  custom_seq    233  283  EGFP_233   \n",
       "2  custom_seq      8   58    EGFP_8   \n",
       "\n",
       "                                      transcript_seq kmer_coord  \\\n",
       "0  AAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGC...   EGFP_450   \n",
       "1  CGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGC...   EGFP_233   \n",
       "2  CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...     EGFP_8   \n",
       "\n",
       "                                           probe_seq  GC_content_full  \\\n",
       "0  AGCTGCACGCTGCCGTCCTCGATGTTGTGGCGGATCTTGAAGTTCA...             0.56   \n",
       "1  AGTTGCCGTCGTCCTTGAAGAAGATGGTGCGCTCCTGGACGTAGCC...             0.58   \n",
       "2  CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...             0.64   \n",
       "\n",
       "   longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \\\n",
       "0                    2            0.64            0.48              True   \n",
       "1                    2            0.48            0.68              True   \n",
       "2                    3            0.60            0.68              True   \n",
       "\n",
       "          transcript_seq_LHS         transcript_seq_RHS  \n",
       "0  AAGGTGAACTTCAAGATCCGCCACA  ACATCGAGGACGGCAGCGTGCAGCT  \n",
       "1  CGAAGGCTACGTCCAGGAGCGCACC  ATCTTCTTCAAGGACGACGGCAACT  \n",
       "2  CCTGGTCGAGCTGGACGGCGACGTA  AACGGCCACAAGTTCAGCGTGTCCG  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_probes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "binary-mistake",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCT\n",
      "CGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACT\n",
      "CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCG\n"
     ]
    }
   ],
   "source": [
    "for seq in selected_probes_df['transcript_seq']:\n",
    "    print(seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "outside-treatment",
   "metadata": {},
   "source": [
    "And we are done! We have now selected three potential probes for our gene.\n",
    "\n",
    "If our probes are meant to be split, we can additionally generate these columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "planned-messaging",
   "metadata": {},
   "outputs": [],
   "source": [
    "if split_nt is not None: \n",
    "    ## Make split probes (and add adapters if provided)\n",
    "    selected_probes_df['probe_seq_LHS'] = [(LHS_pref + seq[0:split_nt] + LHS_suff) for seq in selected_probes_df['probe_seq']]\n",
    "    selected_probes_df['probe_seq_RHS'] = [(RHS_pref +seq[split_nt: probe_length] + RHS_suff) for seq in selected_probes_df['probe_seq']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "molecular-brook",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Also add gene_ID for completeness\n",
    "selected_probes_df['gene_ID'] = gene_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "finished-recruitment",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export selected probes as dataframe:\n",
    "selected_probes_df.to_csv((out_dir + 'kmers_selected_probes.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "broadband-spice",
   "metadata": {},
   "source": [
    "We always recommend additionally performing a manual BLAST of these probe sequences to make sure that there are no off-target effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ordinary-cooperation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "      <th>transcript_seq_LHS</th>\n",
       "      <th>transcript_seq_RHS</th>\n",
       "      <th>probe_seq_LHS</th>\n",
       "      <th>probe_seq_RHS</th>\n",
       "      <th>gene_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>450</td>\n",
       "      <td>500</td>\n",
       "      <td>EGFP_450</td>\n",
       "      <td>AAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGC...</td>\n",
       "      <td>EGFP_450</td>\n",
       "      <td>AGCTGCACGCTGCCGTCCTCGATGTTGTGGCGGATCTTGAAGTTCA...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>2</td>\n",
       "      <td>0.64</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>AAGGTGAACTTCAAGATCCGCCACA</td>\n",
       "      <td>ACATCGAGGACGGCAGCGTGCAGCT</td>\n",
       "      <td>CCTTGGCACCCGAGAATTCCAAGCTGCACGCTGCCGTCCTCGATGT</td>\n",
       "      <td>/5Phos/TGTGGCGGATCTTGAAGTTCACCTTAAAAAAAAAAAAAA...</td>\n",
       "      <td>EGFP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>233</td>\n",
       "      <td>283</td>\n",
       "      <td>EGFP_233</td>\n",
       "      <td>CGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGC...</td>\n",
       "      <td>EGFP_233</td>\n",
       "      <td>AGTTGCCGTCGTCCTTGAAGAAGATGGTGCGCTCCTGGACGTAGCC...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>CGAAGGCTACGTCCAGGAGCGCACC</td>\n",
       "      <td>ATCTTCTTCAAGGACGACGGCAACT</td>\n",
       "      <td>CCTTGGCACCCGAGAATTCCAAGTTGCCGTCGTCCTTGAAGAAGAT</td>\n",
       "      <td>/5Phos/GGTGCGCTCCTGGACGTAGCCTTCGAAAAAAAAAAAAAA...</td>\n",
       "      <td>EGFP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>custom_seq</td>\n",
       "      <td>8</td>\n",
       "      <td>58</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...</td>\n",
       "      <td>EGFP_8</td>\n",
       "      <td>CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...</td>\n",
       "      <td>0.64</td>\n",
       "      <td>3</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.68</td>\n",
       "      <td>True</td>\n",
       "      <td>CCTGGTCGAGCTGGACGGCGACGTA</td>\n",
       "      <td>AACGGCCACAAGTTCAGCGTGTCCG</td>\n",
       "      <td>CCTTGGCACCCGAGAATTCCACGGACACGCTGAACTTGTGGCCGTT</td>\n",
       "      <td>/5Phos/TACGTCGCCGTCCAGCTCGACCAGGAAAAAAAAAAAAAA...</td>\n",
       "      <td>EGFP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname  start  end      name  \\\n",
       "0  custom_seq    450  500  EGFP_450   \n",
       "1  custom_seq    233  283  EGFP_233   \n",
       "2  custom_seq      8   58    EGFP_8   \n",
       "\n",
       "                                      transcript_seq kmer_coord  \\\n",
       "0  AAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGC...   EGFP_450   \n",
       "1  CGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGC...   EGFP_233   \n",
       "2  CCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTG...     EGFP_8   \n",
       "\n",
       "                                           probe_seq  GC_content_full  \\\n",
       "0  AGCTGCACGCTGCCGTCCTCGATGTTGTGGCGGATCTTGAAGTTCA...             0.56   \n",
       "1  AGTTGCCGTCGTCCTTGAAGAAGATGGTGCGCTCCTGGACGTAGCC...             0.58   \n",
       "2  CGGACACGCTGAACTTGTGGCCGTTTACGTCGCCGTCCAGCTCGAC...             0.64   \n",
       "\n",
       "   longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \\\n",
       "0                    2            0.64            0.48              True   \n",
       "1                    2            0.48            0.68              True   \n",
       "2                    3            0.60            0.68              True   \n",
       "\n",
       "          transcript_seq_LHS         transcript_seq_RHS  \\\n",
       "0  AAGGTGAACTTCAAGATCCGCCACA  ACATCGAGGACGGCAGCGTGCAGCT   \n",
       "1  CGAAGGCTACGTCCAGGAGCGCACC  ATCTTCTTCAAGGACGACGGCAACT   \n",
       "2  CCTGGTCGAGCTGGACGGCGACGTA  AACGGCCACAAGTTCAGCGTGTCCG   \n",
       "\n",
       "                                    probe_seq_LHS  \\\n",
       "0  CCTTGGCACCCGAGAATTCCAAGCTGCACGCTGCCGTCCTCGATGT   \n",
       "1  CCTTGGCACCCGAGAATTCCAAGTTGCCGTCGTCCTTGAAGAAGAT   \n",
       "2  CCTTGGCACCCGAGAATTCCACGGACACGCTGAACTTGTGGCCGTT   \n",
       "\n",
       "                                       probe_seq_RHS gene_ID  \n",
       "0  /5Phos/TGTGGCGGATCTTGAAGTTCACCTTAAAAAAAAAAAAAA...    EGFP  \n",
       "1  /5Phos/GGTGCGCTCCTGGACGTAGCCTTCGAAAAAAAAAAAAAA...    EGFP  \n",
       "2  /5Phos/TACGTCGCCGTCCAGCTCGACCAGGAAAAAAAAAAAAAA...    EGFP  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_probes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "strong-dinner",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gene2probe_env",
   "language": "python",
   "name": "gene2probe_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
