{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e216d8aa-ec15-4d54-a8ff-221efdb48c7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import sys\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pybedtools\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import gc_fraction\n",
    "from gene2probe import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f708aff-11ce-4ac7-8e30-952a5902419e",
   "metadata": {},
   "source": [
    "In the previous tutorials, we have seen how to generate probes starting from a [gene](https://github.com/Teichlab/gene2probe/blob/main/notebooks/002_make_exon_probes_XIST.ipynb) or a [specific region](https://github.com/Teichlab/gene2probe/blob/main/notebooks/003_make_probes_for_custom_region_CD45RA.ipynb).\n",
    "\n",
    "But what if we are interested in designing probes against a sequence that is not included in the human genome (e.g., GFP)?\n",
    "We can achieve that by directly providing the sequence as input and following all subsequent steps.\n",
    "\n",
    "We note that we don't recommend this approach for the detection of bacterial or viral reads for metagenomics analyses, as there are many additional considerations that have to be taken into account to ensure that a probe is specifically detecting a certain species."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "865d68f2-a5f1-45f2-9996-f3cf19967e33",
   "metadata": {},
   "source": [
    "### 1. Specify parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "26aa4295-dc32-413a-92c9-71c01dad41e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Specify gene of interest and feature of interest\n",
    "gene_ID = 'EGFP'\n",
    "sequence = 'gtgcccatcctggtcgagctggacggcgacgtaaacggccacaagttcagcgtgtccggcgagggcgagggcgatgccacctacggcaagctgaccctgaagttcatctgcaccaccggcaagctgcccgtgccctggcccaccctcgtgaccaccctgacctacggcgtgcagtgcttcagccgctaccccgaccacatgaagcagcacgacttcttcaagtccgccatgcccgaaggctacgtccaggagcgcaccatcttcttcaaggacgacggcaactacaagacccgcgccgaggtgaagttcgagggcgacaccctggtgaaccgcatcgagctgaagggcatcgacttcaaggaggacggcaacatcctggggcacaagctggagtacaactacaacagccacaacgtctatatcatggccgacaagcagaagaacggcatcaaggtgaacttcaagatccgccacaacatcgaggacggcagcgtgcagctcgccgaccactaccagcagaacacccccatcggcgacggccccgtgctgctgcccgacaaccactacctgagcacccagtccgccctgagcaaagaccccaacgagaagcgcgatcacatggtcctgctggagttcgtgaccgccgccgggatcactctcggcatggacgagctgtacaagtaa'\n",
    "\n",
    "## specify output directory\n",
    "out_dir = '../sample_run/probeDesign_' + gene_ID +'/'\n",
    "## Create output directory\n",
    "os.makedirs(out_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17f89fba-8fd7-4a21-9f27-67060bd21a44",
   "metadata": {},
   "source": [
    "Additionally, we need to provide the path to several resource files. Many of these files can be obtained from [UCSC table browser](https://genome.ucsc.edu/cgi-bin/hgTables).\n",
    "\n",
    "We also need a blast database, such as the one we generated in the [previous tutorial](https://github.com/Teichlab/gene2probe/blob/main/notebooks/001_make_blast_database.ipynb)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1dffb1b9-9ecc-490d-bea4-7aefb5a26028",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Required resources (most can be downloaded from \n",
    "gtf = '../hg38_resources/hg38.ncbiRefSeq.gtf' ## Gene annotation in gtf file\n",
    "## We recommend using RefSeq as this is manually curated and more likely to contain an isoform that is present across most cell types\n",
    "## Alternatively, one can filter based on RNA-seq data for a cell type/tissue of interest\n",
    "fasta = '../hg38_resources/hg38.fa' ## Genome in fasta file\n",
    "snp_db = '../hg38_resources/hg38_snp151Common.bed' ## Database of known SNPs and small indels\n",
    "repeats = '../hg38_resources/hg38_rmsk.bed' ## bed file with repeats/low complexity regions to be excluded\n",
    "gaps = '../hg38_resources/hg38_rmsk.bed' ## bed file with gaps in the genome assembly to be excluded\n",
    "blast_db = '../hg38_resources/blastdb/hg38_ncbiRefSeq_transcripts_db' ## Database of all human transcripts to blast against"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a65bb4d6-13fd-4d49-9a1b-496c7d0e6a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Path to blast binaries.\n",
    "## Replace with your conda environment\n",
    "## This can also be omitted if you started the jupyter session from within the gene2probe conda environment\n",
    "blast_exec_path = '/nfs/team205/is10/miniconda/envs/gene2probe_env/bin/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa752587-9f96-4ca4-8d44-fae0021f0cf2",
   "metadata": {},
   "source": [
    "Finally, we need to provide a set of parameters related to our probe's length, at which nucleotide it's split (if at all), the acceptable range for GC content and any specific requirements for individual nucleotides.\n",
    "\n",
    "Here we are following the [recommendations of 10x Genomics for custom probes for VisiumHD/VisiumFFPE/Flex](https://cdn.10xgenomics.com/image/upload/v1697739385/support-documents/CG000621_CustomProbeDesign_TechNote_RevC.pdf)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cba53a9d-2f6d-4611-8b97-4a6ce8ac75e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Additional parameters regarding how the probe should look like\n",
    "probe_length = 50 ## Length of probe in nucleotides\n",
    "split_nt = 25 ## Index of nucleotide to split the probe at (start of RHS) - set to None if splitting probe is not needed\n",
    "min_GC = 0.44 ## Minimum GC content for probe (if split probe, applied to both LHS and RHS)\n",
    "max_GC = 0.72 ## Maximum GC content for probe (if split probe, applied to both LHS and RHS)\n",
    "required_nts = {24: 'T'} ## Dictionary of index (0-based) for required nts - by default, 25th nucleotide must be a T - set to None if no requirements\n",
    "probe_offset = 100 ## Minimum distance between probes - 10 bp is the recommended minimum by 10x, this can also be adjusted depending on how many probes pass other cutoffs\n",
    "n_desired_probes =3 ## Number of probes to be designed.\n",
    "min_mismatches = 5 ## Minimum number of mismatches (in at least LHS or RHS) - here we require in both to be more conservative"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6ddda454-c3e3-4a6c-aa4b-204e07d1d622",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Optionally, we can also specify adapters that have to be added to the probes.\n",
    "## For example, for visiumHD:\n",
    "LHS_pref = 'CCTTGGCACCCGAGAATTCCA' ## Will be added to the 5' of the LHS probe\n",
    "LHS_suff = '' ## Will be added to the 3' of the LHS probe\n",
    "RHS_pref = '/5Phos/' ## Will be added to the 5' of the RHS probe\n",
    "RHS_suff = 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA' ## Will be added to the 3' of the RHS probe\n",
    "\n",
    "## Leave as empty strings if you don't want to use them"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba5cc208-c3d8-42d8-9cb2-67c47aead9c5",
   "metadata": {},
   "source": [
    "### 2. Generate k-mers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0392b3e9-f101-4386-a981-1338c431ec02",
   "metadata": {},
   "source": [
    "Now we can start by reading the gene annotation and filtering for our gene of interest."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4bb522db-fc21-421c-b37e-8154a58ae63f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Read gtf file\n",
    "gene_anno = read_gtf(gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "049129b1-a807-491e-a725-0a8d9f7c1113",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"TRNP\"; transcript_id \"rna-TRNP\";  gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"TRNP\"; transcript_id \"rna-TRNP\"; exon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"TRNT\"; transcript_id \"rna-TRNT\";  gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"TRNT\"; transcript_id \"rna-TRNT\"; exon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"CYTB\"; transcript_id \"rna-CYTB\";  gen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886697</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>29321</td>\n",
       "      <td>29370</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"WASH7P\"; transcript_id \"NR_024540.1\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886698</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886699</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>11874</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886700</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4886701</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq.2022-10-28</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4886702 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname                 source     feature  start    end score strand  \\\n",
       "0          chrM  ncbiRefSeq.2022-10-28  transcript  15956  16023     .      -   \n",
       "1          chrM  ncbiRefSeq.2022-10-28        exon  15956  16023     .      -   \n",
       "2          chrM  ncbiRefSeq.2022-10-28  transcript  15888  15953     .      +   \n",
       "3          chrM  ncbiRefSeq.2022-10-28        exon  15888  15953     .      +   \n",
       "4          chrM  ncbiRefSeq.2022-10-28  transcript  14747  15887     .      +   \n",
       "...         ...                    ...         ...    ...    ...   ...    ...   \n",
       "4886697    chr1  ncbiRefSeq.2022-10-28        exon  29321  29370     .      -   \n",
       "4886698    chr1  ncbiRefSeq.2022-10-28  transcript  11874  14409     .      +   \n",
       "4886699    chr1  ncbiRefSeq.2022-10-28        exon  11874  12227     .      +   \n",
       "4886700    chr1  ncbiRefSeq.2022-10-28        exon  12613  12721     .      +   \n",
       "4886701    chr1  ncbiRefSeq.2022-10-28        exon  13221  14409     .      +   \n",
       "\n",
       "        frame                                          attribute  \n",
       "0           .  gene_id \"TRNP\"; transcript_id \"rna-TRNP\";  gen...  \n",
       "1           .  gene_id \"TRNP\"; transcript_id \"rna-TRNP\"; exon...  \n",
       "2           .  gene_id \"TRNT\"; transcript_id \"rna-TRNT\";  gen...  \n",
       "3           .  gene_id \"TRNT\"; transcript_id \"rna-TRNT\"; exon...  \n",
       "4           .  gene_id \"CYTB\"; transcript_id \"rna-CYTB\";  gen...  \n",
       "...       ...                                                ...  \n",
       "4886697     .  gene_id \"WASH7P\"; transcript_id \"NR_024540.1\";...  \n",
       "4886698     .  gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...  \n",
       "4886699     .  gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...  \n",
       "4886700     .  gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...  \n",
       "4886701     .  gene_id \"DDX11L1\"; transcript_id \"NR_046018.2\"...  \n",
       "\n",
       "[4886702 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b510a00e-bf9f-4476-85f8-8d643859dc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Extract regions corresponding to gene of interest (symbol: gene_name, Ensembl ID: gene_ID), subset to feature of interest and convert to bed style dataframe:\n",
    "roi_bed = get_region_of_interest(gene_anno, gene_ID, gene_id_type = 'gene_name', feature=mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "60894d08-da20-462e-beb6-b2e672db60f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73827984</td>\n",
       "      <td>XIST_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73829067</td>\n",
       "      <td>73829231</td>\n",
       "      <td>XIST_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73831065</td>\n",
       "      <td>73831274</td>\n",
       "      <td>XIST_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73833237</td>\n",
       "      <td>73833374</td>\n",
       "      <td>XIST_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73837439</td>\n",
       "      <td>73837503</td>\n",
       "      <td>XIST_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73841381</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     start       end    name score strand\n",
       "0    chrX  73820650  73827984  XIST_0     .      -\n",
       "1    chrX  73829067  73829231  XIST_1     .      -\n",
       "2    chrX  73831065  73831274  XIST_2     .      -\n",
       "3    chrX  73833237  73833374  XIST_3     .      -\n",
       "4    chrX  73837439  73837503  XIST_4     .      -\n",
       "5    chrX  73841381  73852753  XIST_5     .      -"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roi_bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "019cdc9b-d3fa-4d46-91f7-5308681f6e6c",
   "metadata": {},
   "source": [
    "After extracting the coordinates of interest and converting to a bed-like format, we can generate all possible kmers that fall within these regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "60f606b4-d196-4382-bcae-f64364dc5768",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers = generate_kmers(roi_bed, k=probe_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c981be58-decf-4f36-bce9-1efabaa02627",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18981</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18982</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18983</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852701</td>\n",
       "      <td>73852751</td>\n",
       "      <td>XIST_5_11320</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18984</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852702</td>\n",
       "      <td>73852752</td>\n",
       "      <td>XIST_5_11321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18985</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852703</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5_11322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18986 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -\n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -\n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -\n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -\n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -\n",
       "...       ...       ...       ...           ...   ...    ...\n",
       "18981    chrX  73852699  73852749  XIST_5_11318     .      -\n",
       "18982    chrX  73852700  73852750  XIST_5_11319     .      -\n",
       "18983    chrX  73852701  73852751  XIST_5_11320     .      -\n",
       "18984    chrX  73852702  73852752  XIST_5_11321     .      -\n",
       "18985    chrX  73852703  73852753  XIST_5_11322     .      -\n",
       "\n",
       "[18986 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "69c58199-798b-45c7-8799-bcd30e8dfb3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export unfiltered\n",
    "kmers.to_csv((out_dir + 'kmers_all.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae962a2c-349a-411f-a22b-5434990e91ea",
   "metadata": {},
   "source": [
    "### 3. Exclude annotated repeats/polymorphism"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8715663-c967-47e5-be22-1094b2721689",
   "metadata": {},
   "source": [
    "We can next exclude kmers overlapping undesired regions (repeats, low complexity regions, common polymorphism, gaps in the assembly) from further consideration.\n",
    "\n",
    "Ideally, we will exclude everything that overlaps a repeat or polymorphism, but if we only have too few kmers available, we might need to relax these requirements (e.g., to only exclude kmers overlapping SNPs around the ligation junction, if the probes are split)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b230cbd8-3b8f-477d-adfb-560857aa7887",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17166</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17167</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17168</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852701</td>\n",
       "      <td>73852751</td>\n",
       "      <td>XIST_5_11320</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17169</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852702</td>\n",
       "      <td>73852752</td>\n",
       "      <td>XIST_5_11321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17170</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852703</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5_11322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17171 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -\n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -\n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -\n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -\n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -\n",
       "...       ...       ...       ...           ...   ...    ...\n",
       "17166    chrX  73852699  73852749  XIST_5_11318     .      -\n",
       "17167    chrX  73852700  73852750  XIST_5_11319     .      -\n",
       "17168    chrX  73852701  73852751  XIST_5_11320     .      -\n",
       "17169    chrX  73852702  73852752  XIST_5_11321     .      -\n",
       "17170    chrX  73852703  73852753  XIST_5_11322     .      -\n",
       "\n",
       "[17171 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## For example, we could have removed all kmers overlapping a repeat/low complexity region within 5 nts of the ligation junction:\n",
    "remove_overlaps(kmers, repeats, core=[20,30])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "70e19fb2-25cf-4285-8f1a-596ffd9b807b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## In this case we have a lot of possible kmers, so we will remove those with overlaps in any part of the probe:\n",
    "kmers = remove_overlaps(kmers, repeats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "95b1779e-46ae-435b-bd9d-d5108cb6fa01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16853</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16854</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16855</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852701</td>\n",
       "      <td>73852751</td>\n",
       "      <td>XIST_5_11320</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16856</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852702</td>\n",
       "      <td>73852752</td>\n",
       "      <td>XIST_5_11321</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16857</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852703</td>\n",
       "      <td>73852753</td>\n",
       "      <td>XIST_5_11322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16858 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -\n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -\n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -\n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -\n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -\n",
       "...       ...       ...       ...           ...   ...    ...\n",
       "16853    chrX  73852699  73852749  XIST_5_11318     .      -\n",
       "16854    chrX  73852700  73852750  XIST_5_11319     .      -\n",
       "16855    chrX  73852701  73852751  XIST_5_11320     .      -\n",
       "16856    chrX  73852702  73852752  XIST_5_11321     .      -\n",
       "16857    chrX  73852703  73852753  XIST_5_11322     .      -\n",
       "\n",
       "[16858 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c0418e87-d0bc-403f-9cb1-72247aa30d28",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Doing the same for gaps in the assembly (very unlikely since we are starting with annotated exons)\n",
    "kmers = remove_overlaps(kmers, gaps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c9ae438c-62c7-4d60-8901-776f9c09e952",
   "metadata": {},
   "outputs": [],
   "source": [
    "## And more importantly, against common polymorphism (SNPs, short indels)\n",
    "kmers = remove_overlaps(kmers, snp_db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "969e23e8-4a03-44a8-a25a-110c66ed06f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15711</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852696</td>\n",
       "      <td>73852746</td>\n",
       "      <td>XIST_5_11315</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15712</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852697</td>\n",
       "      <td>73852747</td>\n",
       "      <td>XIST_5_11316</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15713</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852698</td>\n",
       "      <td>73852748</td>\n",
       "      <td>XIST_5_11317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15714</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15715</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15716 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -\n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -\n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -\n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -\n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -\n",
       "...       ...       ...       ...           ...   ...    ...\n",
       "15711    chrX  73852696  73852746  XIST_5_11315     .      -\n",
       "15712    chrX  73852697  73852747  XIST_5_11316     .      -\n",
       "15713    chrX  73852698  73852748  XIST_5_11317     .      -\n",
       "15714    chrX  73852699  73852749  XIST_5_11318     .      -\n",
       "15715    chrX  73852700  73852750  XIST_5_11319     .      -\n",
       "\n",
       "[15716 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5170568-ccfc-4e68-8738-e0916c4e5796",
   "metadata": {},
   "source": [
    "### 4. Filter for desirable sequence features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "090e41ad-d72b-4b91-bbb8-f6c259530c96",
   "metadata": {},
   "source": [
    "Having excluded undesirable kmers based on intersection with genomic annotations, the next step is to consider their sequence features.\n",
    "\n",
    "For this, we first extract the sequences of each k-mer, and then estimate features such as GC content and the presence of desired nucleotides in specific positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6b91e694-c124-483a-941e-5b134aa3e69e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Get DNA for the transcript\n",
    "kmers_bed = pybedtools.BedTool.from_dataframe(kmers)\n",
    "kmers_seq = kmers_bed.sequence(fi=fasta, s=True) \n",
    "\n",
    "## We can read in the sequences and simultaneously monitor GC content and count the longest homopolymer stretch\n",
    "kmers_seq_stats = get_sequence_stats(kmers_seq.seqfn, probe_length, split_nt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6db0119a-9c26-47f8-bb25-363d738fede9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Combining with our dataframe\n",
    "kmers = pd.merge(kmers, kmers_seq_stats, left_index=True, right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9a98f6eb-dcdd-4406-9c7c-8e3dd37e18e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820650-73820700(-)</td>\n",
       "      <td>GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...</td>\n",
       "      <td>CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820651</td>\n",
       "      <td>73820701</td>\n",
       "      <td>XIST_0_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820651-73820701(-)</td>\n",
       "      <td>TGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAA...</td>\n",
       "      <td>AAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTC...</td>\n",
       "      <td>0.28</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820652-73820702(-)</td>\n",
       "      <td>CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...</td>\n",
       "      <td>AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820653-73820703(-)</td>\n",
       "      <td>ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...</td>\n",
       "      <td>AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820654-73820704(-)</td>\n",
       "      <td>AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...</td>\n",
       "      <td>GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15711</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852696</td>\n",
       "      <td>73852746</td>\n",
       "      <td>XIST_5_11315</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852696-73852746(-)</td>\n",
       "      <td>TTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCT...</td>\n",
       "      <td>CGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15712</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852697</td>\n",
       "      <td>73852747</td>\n",
       "      <td>XIST_5_11316</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852697-73852747(-)</td>\n",
       "      <td>GTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTC...</td>\n",
       "      <td>GAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15713</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852698</td>\n",
       "      <td>73852748</td>\n",
       "      <td>XIST_5_11317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852698-73852748(-)</td>\n",
       "      <td>AGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACT...</td>\n",
       "      <td>AGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAG...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15714</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852699</td>\n",
       "      <td>73852749</td>\n",
       "      <td>XIST_5_11318</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852699-73852749(-)</td>\n",
       "      <td>CAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTAC...</td>\n",
       "      <td>GAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15715</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852700</td>\n",
       "      <td>73852750</td>\n",
       "      <td>XIST_5_11319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852700-73852750(-)</td>\n",
       "      <td>TCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTA...</td>\n",
       "      <td>AGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAA...</td>\n",
       "      <td>0.44</td>\n",
       "      <td>3</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15716 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname     start       end          name score strand  \\\n",
       "0        chrX  73820650  73820700      XIST_0_0     .      -   \n",
       "1        chrX  73820651  73820701      XIST_0_1     .      -   \n",
       "2        chrX  73820652  73820702      XIST_0_2     .      -   \n",
       "3        chrX  73820653  73820703      XIST_0_3     .      -   \n",
       "4        chrX  73820654  73820704      XIST_0_4     .      -   \n",
       "...       ...       ...       ...           ...   ...    ...   \n",
       "15711    chrX  73852696  73852746  XIST_5_11315     .      -   \n",
       "15712    chrX  73852697  73852747  XIST_5_11316     .      -   \n",
       "15713    chrX  73852698  73852748  XIST_5_11317     .      -   \n",
       "15714    chrX  73852699  73852749  XIST_5_11318     .      -   \n",
       "15715    chrX  73852700  73852750  XIST_5_11319     .      -   \n",
       "\n",
       "                      kmer_coord  \\\n",
       "0      chrX:73820650-73820700(-)   \n",
       "1      chrX:73820651-73820701(-)   \n",
       "2      chrX:73820652-73820702(-)   \n",
       "3      chrX:73820653-73820703(-)   \n",
       "4      chrX:73820654-73820704(-)   \n",
       "...                          ...   \n",
       "15711  chrX:73852696-73852746(-)   \n",
       "15712  chrX:73852697-73852747(-)   \n",
       "15713  chrX:73852698-73852748(-)   \n",
       "15714  chrX:73852699-73852749(-)   \n",
       "15715  chrX:73852700-73852750(-)   \n",
       "\n",
       "                                          transcript_seq  \\\n",
       "0      GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...   \n",
       "1      TGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAA...   \n",
       "2      CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...   \n",
       "3      ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...   \n",
       "4      AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...   \n",
       "...                                                  ...   \n",
       "15711  TTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCT...   \n",
       "15712  GTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTC...   \n",
       "15713  AGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACT...   \n",
       "15714  CAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTAC...   \n",
       "15715  TCAGTTCTTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTA...   \n",
       "\n",
       "                                               probe_seq  GC_content_full  \\\n",
       "0      CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...             0.30   \n",
       "1      AAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTC...             0.28   \n",
       "2      AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...             0.30   \n",
       "3      AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...             0.30   \n",
       "4      GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...             0.30   \n",
       "...                                                  ...              ...   \n",
       "15711  CGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTA...             0.46   \n",
       "15712  GAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAA...             0.46   \n",
       "15713  AGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAG...             0.44   \n",
       "15714  GAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGA...             0.46   \n",
       "15715  AGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCTTTAAGAA...             0.44   \n",
       "\n",
       "       longest_homopolymer  GC_content_LHS  GC_content_RHS  \n",
       "0                        4            0.20            0.40  \n",
       "1                        4            0.16            0.40  \n",
       "2                        4            0.16            0.44  \n",
       "3                        4            0.16            0.44  \n",
       "4                        4            0.16            0.44  \n",
       "...                    ...             ...             ...  \n",
       "15711                    3            0.44            0.48  \n",
       "15712                    3            0.44            0.48  \n",
       "15713                    3            0.44            0.44  \n",
       "15714                    3            0.44            0.48  \n",
       "15715                    3            0.44            0.44  \n",
       "\n",
       "[15716 rows x 13 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "44dfc696-c7b4-4b79-a382-8d3d9511db2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "has_required_nts\n",
      "False    11525\n",
      "True      4191\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## Check for required nucleotides in specific positions:\n",
    "if required_nts is not None:\n",
    "    kmers['has_required_nts'] = check_for_required_nts(kmers, required_nts)\n",
    "    print(kmers['has_required_nts'].value_counts())\n",
    "    ## Filter for required nucleotides\n",
    "    kmers = kmers[kmers['has_required_nts']==True].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a937f445-2786-400d-ae11-38233957bbe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export kmers before filtering\n",
    "kmers.to_csv((out_dir + 'kmers_candidates_unfiltered.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0fa42cd0-1b76-4948-b33f-ef5d057d9f01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820650</td>\n",
       "      <td>73820700</td>\n",
       "      <td>XIST_0_0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820650-73820700(-)</td>\n",
       "      <td>GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...</td>\n",
       "      <td>CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.40</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820652</td>\n",
       "      <td>73820702</td>\n",
       "      <td>XIST_0_2</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820652-73820702(-)</td>\n",
       "      <td>CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...</td>\n",
       "      <td>AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820653</td>\n",
       "      <td>73820703</td>\n",
       "      <td>XIST_0_3</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820653-73820703(-)</td>\n",
       "      <td>ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...</td>\n",
       "      <td>AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820654</td>\n",
       "      <td>73820704</td>\n",
       "      <td>XIST_0_4</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820654-73820704(-)</td>\n",
       "      <td>AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...</td>\n",
       "      <td>GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>4</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73820655</td>\n",
       "      <td>73820705</td>\n",
       "      <td>XIST_0_5</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73820655-73820705(-)</td>\n",
       "      <td>AAACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTG...</td>\n",
       "      <td>TTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCCA...</td>\n",
       "      <td>0.28</td>\n",
       "      <td>4</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4186</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852664</td>\n",
       "      <td>73852714</td>\n",
       "      <td>XIST_5_11283</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852664-73852714(-)</td>\n",
       "      <td>ATATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCT...</td>\n",
       "      <td>AGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGAA...</td>\n",
       "      <td>0.46</td>\n",
       "      <td>4</td>\n",
       "      <td>0.44</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4187</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852680</td>\n",
       "      <td>73852730</td>\n",
       "      <td>XIST_5_11299</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852680-73852730(-)</td>\n",
       "      <td>ATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCTGGAAGCT...</td>\n",
       "      <td>AGGAAGCTTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGC...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.44</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4188</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4189</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4191 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqname     start       end          name score strand  \\\n",
       "0       chrX  73820650  73820700      XIST_0_0     .      -   \n",
       "1       chrX  73820652  73820702      XIST_0_2     .      -   \n",
       "2       chrX  73820653  73820703      XIST_0_3     .      -   \n",
       "3       chrX  73820654  73820704      XIST_0_4     .      -   \n",
       "4       chrX  73820655  73820705      XIST_0_5     .      -   \n",
       "...      ...       ...       ...           ...   ...    ...   \n",
       "4186    chrX  73852664  73852714  XIST_5_11283     .      -   \n",
       "4187    chrX  73852680  73852730  XIST_5_11299     .      -   \n",
       "4188    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "4189    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "4190    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                     kmer_coord  \\\n",
       "0     chrX:73820650-73820700(-)   \n",
       "1     chrX:73820652-73820702(-)   \n",
       "2     chrX:73820653-73820703(-)   \n",
       "3     chrX:73820654-73820704(-)   \n",
       "4     chrX:73820655-73820705(-)   \n",
       "...                         ...   \n",
       "4186  chrX:73852664-73852714(-)   \n",
       "4187  chrX:73852680-73852730(-)   \n",
       "4188  chrX:73852687-73852737(-)   \n",
       "4189  chrX:73852689-73852739(-)   \n",
       "4190  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                         transcript_seq  \\\n",
       "0     GGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAAAC...   \n",
       "1     CTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAAA...   \n",
       "2     ACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGAA...   \n",
       "3     AACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTGA...   \n",
       "4     AAACTGGGGAGTTGGTTGCTATTGTAAAATAAAATATACTGTTTTG...   \n",
       "...                                                 ...   \n",
       "4186  ATATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCT...   \n",
       "4187  ATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCTGGAAGCT...   \n",
       "4188  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "4189  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "4190  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                              probe_seq  GC_content_full  \\\n",
       "0     CAAAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACT...             0.30   \n",
       "1     AAGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCC...             0.30   \n",
       "2     AGTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCC...             0.30   \n",
       "3     GTTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCC...             0.30   \n",
       "4     TTTTCAAAACAGTATATTTTATTTTACAATAGCAACCAACTCCCCA...             0.28   \n",
       "...                                                 ...              ...   \n",
       "4186  AGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGAA...             0.46   \n",
       "4187  AGGAAGCTTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGC...             0.52   \n",
       "4188  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "4189  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "4190  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "      longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                       4            0.20            0.40              True  \n",
       "1                       4            0.16            0.44              True  \n",
       "2                       4            0.16            0.44              True  \n",
       "3                       4            0.16            0.44              True  \n",
       "4                       4            0.12            0.44              True  \n",
       "...                   ...             ...             ...               ...  \n",
       "4186                    4            0.44            0.48              True  \n",
       "4187                    4            0.60            0.44              True  \n",
       "4188                    4            0.48            0.60              True  \n",
       "4189                    4            0.48            0.64              True  \n",
       "4190                    4            0.48            0.52              True  \n",
       "\n",
       "[4191 rows x 14 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6b903375-7a4b-4878-9479-574bb359c563",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter for GC content\n",
    "kmers = filter_by_GC_content(kmers, min_GC, max_GC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1ed1ae93-3577-4a85-a65a-9e506d4c317a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821073</td>\n",
       "      <td>73821123</td>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821073-73821123(-)</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...</td>\n",
       "      <td>AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821085</td>\n",
       "      <td>73821135</td>\n",
       "      <td>XIST_0_435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821085-73821135(-)</td>\n",
       "      <td>TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...</td>\n",
       "      <td>CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821087</td>\n",
       "      <td>73821137</td>\n",
       "      <td>XIST_0_437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821087-73821137(-)</td>\n",
       "      <td>TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...</td>\n",
       "      <td>CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821088</td>\n",
       "      <td>73821138</td>\n",
       "      <td>XIST_0_438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821088-73821138(-)</td>\n",
       "      <td>ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...</td>\n",
       "      <td>ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821091</td>\n",
       "      <td>73821141</td>\n",
       "      <td>XIST_0_441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821091-73821141(-)</td>\n",
       "      <td>TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...</td>\n",
       "      <td>CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852654</td>\n",
       "      <td>73852704</td>\n",
       "      <td>XIST_5_11273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852654-73852704(-)</td>\n",
       "      <td>CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...</td>\n",
       "      <td>CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852663</td>\n",
       "      <td>73852713</td>\n",
       "      <td>XIST_5_11282</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852663-73852713(-)</td>\n",
       "      <td>TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...</td>\n",
       "      <td>CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>534 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73821073  73821123    XIST_0_423     .      -   \n",
       "1      chrX  73821085  73821135    XIST_0_435     .      -   \n",
       "2      chrX  73821087  73821137    XIST_0_437     .      -   \n",
       "3      chrX  73821088  73821138    XIST_0_438     .      -   \n",
       "4      chrX  73821091  73821141    XIST_0_441     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "529    chrX  73852654  73852704  XIST_5_11273     .      -   \n",
       "530    chrX  73852663  73852713  XIST_5_11282     .      -   \n",
       "531    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "532    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "533    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73821073-73821123(-)   \n",
       "1    chrX:73821085-73821135(-)   \n",
       "2    chrX:73821087-73821137(-)   \n",
       "3    chrX:73821088-73821138(-)   \n",
       "4    chrX:73821091-73821141(-)   \n",
       "..                         ...   \n",
       "529  chrX:73852654-73852704(-)   \n",
       "530  chrX:73852663-73852713(-)   \n",
       "531  chrX:73852687-73852737(-)   \n",
       "532  chrX:73852689-73852739(-)   \n",
       "533  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...   \n",
       "1    TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...   \n",
       "2    TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...   \n",
       "3    ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...   \n",
       "4    TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...   \n",
       "..                                                 ...   \n",
       "529  CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...   \n",
       "530  TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...   \n",
       "531  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "532  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "533  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...             0.48   \n",
       "1    CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...             0.54   \n",
       "2    CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...             0.52   \n",
       "3    ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...             0.50   \n",
       "4    CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...             0.50   \n",
       "..                                                 ...              ...   \n",
       "529  CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...             0.58   \n",
       "530  CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...             0.48   \n",
       "531  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "532  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "533  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                      4            0.48            0.48              True  \n",
       "1                      4            0.60            0.48              True  \n",
       "2                      4            0.56            0.48              True  \n",
       "3                      4            0.52            0.48              True  \n",
       "4                      4            0.52            0.48              True  \n",
       "..                   ...             ...             ...               ...  \n",
       "529                    4            0.52            0.64              True  \n",
       "530                    4            0.48            0.48              True  \n",
       "531                    4            0.48            0.60              True  \n",
       "532                    4            0.48            0.64              True  \n",
       "533                    4            0.48            0.52              True  \n",
       "\n",
       "[534 rows x 14 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1a7eb377-dce8-4e92-aa44-bbdec9172d27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821073</td>\n",
       "      <td>73821123</td>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821073-73821123(-)</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...</td>\n",
       "      <td>AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821085</td>\n",
       "      <td>73821135</td>\n",
       "      <td>XIST_0_435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821085-73821135(-)</td>\n",
       "      <td>TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...</td>\n",
       "      <td>CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821087</td>\n",
       "      <td>73821137</td>\n",
       "      <td>XIST_0_437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821087-73821137(-)</td>\n",
       "      <td>TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...</td>\n",
       "      <td>CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821088</td>\n",
       "      <td>73821138</td>\n",
       "      <td>XIST_0_438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821088-73821138(-)</td>\n",
       "      <td>ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...</td>\n",
       "      <td>ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821091</td>\n",
       "      <td>73821141</td>\n",
       "      <td>XIST_0_441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821091-73821141(-)</td>\n",
       "      <td>TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...</td>\n",
       "      <td>CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852654</td>\n",
       "      <td>73852704</td>\n",
       "      <td>XIST_5_11273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852654-73852704(-)</td>\n",
       "      <td>CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...</td>\n",
       "      <td>CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852663</td>\n",
       "      <td>73852713</td>\n",
       "      <td>XIST_5_11282</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852663-73852713(-)</td>\n",
       "      <td>TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...</td>\n",
       "      <td>CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>532</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>533</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>534 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73821073  73821123    XIST_0_423     .      -   \n",
       "1      chrX  73821085  73821135    XIST_0_435     .      -   \n",
       "2      chrX  73821087  73821137    XIST_0_437     .      -   \n",
       "3      chrX  73821088  73821138    XIST_0_438     .      -   \n",
       "4      chrX  73821091  73821141    XIST_0_441     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "529    chrX  73852654  73852704  XIST_5_11273     .      -   \n",
       "530    chrX  73852663  73852713  XIST_5_11282     .      -   \n",
       "531    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "532    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "533    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73821073-73821123(-)   \n",
       "1    chrX:73821085-73821135(-)   \n",
       "2    chrX:73821087-73821137(-)   \n",
       "3    chrX:73821088-73821138(-)   \n",
       "4    chrX:73821091-73821141(-)   \n",
       "..                         ...   \n",
       "529  chrX:73852654-73852704(-)   \n",
       "530  chrX:73852663-73852713(-)   \n",
       "531  chrX:73852687-73852737(-)   \n",
       "532  chrX:73852689-73852739(-)   \n",
       "533  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...   \n",
       "1    TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...   \n",
       "2    TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...   \n",
       "3    ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...   \n",
       "4    TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...   \n",
       "..                                                 ...   \n",
       "529  CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...   \n",
       "530  TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...   \n",
       "531  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "532  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "533  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...             0.48   \n",
       "1    CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...             0.54   \n",
       "2    CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...             0.52   \n",
       "3    ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...             0.50   \n",
       "4    CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...             0.50   \n",
       "..                                                 ...              ...   \n",
       "529  CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...             0.58   \n",
       "530  CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...             0.48   \n",
       "531  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "532  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "533  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \n",
       "0                      4            0.48            0.48              True  \n",
       "1                      4            0.60            0.48              True  \n",
       "2                      4            0.56            0.48              True  \n",
       "3                      4            0.52            0.48              True  \n",
       "4                      4            0.52            0.48              True  \n",
       "..                   ...             ...             ...               ...  \n",
       "529                    4            0.52            0.64              True  \n",
       "530                    4            0.48            0.48              True  \n",
       "531                    4            0.48            0.60              True  \n",
       "532                    4            0.48            0.64              True  \n",
       "533                    4            0.48            0.52              True  \n",
       "\n",
       "[534 rows x 14 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Candidate kmers\n",
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e482ac72-5ed2-46dd-a8bb-1950d1a8aa14",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers.to_csv((out_dir + 'kmers_candidates_filtered.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fbba816-f29f-40cc-be17-6e44729c6d87",
   "metadata": {},
   "source": [
    "### 5. Remove probes with potential off-targets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bede588-f85c-4321-a15d-d48498d802d2",
   "metadata": {},
   "source": [
    "Having identified a set of kmers that fulfill our sequence requirements, we can next proceed with testing whether they are specific to our transcript/exon of interest.\n",
    "\n",
    "For this we rely on using BLAST. \n",
    "\n",
    "We recommend blasting against transcripts (i.e., exons and introns combined) to be as conservative as possible in terms of off-targets. \n",
    "However, in cases where it is not possible to obtain enough suitable kmers (e.g., for  short transcripts), it is reasonable to relax this requirement by BLASTing against exons only (a much smaller search space).\n",
    "\n",
    "At this step, we also want to consider whether our probes are split (as in the current specifications for VisiumHD) or a single oligo. If probes are split, it's best to BLAST each side separately, to make sure that both sides are specific."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3ace2aec-f433-49c6-90c6-8a971232be41",
   "metadata": {},
   "outputs": [],
   "source": [
    "## The first thing to do is to export our sequences in fasta format, so that we can use them for BLAST\n",
    "write_fasta(kmers['name'], kmers['transcript_seq'], (out_dir + 'kmers_candidates_filtered_transcript_seqs.fa'))\n",
    "## If our probes are meant to be split, we should additionally blast them separately \n",
    "## Note that the LHS/RHS in the transcript are reversed compared to the probe (i.e., the LHS of the transcript is complementary to the RHS of the probe)\n",
    "if split_nt is not None: \n",
    "    ## Make split probes\n",
    "    kmers['transcript_seq_LHS'] = [seq[0:split_nt] for seq in kmers['transcript_seq']]\n",
    "    kmers['transcript_seq_RHS'] = [seq[split_nt: probe_length] for seq in kmers['transcript_seq']]\n",
    "\n",
    "    ## We are exporting the transcript sequence as that's the one that has to be blasted against the human transcriptome\n",
    "    write_fasta(kmers['name'], kmers['transcript_seq_LHS'], (out_dir + 'kmers_candidates_filtered_transcript_seqs_LHS.fa'))\n",
    "    write_fasta(kmers['name'], kmers['transcript_seq_RHS'], (out_dir + 'kmers_candidates_filtered_transcript_seqs_RHS.fa'))    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0f5255b6-3eb9-4912-a75a-c2dcf6cad93a",
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_res = {}\n",
    "## First, blast the full probe\n",
    "blast_res['full'] = run_blast(fasta=(out_dir + 'kmers_candidates_filtered_transcript_seqs.fa'),\n",
    "                              blastdb = blast_db,\n",
    "                              path2blastn=(blast_exec_path + 'blastn'),\n",
    "                              outfile = (out_dir + 'kmers_candidates_filtered_blast_output.txt'))\n",
    "\n",
    "## Additionally, if probe is split, blast each side separately\n",
    "if split_nt is not None: \n",
    "    blast_res['LHS'] = run_blast(fasta=(out_dir + 'kmers_candidates_filtered_transcript_seqs_LHS.fa'),\n",
    "                                     blastdb = blast_db,\n",
    "                                     path2blastn=(blast_exec_path + 'blastn'),\n",
    "                                     outfile = (out_dir + 'kmers_candidates_filtered_blast_output_LHS.txt'))\n",
    "    blast_res['RHS'] = run_blast(fasta=(out_dir + 'kmers_candidates_filtered_transcript_seqs_RHS.fa'),\n",
    "                                     blastdb = blast_db,\n",
    "                                     path2blastn=(blast_exec_path + 'blastn'),\n",
    "                                     outfile = (out_dir + 'kmers_candidates_filtered_blast_output_RHS.txt'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bf80b05a-1ed0-4e4d-8937-b39b7aa751b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following genes were detected in mode: full\n",
      "sgeneid\n",
      "XIST       1228\n",
      "MYT1L       174\n",
      "ANO4        102\n",
      "SLC45A4      60\n",
      "BRAF         57\n",
      "JARID2       56\n",
      "FAAH2        54\n",
      "LRRC8A       45\n",
      "PACSIN2      40\n",
      "SORBS2       38\n",
      "Name: count, dtype: int64\n",
      "The following genes were detected in mode: LHS\n",
      "sgeneid\n",
      "XIST      887\n",
      "KIFC3      78\n",
      "AGPAT3     54\n",
      "ELAVL4     54\n",
      "NHSL1      52\n",
      "TLR5       44\n",
      "MYT1L      29\n",
      "NR2C2      25\n",
      "ANKS1B     21\n",
      "JARID2     20\n",
      "Name: count, dtype: int64\n",
      "The following genes were detected in mode: RHS\n",
      "sgeneid\n",
      "XIST       803\n",
      "NT5C2      108\n",
      "TRERF1      52\n",
      "DOCK9       43\n",
      "MAP7        39\n",
      "PLEKHA7     38\n",
      "GSE1        29\n",
      "NHSL1       26\n",
      "COBL        22\n",
      "FGF14       22\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "for k in blast_res.keys():\n",
    "    print((\"The following genes were detected in mode: \" +  k))\n",
    "    print(blast_res[k]['sgeneid'].value_counts().head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1ea1f7f6-cc0d-4f98-8639-c5087f1ea223",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>sgeneid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>31631</td>\n",
       "      <td>31680</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XIST_0_435</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>31619</td>\n",
       "      <td>31668</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XIST_0_437</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>31617</td>\n",
       "      <td>31666</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XIST_0_438</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>31616</td>\n",
       "      <td>31665</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XIST_0_441</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>31613</td>\n",
       "      <td>31662</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2364</th>\n",
       "      <td>XIST_5_11273</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "      <td>99</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2365</th>\n",
       "      <td>XIST_5_11282</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>41</td>\n",
       "      <td>90</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2366</th>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>17</td>\n",
       "      <td>66</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2367</th>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>15</td>\n",
       "      <td>64</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2368</th>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>XIST::chrX:73820650-73852753(-)</td>\n",
       "      <td>100.0</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>11</td>\n",
       "      <td>60</td>\n",
       "      <td>8.510000e-17</td>\n",
       "      <td>91.5</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2369 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              name                           sseqid  pident  length  mismatch  \\\n",
       "0       XIST_0_423  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "1       XIST_0_435  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "2       XIST_0_437  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "3       XIST_0_438  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "4       XIST_0_441  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "...            ...                              ...     ...     ...       ...   \n",
       "2364  XIST_5_11273  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "2365  XIST_5_11282  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "2366  XIST_5_11306  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "2367  XIST_5_11308  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "2368  XIST_5_11312  XIST::chrX:73820650-73852753(-)   100.0      50         0   \n",
       "\n",
       "      gapopen  qstart  qend  sstart   send        evalue  bitscore sgeneid  \n",
       "0           0       1    50   31631  31680  8.510000e-17      91.5    XIST  \n",
       "1           0       1    50   31619  31668  8.510000e-17      91.5    XIST  \n",
       "2           0       1    50   31617  31666  8.510000e-17      91.5    XIST  \n",
       "3           0       1    50   31616  31665  8.510000e-17      91.5    XIST  \n",
       "4           0       1    50   31613  31662  8.510000e-17      91.5    XIST  \n",
       "...       ...     ...   ...     ...    ...           ...       ...     ...  \n",
       "2364        0       1    50      50     99  8.510000e-17      91.5    XIST  \n",
       "2365        0       1    50      41     90  8.510000e-17      91.5    XIST  \n",
       "2366        0       1    50      17     66  8.510000e-17      91.5    XIST  \n",
       "2367        0       1    50      15     64  8.510000e-17      91.5    XIST  \n",
       "2368        0       1    50      11     60  8.510000e-17      91.5    XIST  \n",
       "\n",
       "[2369 rows x 13 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_res['full']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0d97396-3b89-4a62-9581-0027bff69048",
   "metadata": {},
   "source": [
    "Not all BLAST hits will be off-targets. Hopefully, our gene of interest is included in the BLAST output. We therefore need to filter for hits with different gene IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f2d394e1-35a7-461a-b94e-fa9db71984b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "offtargets = []\n",
    "for k in blast_res.keys():\n",
    "    offtargets += (detect_offtargets(blast_res[k], gene_ID, min_mismatches=min_mismatches))\n",
    "## Remove redundancies\n",
    "offtargets = list(set(offtargets))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "73ace4d0-5850-47e7-9770-3358b78237c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "162"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(offtargets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b25ebea4-fa7c-44f9-b7f0-e6fb32149341",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Remove off-targets\n",
    "kmers = kmers[kmers['name'].isin(offtargets)==False].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "1a1d18ca-0824-4094-94f5-d4a5a0645366",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "      <th>transcript_seq_LHS</th>\n",
       "      <th>transcript_seq_RHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821073</td>\n",
       "      <td>73821123</td>\n",
       "      <td>XIST_0_423</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821073-73821123(-)</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...</td>\n",
       "      <td>AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>TGGACTCAGTAACACCCCTTTCTTC</td>\n",
       "      <td>AGCTGGGGATGGGGAATGGATTATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821085</td>\n",
       "      <td>73821135</td>\n",
       "      <td>XIST_0_435</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821085-73821135(-)</td>\n",
       "      <td>TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...</td>\n",
       "      <td>CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>TGTGGTATCAGCTGGACTCAGTAAC</td>\n",
       "      <td>ACCCCTTTCTTCAGCTGGGGATGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821087</td>\n",
       "      <td>73821137</td>\n",
       "      <td>XIST_0_437</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821087-73821137(-)</td>\n",
       "      <td>TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...</td>\n",
       "      <td>CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>4</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>TGTGTGGTATCAGCTGGACTCAGTA</td>\n",
       "      <td>ACACCCCTTTCTTCAGCTGGGGATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821088</td>\n",
       "      <td>73821138</td>\n",
       "      <td>XIST_0_438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821088-73821138(-)</td>\n",
       "      <td>ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...</td>\n",
       "      <td>ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>ATGTGTGGTATCAGCTGGACTCAGT</td>\n",
       "      <td>AACACCCCTTTCTTCAGCTGGGGAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73821091</td>\n",
       "      <td>73821141</td>\n",
       "      <td>XIST_0_441</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73821091-73821141(-)</td>\n",
       "      <td>TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...</td>\n",
       "      <td>CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>TCAATGTGTGGTATCAGCTGGACTC</td>\n",
       "      <td>AGTAACACCCCTTTCTTCAGCTGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852654</td>\n",
       "      <td>73852704</td>\n",
       "      <td>XIST_5_11273</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852654-73852704(-)</td>\n",
       "      <td>CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...</td>\n",
       "      <td>CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>4</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "      <td>CTCTCTCGGGGCTGGAAGCTTCCTG</td>\n",
       "      <td>ACTGAAGATCTCTCTGCACTTGGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852663</td>\n",
       "      <td>73852713</td>\n",
       "      <td>XIST_5_11282</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852663-73852713(-)</td>\n",
       "      <td>TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...</td>\n",
       "      <td>CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...</td>\n",
       "      <td>0.48</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>TATTTCTTACTCTCTCGGGGCTGGA</td>\n",
       "      <td>AGCTTCCTGACTGAAGATCTCTCTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852687</td>\n",
       "      <td>73852737</td>\n",
       "      <td>XIST_5_11306</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852687-73852737(-)</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...</td>\n",
       "      <td>TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>CGCTGCAATTCGCTGCTGCAGCCAT</td>\n",
       "      <td>ATTTCTTACTCTCTCGGGGCTGGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852689</td>\n",
       "      <td>73852739</td>\n",
       "      <td>XIST_5_11308</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852689-73852739(-)</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...</td>\n",
       "      <td>CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.64</td>\n",
       "      <td>True</td>\n",
       "      <td>AGCGCTGCAATTCGCTGCTGCAGCC</td>\n",
       "      <td>ATATTTCTTACTCTCTCGGGGCTGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852693</td>\n",
       "      <td>73852743</td>\n",
       "      <td>XIST_5_11312</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852693-73852743(-)</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...</td>\n",
       "      <td>CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...</td>\n",
       "      <td>0.50</td>\n",
       "      <td>4</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "      <td>TTAAAGCGCTGCAATTCGCTGCTGC</td>\n",
       "      <td>AGCCATATTTCTTACTCTCTCGGGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>372 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73821073  73821123    XIST_0_423     .      -   \n",
       "1      chrX  73821085  73821135    XIST_0_435     .      -   \n",
       "2      chrX  73821087  73821137    XIST_0_437     .      -   \n",
       "3      chrX  73821088  73821138    XIST_0_438     .      -   \n",
       "4      chrX  73821091  73821141    XIST_0_441     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "367    chrX  73852654  73852704  XIST_5_11273     .      -   \n",
       "368    chrX  73852663  73852713  XIST_5_11282     .      -   \n",
       "369    chrX  73852687  73852737  XIST_5_11306     .      -   \n",
       "370    chrX  73852689  73852739  XIST_5_11308     .      -   \n",
       "371    chrX  73852693  73852743  XIST_5_11312     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73821073-73821123(-)   \n",
       "1    chrX:73821085-73821135(-)   \n",
       "2    chrX:73821087-73821137(-)   \n",
       "3    chrX:73821088-73821138(-)   \n",
       "4    chrX:73821091-73821141(-)   \n",
       "..                         ...   \n",
       "367  chrX:73852654-73852704(-)   \n",
       "368  chrX:73852663-73852713(-)   \n",
       "369  chrX:73852687-73852737(-)   \n",
       "370  chrX:73852689-73852739(-)   \n",
       "371  chrX:73852693-73852743(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    TGGACTCAGTAACACCCCTTTCTTCAGCTGGGGATGGGGAATGGAT...   \n",
       "1    TGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGGGA...   \n",
       "2    TGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGGG...   \n",
       "3    ATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGCTGG...   \n",
       "4    TCAATGTGTGGTATCAGCTGGACTCAGTAACACCCCTTTCTTCAGC...   \n",
       "..                                                 ...   \n",
       "367  CTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTCTCTGCACTT...   \n",
       "368  TATTTCTTACTCTCTCGGGGCTGGAAGCTTCCTGACTGAAGATCTC...   \n",
       "369  CGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGGCT...   \n",
       "370  AGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTCGGGG...   \n",
       "371  TTAAAGCGCTGCAATTCGCTGCTGCAGCCATATTTCTTACTCTCTC...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    AATAATCCATTCCCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAG...             0.48   \n",
       "1    CCCATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATAC...             0.54   \n",
       "2    CATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCA...             0.52   \n",
       "3    ATCCCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCAC...             0.50   \n",
       "4    CCCAGCTGAAGAAAGGGGTGTTACTGAGTCCAGCTGATACCACACA...             0.50   \n",
       "..                                                 ...              ...   \n",
       "367  CCCCAAGTGCAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAG...             0.58   \n",
       "368  CAGAGAGATCTTCAGTCAGGAAGCTTCCAGCCCCGAGAGAGTAAGA...             0.48   \n",
       "369  TTCCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGC...             0.54   \n",
       "370  CCAGCCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAG...             0.56   \n",
       "371  CCCCGAGAGAGTAAGAAATATGGCTGCAGCAGCGAATTGCAGCGCT...             0.50   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \\\n",
       "0                      4            0.48            0.48              True   \n",
       "1                      4            0.60            0.48              True   \n",
       "2                      4            0.56            0.48              True   \n",
       "3                      4            0.52            0.48              True   \n",
       "4                      4            0.52            0.48              True   \n",
       "..                   ...             ...             ...               ...   \n",
       "367                    4            0.52            0.64              True   \n",
       "368                    4            0.48            0.48              True   \n",
       "369                    4            0.48            0.60              True   \n",
       "370                    4            0.48            0.64              True   \n",
       "371                    4            0.48            0.52              True   \n",
       "\n",
       "            transcript_seq_LHS         transcript_seq_RHS  \n",
       "0    TGGACTCAGTAACACCCCTTTCTTC  AGCTGGGGATGGGGAATGGATTATT  \n",
       "1    TGTGGTATCAGCTGGACTCAGTAAC  ACCCCTTTCTTCAGCTGGGGATGGG  \n",
       "2    TGTGTGGTATCAGCTGGACTCAGTA  ACACCCCTTTCTTCAGCTGGGGATG  \n",
       "3    ATGTGTGGTATCAGCTGGACTCAGT  AACACCCCTTTCTTCAGCTGGGGAT  \n",
       "4    TCAATGTGTGGTATCAGCTGGACTC  AGTAACACCCCTTTCTTCAGCTGGG  \n",
       "..                         ...                        ...  \n",
       "367  CTCTCTCGGGGCTGGAAGCTTCCTG  ACTGAAGATCTCTCTGCACTTGGGG  \n",
       "368  TATTTCTTACTCTCTCGGGGCTGGA  AGCTTCCTGACTGAAGATCTCTCTG  \n",
       "369  CGCTGCAATTCGCTGCTGCAGCCAT  ATTTCTTACTCTCTCGGGGCTGGAA  \n",
       "370  AGCGCTGCAATTCGCTGCTGCAGCC  ATATTTCTTACTCTCTCGGGGCTGG  \n",
       "371  TTAAAGCGCTGCAATTCGCTGCTGC  AGCCATATTTCTTACTCTCTCGGGG  \n",
       "\n",
       "[372 rows x 16 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77d4ead9-b6b5-4106-9171-fa3cd07eb5e9",
   "metadata": {},
   "source": [
    "### 6. Select non-overlapping probes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56719d4a-ad9f-4ff0-9e89-0c2fcbc55041",
   "metadata": {},
   "source": [
    "At this point, we have effectively acquired a set of usable probes. They don't overlap undesirable regions (repeats/polymorphism), have desirable sequence features (GC content, specific nucleotides) and are specific to our gene of interest.\n",
    "\n",
    "In this particular case, we still have a lot of possible k-mers (much more than the number of probes we intend to design). We can therefore choose to prioritise k-mers with shorter homopolymer stretches, as these are also discouraged by the [10x recommendations](https://cdn.10xgenomics.com/image/upload/v1697739385/support-documents/CG000621_CustomProbeDesign_TechNote_RevC.pdf).\n",
    "\n",
    "However, at this stage you might want to consider ranking probes in a diffferent way, depending on your application.\n",
    "\n",
    "After having ranked our k-mers in whatever way we think is reasonable at this stage, we can proceed with selecting the top probe, then removing all overlapping/adjacent probes (within a window determined by `probe_offset`).\n",
    "\n",
    "Since we have so many available probes, we will be increasing `probe_offset` from `100 (default)` to `1000`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c0d8c8f9-0c4c-46a4-bb8f-16782e47d379",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Sort in increasing homopolymer length\n",
    "kmers = kmers.sort_values('longest_homopolymer', ascending=True).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d951231a-8c94-4473-978d-b202c2ded194",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "      <th>transcript_seq_LHS</th>\n",
       "      <th>transcript_seq_RHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73827196</td>\n",
       "      <td>73827246</td>\n",
       "      <td>XIST_0_6546</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73827196-73827246(-)</td>\n",
       "      <td>CAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATG...</td>\n",
       "      <td>CTGGCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGT...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "      <td>CAGAACGTCGGCTATCAGAGCAAGC</td>\n",
       "      <td>ATTGTGGAGCGGTTCCTTATGCCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73827173</td>\n",
       "      <td>73827223</td>\n",
       "      <td>XIST_0_6523</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73827173-73827223(-)</td>\n",
       "      <td>GCATTGTGGAGCGGTTCCTTATGCCAGGCTGCCATGTGAGATGATC...</td>\n",
       "      <td>CTTGGATCATCTCACATGGCAGCCTGGCATAAGGAACCGCTCCACA...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "      <td>GCATTGTGGAGCGGTTCCTTATGCC</td>\n",
       "      <td>AGGCTGCCATGTGAGATGATCCAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73827178</td>\n",
       "      <td>73827228</td>\n",
       "      <td>XIST_0_6528</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73827178-73827228(-)</td>\n",
       "      <td>AGCAAGCATTGTGGAGCGGTTCCTTATGCCAGGCTGCCATGTGAGA...</td>\n",
       "      <td>ATCATCTCACATGGCAGCCTGGCATAAGGAACCGCTCCACAATGCT...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>2</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "      <td>AGCAAGCATTGTGGAGCGGTTCCTT</td>\n",
       "      <td>ATGCCAGGCTGCCATGTGAGATGAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73827189</td>\n",
       "      <td>73827239</td>\n",
       "      <td>XIST_0_6539</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73827189-73827239(-)</td>\n",
       "      <td>TCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATGCCAGGCT...</td>\n",
       "      <td>TGGCAGCCTGGCATAAGGAACCGCTCCACAATGCTTGCTCTGATAG...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>2</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "      <td>TCGGCTATCAGAGCAAGCATTGTGG</td>\n",
       "      <td>AGCGGTTCCTTATGCCAGGCTGCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850525</td>\n",
       "      <td>73850575</td>\n",
       "      <td>XIST_5_9144</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850525-73850575(-)</td>\n",
       "      <td>GTTGGCCAACAGGTGGCAGAAGAGGAATTCCTGCCTTCCTCAAGAG...</td>\n",
       "      <td>GTTCCTCTTGAGGAAGGCAGGAATTCCTCTTCTGCCACCTGTTGGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>GTTGGCCAACAGGTGGCAGAAGAGG</td>\n",
       "      <td>AATTCCTGCCTTCCTCAAGAGGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73845767</td>\n",
       "      <td>73845817</td>\n",
       "      <td>XIST_5_4386</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73845767-73845817(-)</td>\n",
       "      <td>CACTCCCAGTCTTCCTTTCCCTTCCAGCAGGGAGTGCCCCCTCCAT...</td>\n",
       "      <td>TCTTATGGAGGGGGCACTCCCTGCTGGAAGGGAAAGGAAGACTGGG...</td>\n",
       "      <td>0.58</td>\n",
       "      <td>5</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "      <td>CACTCCCAGTCTTCCTTTCCCTTCC</td>\n",
       "      <td>AGCAGGGAGTGCCCCCTCCATAAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73852326</td>\n",
       "      <td>73852376</td>\n",
       "      <td>XIST_5_10945</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73852326-73852376(-)</td>\n",
       "      <td>GGGGCTGCGGATACCTGGTTTTATTATTTTTTCTTTGCCCAACGGG...</td>\n",
       "      <td>CGGCCCCGTTGGGCAAAGAAAAAATAATAAAACCAGGTATCCGCAG...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>6</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "      <td>GGGGCTGCGGATACCTGGTTTTATT</td>\n",
       "      <td>ATTTTTTCTTTGCCCAACGGGGCCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850758</td>\n",
       "      <td>73850808</td>\n",
       "      <td>XIST_5_9377</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850758-73850808(-)</td>\n",
       "      <td>CTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccacc...</td>\n",
       "      <td>ggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAAT...</td>\n",
       "      <td>0.60</td>\n",
       "      <td>6</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.52</td>\n",
       "      <td>True</td>\n",
       "      <td>CTGTATTGGTGCCTCACCTAAGGCT</td>\n",
       "      <td>AAGTATACCTCCccccccacccccc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850763</td>\n",
       "      <td>73850813</td>\n",
       "      <td>XIST_5_9382</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850763-73850813(-)</td>\n",
       "      <td>TTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCcccc...</td>\n",
       "      <td>gtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGA...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>6</td>\n",
       "      <td>0.60</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>TTCCTCTGTATTGGTGCCTCACCTA</td>\n",
       "      <td>AGGCTAAGTATACCTCCccccccac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850764</td>\n",
       "      <td>73850814</td>\n",
       "      <td>XIST_5_9383</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850764-73850814(-)</td>\n",
       "      <td>ATTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccc...</td>\n",
       "      <td>tggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGAG...</td>\n",
       "      <td>0.52</td>\n",
       "      <td>6</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.48</td>\n",
       "      <td>True</td>\n",
       "      <td>ATTCCTCTGTATTGGTGCCTCACCT</td>\n",
       "      <td>AAGGCTAAGTATACCTCCcccccca</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>372 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqname     start       end          name score strand  \\\n",
       "0      chrX  73827196  73827246   XIST_0_6546     .      -   \n",
       "1      chrX  73827173  73827223   XIST_0_6523     .      -   \n",
       "2      chrX  73827178  73827228   XIST_0_6528     .      -   \n",
       "3      chrX  73827189  73827239   XIST_0_6539     .      -   \n",
       "4      chrX  73850525  73850575   XIST_5_9144     .      -   \n",
       "..      ...       ...       ...           ...   ...    ...   \n",
       "367    chrX  73845767  73845817   XIST_5_4386     .      -   \n",
       "368    chrX  73852326  73852376  XIST_5_10945     .      -   \n",
       "369    chrX  73850758  73850808   XIST_5_9377     .      -   \n",
       "370    chrX  73850763  73850813   XIST_5_9382     .      -   \n",
       "371    chrX  73850764  73850814   XIST_5_9383     .      -   \n",
       "\n",
       "                    kmer_coord  \\\n",
       "0    chrX:73827196-73827246(-)   \n",
       "1    chrX:73827173-73827223(-)   \n",
       "2    chrX:73827178-73827228(-)   \n",
       "3    chrX:73827189-73827239(-)   \n",
       "4    chrX:73850525-73850575(-)   \n",
       "..                         ...   \n",
       "367  chrX:73845767-73845817(-)   \n",
       "368  chrX:73852326-73852376(-)   \n",
       "369  chrX:73850758-73850808(-)   \n",
       "370  chrX:73850763-73850813(-)   \n",
       "371  chrX:73850764-73850814(-)   \n",
       "\n",
       "                                        transcript_seq  \\\n",
       "0    CAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATG...   \n",
       "1    GCATTGTGGAGCGGTTCCTTATGCCAGGCTGCCATGTGAGATGATC...   \n",
       "2    AGCAAGCATTGTGGAGCGGTTCCTTATGCCAGGCTGCCATGTGAGA...   \n",
       "3    TCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATGCCAGGCT...   \n",
       "4    GTTGGCCAACAGGTGGCAGAAGAGGAATTCCTGCCTTCCTCAAGAG...   \n",
       "..                                                 ...   \n",
       "367  CACTCCCAGTCTTCCTTTCCCTTCCAGCAGGGAGTGCCCCCTCCAT...   \n",
       "368  GGGGCTGCGGATACCTGGTTTTATTATTTTTTCTTTGCCCAACGGG...   \n",
       "369  CTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccccccacc...   \n",
       "370  TTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCcccc...   \n",
       "371  ATTCCTCTGTATTGGTGCCTCACCTAAGGCTAAGTATACCTCCccc...   \n",
       "\n",
       "                                             probe_seq  GC_content_full  \\\n",
       "0    CTGGCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGT...             0.54   \n",
       "1    CTTGGATCATCTCACATGGCAGCCTGGCATAAGGAACCGCTCCACA...             0.54   \n",
       "2    ATCATCTCACATGGCAGCCTGGCATAAGGAACCGCTCCACAATGCT...             0.52   \n",
       "3    TGGCAGCCTGGCATAAGGAACCGCTCCACAATGCTTGCTCTGATAG...             0.56   \n",
       "4    GTTCCTCTTGAGGAAGGCAGGAATTCCTCTTCTGCCACCTGTTGGC...             0.54   \n",
       "..                                                 ...              ...   \n",
       "367  TCTTATGGAGGGGGCACTCCCTGCTGGAAGGGAAAGGAAGACTGGG...             0.58   \n",
       "368  CGGCCCCGTTGGGCAAAGAAAAAATAATAAAACCAGGTATCCGCAG...             0.52   \n",
       "369  ggggggtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAAT...             0.60   \n",
       "370  gtggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGA...             0.54   \n",
       "371  tggggggGGAGGTATACTTAGCCTTAGGTGAGGCACCAATACAGAG...             0.52   \n",
       "\n",
       "     longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \\\n",
       "0                      2            0.52            0.56              True   \n",
       "1                      2            0.52            0.56              True   \n",
       "2                      2            0.52            0.52              True   \n",
       "3                      2            0.60            0.52              True   \n",
       "4                      2            0.48            0.60              True   \n",
       "..                   ...             ...             ...               ...   \n",
       "367                    5            0.60            0.56              True   \n",
       "368                    6            0.52            0.52              True   \n",
       "369                    6            0.68            0.52              True   \n",
       "370                    6            0.60            0.48              True   \n",
       "371                    6            0.56            0.48              True   \n",
       "\n",
       "            transcript_seq_LHS         transcript_seq_RHS  \n",
       "0    CAGAACGTCGGCTATCAGAGCAAGC  ATTGTGGAGCGGTTCCTTATGCCAG  \n",
       "1    GCATTGTGGAGCGGTTCCTTATGCC  AGGCTGCCATGTGAGATGATCCAAG  \n",
       "2    AGCAAGCATTGTGGAGCGGTTCCTT  ATGCCAGGCTGCCATGTGAGATGAT  \n",
       "3    TCGGCTATCAGAGCAAGCATTGTGG  AGCGGTTCCTTATGCCAGGCTGCCA  \n",
       "4    GTTGGCCAACAGGTGGCAGAAGAGG  AATTCCTGCCTTCCTCAAGAGGAAC  \n",
       "..                         ...                        ...  \n",
       "367  CACTCCCAGTCTTCCTTTCCCTTCC  AGCAGGGAGTGCCCCCTCCATAAGA  \n",
       "368  GGGGCTGCGGATACCTGGTTTTATT  ATTTTTTCTTTGCCCAACGGGGCCG  \n",
       "369  CTGTATTGGTGCCTCACCTAAGGCT  AAGTATACCTCCccccccacccccc  \n",
       "370  TTCCTCTGTATTGGTGCCTCACCTA  AGGCTAAGTATACCTCCccccccac  \n",
       "371  ATTCCTCTGTATTGGTGCCTCACCT  AAGGCTAAGTATACCTCCcccccca  \n",
       "\n",
       "[372 rows x 16 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9d476c7a-f47e-4389-abf1-8b2a567ce668",
   "metadata": {},
   "outputs": [],
   "source": [
    "## We have a lot of probes here - increasing the offset to 1000 bp to space them out\n",
    "probe_offset = 1000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "3504dcc6-d55f-4491-bcdd-5049eab3b40e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/nfs/team205/is10/miniconda/envs/gene2probe_env/lib/python3.11/site-packages/gene2probe/gene2probe.py:436: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  start = int(probe_df['start'].iloc[i])\n",
      "/nfs/team205/is10/miniconda/envs/gene2probe_env/lib/python3.11/site-packages/gene2probe/gene2probe.py:437: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  end = int(probe_df['end'].iloc[i])\n"
     ]
    }
   ],
   "source": [
    "selected_probes_list = []\n",
    "df = kmers.copy()\n",
    "i = 0\n",
    "while i < n_desired_probes:\n",
    "    selected_probes_list.append(df.iloc[0:1,:])\n",
    "    ## Going for a high offset as we have a large span:\n",
    "    df = remove_overlapping_probes(df,\n",
    "                                   df['name'][0], \n",
    "                                   offset=probe_offset).reset_index(drop=True).copy()\n",
    "    i +=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "069f0bab-21de-4f48-a902-767f77ea4260",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_probes_df = pd.concat(selected_probes_list, axis=0).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6ed7dbf9-aef5-4723-ad14-f6892251e7e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "      <th>transcript_seq_LHS</th>\n",
       "      <th>transcript_seq_RHS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73827196</td>\n",
       "      <td>73827246</td>\n",
       "      <td>XIST_0_6546</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73827196-73827246(-)</td>\n",
       "      <td>CAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATG...</td>\n",
       "      <td>CTGGCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGT...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "      <td>CAGAACGTCGGCTATCAGAGCAAGC</td>\n",
       "      <td>ATTGTGGAGCGGTTCCTTATGCCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850525</td>\n",
       "      <td>73850575</td>\n",
       "      <td>XIST_5_9144</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850525-73850575(-)</td>\n",
       "      <td>GTTGGCCAACAGGTGGCAGAAGAGGAATTCCTGCCTTCCTCAAGAG...</td>\n",
       "      <td>GTTCCTCTTGAGGAAGGCAGGAATTCCTCTTCTGCCACCTGTTGGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>GTTGGCCAACAGGTGGCAGAAGAGG</td>\n",
       "      <td>AATTCCTGCCTTCCTCAAGAGGAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73844799</td>\n",
       "      <td>73844849</td>\n",
       "      <td>XIST_5_3418</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73844799-73844849(-)</td>\n",
       "      <td>TGACCGCAGCCATGCACCTTGGTCAATAATGTGTGTAACTGCACAC...</td>\n",
       "      <td>GGCCGTGTGCAGTTACACACATTATTGACCAAGGTGCATGGCTGCG...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>TGACCGCAGCCATGCACCTTGGTCA</td>\n",
       "      <td>ATAATGTGTGTAACTGCACACGGCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     start       end         name score strand  \\\n",
       "0    chrX  73827196  73827246  XIST_0_6546     .      -   \n",
       "1    chrX  73850525  73850575  XIST_5_9144     .      -   \n",
       "2    chrX  73844799  73844849  XIST_5_3418     .      -   \n",
       "\n",
       "                  kmer_coord  \\\n",
       "0  chrX:73827196-73827246(-)   \n",
       "1  chrX:73850525-73850575(-)   \n",
       "2  chrX:73844799-73844849(-)   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATG...   \n",
       "1  GTTGGCCAACAGGTGGCAGAAGAGGAATTCCTGCCTTCCTCAAGAG...   \n",
       "2  TGACCGCAGCCATGCACCTTGGTCAATAATGTGTGTAACTGCACAC...   \n",
       "\n",
       "                                           probe_seq  GC_content_full  \\\n",
       "0  CTGGCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGT...             0.54   \n",
       "1  GTTCCTCTTGAGGAAGGCAGGAATTCCTCTTCTGCCACCTGTTGGC...             0.54   \n",
       "2  GGCCGTGTGCAGTTACACACATTATTGACCAAGGTGCATGGCTGCG...             0.54   \n",
       "\n",
       "   longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \\\n",
       "0                    2            0.52            0.56              True   \n",
       "1                    2            0.48            0.60              True   \n",
       "2                    2            0.48            0.60              True   \n",
       "\n",
       "          transcript_seq_LHS         transcript_seq_RHS  \n",
       "0  CAGAACGTCGGCTATCAGAGCAAGC  ATTGTGGAGCGGTTCCTTATGCCAG  \n",
       "1  GTTGGCCAACAGGTGGCAGAAGAGG  AATTCCTGCCTTCCTCAAGAGGAAC  \n",
       "2  TGACCGCAGCCATGCACCTTGGTCA  ATAATGTGTGTAACTGCACACGGCC  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_probes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "9ab3f7e6-a39f-417c-8b40-92eb9e3e2908",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATGCCAG\n",
      "GTTGGCCAACAGGTGGCAGAAGAGGAATTCCTGCCTTCCTCAAGAGGAAC\n",
      "TGACCGCAGCCATGCACCTTGGTCAATAATGTGTGTAACTGCACACGGCC\n"
     ]
    }
   ],
   "source": [
    "for seq in selected_probes_df['transcript_seq']:\n",
    "    print(seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52a8e31d-f567-410c-bd05-54b1e2731f6f",
   "metadata": {},
   "source": [
    "And we are done! We have now selected three potential probes for our gene.\n",
    "\n",
    "If our probes are meant to be split, we can additionally generate these columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3ec41fa9-1992-49cc-8f8f-88da8fdd07e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "if split_nt is not None: \n",
    "    ## Make split probes (and add adapters if provided)\n",
    "    selected_probes_df['probe_seq_LHS'] = [(LHS_pref + seq[0:split_nt] + LHS_suff) for seq in selected_probes_df['probe_seq']]\n",
    "    selected_probes_df['probe_seq_RHS'] = [(RHS_pref +seq[split_nt: probe_length] + RHS_suff) for seq in selected_probes_df['probe_seq']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "76b0b300-b3ca-49db-a600-c82961d30ce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Also add gene_ID for completeness\n",
    "selected_probes_df['gene_ID'] = gene_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1ce62661-8a1b-4ef6-83e1-2bd603806975",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Export selected probes as dataframe:\n",
    "selected_probes_df.to_csv((out_dir + 'kmers_selected_probes.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77b84ece-a367-4a6b-87c1-1c2a3858ed8a",
   "metadata": {},
   "source": [
    "We always recommend additionally performing a manual BLAST of these probe sequences to make sure that there are no off-target effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "1e5eb6cd-cc4e-4354-bc83-647c2003a9fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>kmer_coord</th>\n",
       "      <th>transcript_seq</th>\n",
       "      <th>probe_seq</th>\n",
       "      <th>GC_content_full</th>\n",
       "      <th>longest_homopolymer</th>\n",
       "      <th>GC_content_LHS</th>\n",
       "      <th>GC_content_RHS</th>\n",
       "      <th>has_required_nts</th>\n",
       "      <th>transcript_seq_LHS</th>\n",
       "      <th>transcript_seq_RHS</th>\n",
       "      <th>probe_seq_LHS</th>\n",
       "      <th>probe_seq_RHS</th>\n",
       "      <th>gene_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73827196</td>\n",
       "      <td>73827246</td>\n",
       "      <td>XIST_0_6546</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73827196-73827246(-)</td>\n",
       "      <td>CAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATG...</td>\n",
       "      <td>CTGGCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGT...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.52</td>\n",
       "      <td>0.56</td>\n",
       "      <td>True</td>\n",
       "      <td>CAGAACGTCGGCTATCAGAGCAAGC</td>\n",
       "      <td>ATTGTGGAGCGGTTCCTTATGCCAG</td>\n",
       "      <td>CCTTGGCACCCGAGAATTCCACTGGCATAAGGAACCGCTCCACAAT</td>\n",
       "      <td>/5Phos/GCTTGCTCTGATAGCCGACGTTCTGAAAAAAAAAAAAAA...</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73850525</td>\n",
       "      <td>73850575</td>\n",
       "      <td>XIST_5_9144</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73850525-73850575(-)</td>\n",
       "      <td>GTTGGCCAACAGGTGGCAGAAGAGGAATTCCTGCCTTCCTCAAGAG...</td>\n",
       "      <td>GTTCCTCTTGAGGAAGGCAGGAATTCCTCTTCTGCCACCTGTTGGC...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>GTTGGCCAACAGGTGGCAGAAGAGG</td>\n",
       "      <td>AATTCCTGCCTTCCTCAAGAGGAAC</td>\n",
       "      <td>CCTTGGCACCCGAGAATTCCAGTTCCTCTTGAGGAAGGCAGGAATT</td>\n",
       "      <td>/5Phos/CCTCTTCTGCCACCTGTTGGCCAACAAAAAAAAAAAAAA...</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73844799</td>\n",
       "      <td>73844849</td>\n",
       "      <td>XIST_5_3418</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73844799-73844849(-)</td>\n",
       "      <td>TGACCGCAGCCATGCACCTTGGTCAATAATGTGTGTAACTGCACAC...</td>\n",
       "      <td>GGCCGTGTGCAGTTACACACATTATTGACCAAGGTGCATGGCTGCG...</td>\n",
       "      <td>0.54</td>\n",
       "      <td>2</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.60</td>\n",
       "      <td>True</td>\n",
       "      <td>TGACCGCAGCCATGCACCTTGGTCA</td>\n",
       "      <td>ATAATGTGTGTAACTGCACACGGCC</td>\n",
       "      <td>CCTTGGCACCCGAGAATTCCAGGCCGTGTGCAGTTACACACATTAT</td>\n",
       "      <td>/5Phos/TGACCAAGGTGCATGGCTGCGGTCAAAAAAAAAAAAAAA...</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname     start       end         name score strand  \\\n",
       "0    chrX  73827196  73827246  XIST_0_6546     .      -   \n",
       "1    chrX  73850525  73850575  XIST_5_9144     .      -   \n",
       "2    chrX  73844799  73844849  XIST_5_3418     .      -   \n",
       "\n",
       "                  kmer_coord  \\\n",
       "0  chrX:73827196-73827246(-)   \n",
       "1  chrX:73850525-73850575(-)   \n",
       "2  chrX:73844799-73844849(-)   \n",
       "\n",
       "                                      transcript_seq  \\\n",
       "0  CAGAACGTCGGCTATCAGAGCAAGCATTGTGGAGCGGTTCCTTATG...   \n",
       "1  GTTGGCCAACAGGTGGCAGAAGAGGAATTCCTGCCTTCCTCAAGAG...   \n",
       "2  TGACCGCAGCCATGCACCTTGGTCAATAATGTGTGTAACTGCACAC...   \n",
       "\n",
       "                                           probe_seq  GC_content_full  \\\n",
       "0  CTGGCATAAGGAACCGCTCCACAATGCTTGCTCTGATAGCCGACGT...             0.54   \n",
       "1  GTTCCTCTTGAGGAAGGCAGGAATTCCTCTTCTGCCACCTGTTGGC...             0.54   \n",
       "2  GGCCGTGTGCAGTTACACACATTATTGACCAAGGTGCATGGCTGCG...             0.54   \n",
       "\n",
       "   longest_homopolymer  GC_content_LHS  GC_content_RHS  has_required_nts  \\\n",
       "0                    2            0.52            0.56              True   \n",
       "1                    2            0.48            0.60              True   \n",
       "2                    2            0.48            0.60              True   \n",
       "\n",
       "          transcript_seq_LHS         transcript_seq_RHS  \\\n",
       "0  CAGAACGTCGGCTATCAGAGCAAGC  ATTGTGGAGCGGTTCCTTATGCCAG   \n",
       "1  GTTGGCCAACAGGTGGCAGAAGAGG  AATTCCTGCCTTCCTCAAGAGGAAC   \n",
       "2  TGACCGCAGCCATGCACCTTGGTCA  ATAATGTGTGTAACTGCACACGGCC   \n",
       "\n",
       "                                    probe_seq_LHS  \\\n",
       "0  CCTTGGCACCCGAGAATTCCACTGGCATAAGGAACCGCTCCACAAT   \n",
       "1  CCTTGGCACCCGAGAATTCCAGTTCCTCTTGAGGAAGGCAGGAATT   \n",
       "2  CCTTGGCACCCGAGAATTCCAGGCCGTGTGCAGTTACACACATTAT   \n",
       "\n",
       "                                       probe_seq_RHS gene_ID  \n",
       "0  /5Phos/GCTTGCTCTGATAGCCGACGTTCTGAAAAAAAAAAAAAA...    XIST  \n",
       "1  /5Phos/CCTCTTCTGCCACCTGTTGGCCAACAAAAAAAAAAAAAA...    XIST  \n",
       "2  /5Phos/TGACCAAGGTGCATGGCTGCGGTCAAAAAAAAAAAAAAA...    XIST  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_probes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8697e206-e33f-413b-a273-ddf232a32aff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gene2probe_env",
   "language": "python",
   "name": "gene2probe_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
